<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="structToolbox">
<title>Data analysis of metabolomics and other omics datasets using the structToolbox • structToolbox</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Data analysis of metabolomics and other omics datasets using the structToolbox">
<meta property="og:description" content="structToolbox">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">structToolbox</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.15.1</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/data_analysis_omics_using_the_structtoolbox.html">Data analysis of metabolomics and other omics datasets using the structToolbox</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/computational-metabolomics/structToolbox/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Data analysis of metabolomics and other omics datasets using the structToolbox</h1>
                        <h4 data-toc-skip class="author">Gavin R
Lloyd</h4>
            <address class="author_afil">
      Phenome Centre Birmingham, University of Birmingham,
UK<br><a class="author_email" href="mailto:#"></a><a href="mailto:g.r.lloyd@bham.ac.uk" class="email">g.r.lloyd@bham.ac.uk</a>
      </address>
                              <h4 data-toc-skip class="author">Andris
Jankevics</h4>
            <address class="author_afil">
      Phenome Centre Birmingham, University of Birmingham,
UK<br><a class="author_email" href="mailto:#"></a><a href="mailto:a.jankevics@bham.ac.uk" class="email">a.jankevics@bham.ac.uk</a>
      </address>
                              <h4 data-toc-skip class="author">Ralf J
Weber</h4>
            <address class="author_afil">
      Phenome Centre Birmingham, University of Birmingham,
UK<br><a class="author_email" href="mailto:#"></a><a href="mailto:r.j.weber@bham.ac.uk" class="email">r.j.weber@bham.ac.uk</a>
      </address>
                  
      
      <small class="dont-index">Source: <a href="https://github.com/computational-metabolomics/structToolbox/blob/HEAD/vignettes/data_analysis_omics_using_the_structtoolbox.Rmd" class="external-link"><code>vignettes/data_analysis_omics_using_the_structtoolbox.Rmd</code></a></small>
      <div class="d-none name"><code>data_analysis_omics_using_the_structtoolbox.Rmd</code></div>
    </div>

    
        <div class="abstract">
      <p class="abstract">Abstract</p>
      <p>Data (pre-)processing and data analysis of Metabolomics and
      other omics datasets using struct and structToolbox, including
      univariate/multivariate statistics and machine learning
      approaches.</p>
    </div>
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>The ‘structToolbox’ includes an extensive set of data
(pre-)processing and analysis tools for metabolomics and other omics,
with a strong emphasis on statistics and machine learning. The methods
and tools have been implemented using class-based templates available
via the <code>struct</code> (Statistics in R Using Class-based
Templates) package. The aim of this vignette is to introduce the reader
to basic and more advanced structToolbox-based operations and
implementations, such as the use of <code>struct</code> objects,
getting/setting methods/parameters, and building workflows for the
analysis of mass spectrometry (MS) and nuclear magnetic resonance
(NMR)-based Metabolomics and proteomics datasets. The workflows
demonstrated here include a wide range of methods and tools including
pre-processing such as filtering, normalisation and scaling, followed by
univariate and/or multivariate statistics, and machine learning
approaches.</p>
</div>
<div class="section level2">
<h2 id="getting-started">Getting started<a class="anchor" aria-label="anchor" href="#getting-started"></a>
</h2>
<p>The latest version of <code>structToolbox</code> compatible with your
current R version can be installed using <code>BiocManager</code>.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># install BiocManager if not present</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># install structToolbox and dependencies</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"structToolbox"</span><span class="op">)</span></span></code></pre></div>
<p>A number of additional packages are needed for this vignette.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## install additional bioc packages for vignette if needed</span></span>
<span><span class="co">#BiocManager::install(c('pmp', 'ropls', 'BiocFileCache'))</span></span>
<span></span>
<span><span class="co">## install additional CRAN packages if needed</span></span>
<span><span class="co">#install.packages(c('cowplot', 'openxlsx'))</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="op">{</span></span>
<span>    <span class="co"># Bioconductor packages</span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/computational-metabolomics/structToolbox" class="external-link">structToolbox</a></span><span class="op">)</span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">pmp</span><span class="op">)</span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://doi.org/10.1021/acs.jproteome.5b00354" class="external-link">ropls</a></span><span class="op">)</span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">BiocFileCache</span><span class="op">)</span></span>
<span>  </span>
<span>    <span class="co"># CRAN libraries</span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">gridExtra</span><span class="op">)</span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://wilkelab.org/cowplot/" class="external-link">cowplot</a></span><span class="op">)</span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ycphs.github.io/openxlsx/index.html" class="external-link">openxlsx</a></span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co"># use the BiocFileCache</span></span>
<span><span class="va">bfc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocFileCache/man/BiocFileCache-class.html" class="external-link">BiocFileCache</a></span><span class="op">(</span>ask <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="introduction-to-struct-objects-including-models-model-sequences-model-charts-and-ontology-">Introduction to <code>struct</code> objects, including models, model
sequences, model charts and ontology.<a class="anchor" aria-label="anchor" href="#introduction-to-struct-objects-including-models-model-sequences-model-charts-and-ontology-"></a>
</h2>
<div class="section level3">
<h3 id="introduction-1">Introduction<a class="anchor" aria-label="anchor" href="#introduction-1"></a>
</h3>
<p>PCA (Principal Component Analysis) and PLS (Partial Least Squares)
are commonly applied methods for exploring and analysing multivariate
datasets. Here we use these two statistical methods to demonstrate the
different types of <code>struct</code> (STatistics in R Using Class
Templates) objects that are available as part of the
<code>structToolbox</code> and how these objects (i.e. class templates)
can be used to conduct unsupervised and supervised multivariate
statistical analysis.</p>
</div>
<div class="section level3">
<h3 id="dataset">Dataset<a class="anchor" aria-label="anchor" href="#dataset"></a>
</h3>
<p>For demonstration purposes we will use the “Iris” dataset. This
famous (Fisher’s or Anderson’s) dataset contains measurements of sepal
length and width and petal length and width, in centimeters, for 50
flowers from each of 3 class of Iris. The class are Iris setosa,
versicolor, and virginica. See here (<a href="https://stat.ethz.ch/R-manual/R-devel/library/datasets/html/iris.html" class="external-link uri">https://stat.ethz.ch/R-manual/R-devel/library/datasets/html/iris.html</a>)
for more information.</p>
<p>Note: this vignette is also compatible with the Direct infusion mass
spectrometry metabolomics “benchmark” dataset described in Kirwan et
al., <em>Sci Data</em> <em>1</em>, 140012 (2014) (<a href="https://doi.org/10.1038/sdata.2014.12" class="external-link uri">https://doi.org/10.1038/sdata.2014.12</a>).</p>
<p>Both datasets are available as part of the structToolbox package and
already prepared as a <code>DatasetExperiment</code> object.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Iris dataset (comment if using MTBLS79 benchmark data)</span></span>
<span><span class="va">D</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/struct/man/iris_DatasetExperiment.html" class="external-link">iris_DatasetExperiment</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">D</span><span class="op">$</span><span class="va">sample_meta</span><span class="op">$</span><span class="va">class</span> <span class="op">=</span> <span class="va">D</span><span class="op">$</span><span class="va">sample_meta</span><span class="op">$</span><span class="va">Species</span></span>
<span></span>
<span><span class="co">## MTBLS (comment if using Iris data)</span></span>
<span><span class="co"># D = MTBLS79_DatasetExperiment(filtered=TRUE)</span></span>
<span><span class="co"># M = pqn_norm(qc_label='QC',factor_name='sample_type') + </span></span>
<span><span class="co">#   knn_impute(neighbours=5) +</span></span>
<span><span class="co">#   glog_transform(qc_label='QC',factor_name='sample_type') +</span></span>
<span><span class="co">#   filter_smeta(mode='exclude',levels='QC',factor_name='sample_type')</span></span>
<span><span class="co"># M = model_apply(M,D)</span></span>
<span><span class="co"># D = predicted(M)</span></span>
<span></span>
<span><span class="co"># show info</span></span>
<span><span class="va">D</span></span></code></pre></div>
<pre><code><span><span class="co">## A "DatasetExperiment" object</span></span>
<span><span class="co">## ----------------------------</span></span>
<span><span class="co">## name:          Fisher's Iris dataset</span></span>
<span><span class="co">## description:   This famous (Fisher's or Anderson's) iris data set gives the measurements in centimeters of</span></span>
<span><span class="co">##                  the variables sepal length and width and petal length and width,</span></span>
<span><span class="co">##                  respectively, for 50 flowers from each of 3 species of iris. The species are</span></span>
<span><span class="co">##                  Iris setosa, versicolor, and virginica.</span></span>
<span><span class="co">## data:          150 rows x 4 columns</span></span>
<span><span class="co">## sample_meta:   150 rows x 2 columns</span></span>
<span><span class="co">## variable_meta: 4 rows x 1 columns</span></span></code></pre>
<div class="section level4">
<h4 id="datasetexperiment-objects">DatasetExperiment objects<a class="anchor" aria-label="anchor" href="#datasetexperiment-objects"></a>
</h4>
<p>The <code>DatasetExperiment</code> object is an extension of the
<code>SummarizedExperiment</code> class used by the Bioconductor
community. It contains three main parts:</p>
<ol style="list-style-type: decimal">
<li>
<code>data</code> A data frame containing the measured data for each
sample.</li>
<li>
<code>sample_meta</code> A data frame of additional information
related to the samples e.g. group labels.</li>
<li>
<code>variable_meta</code> A data frame of additional information
related to the variables (features) e.g. annotations</li>
</ol>
<p>Like all <code>struct</code> objects it also contains
<code>name</code> and <code>description</code> fields (called “slots” is
R language).</p>
<p>A key difference between <code>DatasetExperiment</code> and
<code>SummarizedExperiment</code> objects is that the data is
transposed. i.e. for <code>DatasetExperiment</code> objects the samples
are in rows and the features are in columns, while the opposite is true
for <code>SummarizedExperiment</code> objects.</p>
<p>All slots are accessible using dollar notation.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># show some data</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">D</span><span class="op">$</span><span class="va">data</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##   Sepal.Length Sepal.Width Petal.Length Petal.Width</span></span>
<span><span class="co">## 1          5.1         3.5          1.4         0.2</span></span>
<span><span class="co">## 2          4.9         3.0          1.4         0.2</span></span>
<span><span class="co">## 3          4.7         3.2          1.3         0.2</span></span>
<span><span class="co">## 4          4.6         3.1          1.5         0.2</span></span>
<span><span class="co">## 5          5.0         3.6          1.4         0.2</span></span>
<span><span class="co">## 6          5.4         3.9          1.7         0.4</span></span></code></pre>
</div>
</div>
<div class="section level3">
<h3 id="using-struct-model-objects">Using <code>struct</code> model objects<a class="anchor" aria-label="anchor" href="#using-struct-model-objects"></a>
</h3>
<div class="section level4">
<h4 id="statistical-models">Statistical models<a class="anchor" aria-label="anchor" href="#statistical-models"></a>
</h4>
<p>Before we can apply e.g. PCA we first need to create a PCA object.
This object contains all the inputs, outputs and methods needed to apply
PCA. We can set parameters such as the number of components when the PCA
model is created, but we can also use dollar notation to change/view it
later.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">P</span> <span class="op">=</span> <span class="fu"><a href="../reference/PCA.html">PCA</a></span><span class="op">(</span>number_components<span class="op">=</span><span class="fl">15</span><span class="op">)</span></span>
<span><span class="va">P</span><span class="op">$</span><span class="va">number_components</span><span class="op">=</span><span class="fl">5</span></span>
<span><span class="va">P</span><span class="op">$</span><span class="va">number_components</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 5</span></span></code></pre>
<p>The inputs for a model can be listed using
<code>param_ids(object)</code>:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/struct/man/param_ids.html" class="external-link">param_ids</a></span><span class="op">(</span><span class="va">P</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "number_components"</span></span></code></pre>
<p>Or a summary of the object can be printed to the console:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">P</span></span></code></pre></div>
<pre><code><span><span class="co">## A "PCA" object</span></span>
<span><span class="co">## --------------</span></span>
<span><span class="co">## name:          Principal Component Analysis (PCA)</span></span>
<span><span class="co">## description:   PCA is a multivariate data reduction technique. It summarises the data in a smaller number of</span></span>
<span><span class="co">##                  Principal Components that maximise variance.</span></span>
<span><span class="co">## input params:  number_components </span></span>
<span><span class="co">## outputs:       scores, loadings, eigenvalues, ssx, correlation, that </span></span>
<span><span class="co">## predicted:     that</span></span>
<span><span class="co">## seq_in:        data</span></span></code></pre>
</div>
<div class="section level4">
<h4 id="model-sequences">Model sequences<a class="anchor" aria-label="anchor" href="#model-sequences"></a>
</h4>
<p>Unless you have good reason not to, it is usually sensible to mean
centre the columns of the data before PCA. Using the <code>STRUCT</code>
framework we can create a model sequence that will mean centre and then
apply PCA to the mean centred data.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">M</span> <span class="op">=</span> <span class="fu"><a href="../reference/mean_centre.html">mean_centre</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="../reference/PCA.html">PCA</a></span><span class="op">(</span>number_components <span class="op">=</span> <span class="fl">4</span><span class="op">)</span></span></code></pre></div>
<p>In <code>structToolbox</code> mean centring and PCA are both model
objects, and joining them using “+” creates a model_sequence object. In
a model_sequence the outputs of the first object (mean centring) are
automatically passed to the inputs of the second object (PCA), which
allows you chain together modelling steps in order to build a
workflow.</p>
<p>The objects in the model_sequence can be accessed by indexing, and we
can combine this with dollar notation. For example, the PCA object is
the second object in our sequence and we can access the number of
components as follows:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">M</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">$</span><span class="va">number_components</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 4</span></span></code></pre>
</div>
<div class="section level4">
<h4 id="trainingtesting-models">Training/testing models<a class="anchor" aria-label="anchor" href="#trainingtesting-models"></a>
</h4>
<p>Model and model_sequence objects need to be trained using data in the
form of a <code>DatasetExperiment</code> object. For example, the PCA
model sequence we created (<code>M</code>) can be trained using the iris
<code>DatasetExperiment</code> object (‘D’).</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">M</span> <span class="op">=</span> <span class="fu"><a href="../reference/model_train.html">model_train</a></span><span class="op">(</span><span class="va">M</span>,<span class="va">D</span><span class="op">)</span></span></code></pre></div>
<p>This model sequence has now mean centred the original data and
calculated the PCA scores and loadings.</p>
<p>Model objects can be used to generate predictions for test datasets.
For the PCA model sequence this involves mean centring the test data
using the mean of training data, and the projecting the centred test
data onto the PCA model using the loadings. The outputs are all stored
in the model sequence and can be accessed using dollar notation. For
this example we will just use the training data again (sometimes called
autoprediction), which for PCA allows us to explore the training data in
more detail.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">M</span> <span class="op">=</span> <span class="fu"><a href="../reference/model_predict.html">model_predict</a></span><span class="op">(</span><span class="va">M</span>,<span class="va">D</span><span class="op">)</span></span></code></pre></div>
<p>Sometimes models don’t make use the training/test approach
e.g. univariate statsitics, filtering etc. For these models the
<code>model_apply</code> method can be used instead. For models that do
provide training/test methods, <code>model_apply</code> applies
autoprediction by default i.e. it is a short-cut for applying
<code>model_train</code> and <code>model_predict</code> to the same
data.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">M</span> <span class="op">=</span> <span class="fu"><a href="../reference/model_apply.html">model_apply</a></span><span class="op">(</span><span class="va">M</span>,<span class="va">D</span><span class="op">)</span></span></code></pre></div>
<p>The available outputs for an object can be listed and accessed like
input params, using dollar notation:</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/struct/man/output_ids.html" class="external-link">output_ids</a></span><span class="op">(</span><span class="va">M</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "scores"      "loadings"    "eigenvalues" "ssx"         "correlation"</span></span>
<span><span class="co">## [6] "that"</span></span></code></pre>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">M</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">$</span><span class="va">scores</span></span></code></pre></div>
<pre><code><span><span class="co">## A "DatasetExperiment" object</span></span>
<span><span class="co">## ----------------------------</span></span>
<span><span class="co">## name:          </span></span>
<span><span class="co">## description:   </span></span>
<span><span class="co">## data:          150 rows x 4 columns</span></span>
<span><span class="co">## sample_meta:   150 rows x 2 columns</span></span>
<span><span class="co">## variable_meta: 4 rows x 1 columns</span></span></code></pre>
</div>
<div class="section level4">
<h4 id="model-charts">Model charts<a class="anchor" aria-label="anchor" href="#model-charts"></a>
</h4>
<p>The <code>struct</code> framework includes chart objects. Charts
associated with a model object can be listed.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/struct/man/chart_names.html" class="external-link">chart_names</a></span><span class="op">(</span><span class="va">M</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "pca_biplot"           "pca_correlation_plot" "pca_dstat_plot"      </span></span>
<span><span class="co">## [4] "pca_loadings_plot"    "pca_scores_plot"      "pca_scree_plot"</span></span></code></pre>
<p>Like model objects, chart objects need to be created before they can
be used. Here we will plot the PCA scores plot for our mean centred PCA
model.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">C</span> <span class="op">=</span> <span class="fu"><a href="../reference/pca_scores_plot.html">pca_scores_plot</a></span><span class="op">(</span>factor_name<span class="op">=</span><span class="st">'class'</span><span class="op">)</span> <span class="co"># colour by class</span></span>
<span><span class="fu"><a href="../reference/chart_plot.html">chart_plot</a></span><span class="op">(</span><span class="va">C</span>,<span class="va">M</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<p><img src="data_analysis_omics_using_the_structtoolbox_files/figure-html/unnamed-chunk-16-1.png"><!-- --></p>
<p>Note that indexing the PCA model is required because the
<code>pca_scores_plot</code> object requires a PCA object as input, not
a model_sequence.</p>
<p>If we make changes to the input parameters of the chart,
<code>chart_plot</code> must be called again to see the effects.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># add petal width to meta data of pca scores</span></span>
<span><span class="va">M</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">$</span><span class="va">scores</span><span class="op">$</span><span class="va">sample_meta</span><span class="op">$</span><span class="va">example</span><span class="op">=</span><span class="va">D</span><span class="op">$</span><span class="va">data</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span></span>
<span><span class="co"># update plot</span></span>
<span><span class="va">C</span><span class="op">$</span><span class="va">factor_name</span><span class="op">=</span><span class="st">'example'</span></span>
<span><span class="fu"><a href="../reference/chart_plot.html">chart_plot</a></span><span class="op">(</span><span class="va">C</span>,<span class="va">M</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<p><img src="data_analysis_omics_using_the_structtoolbox_files/figure-html/unnamed-chunk-17-1.png"><!-- --></p>
<p>The <code>chart_plot</code> method returns a ggplot object so that
you can easily combine it with other plots using the
<code>gridExtra</code> or <code>cowplot</code> packages for example.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># scores plot</span></span>
<span><span class="va">C1</span> <span class="op">=</span> <span class="fu"><a href="../reference/pca_scores_plot.html">pca_scores_plot</a></span><span class="op">(</span>factor_name<span class="op">=</span><span class="st">'class'</span><span class="op">)</span> <span class="co"># colour by class</span></span>
<span><span class="va">g1</span> <span class="op">=</span> <span class="fu"><a href="../reference/chart_plot.html">chart_plot</a></span><span class="op">(</span><span class="va">C1</span>,<span class="va">M</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># scree plot</span></span>
<span><span class="va">C2</span> <span class="op">=</span> <span class="fu"><a href="../reference/pca_scree_plot.html">pca_scree_plot</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">g2</span> <span class="op">=</span> <span class="fu"><a href="../reference/chart_plot.html">chart_plot</a></span><span class="op">(</span><span class="va">C2</span>,<span class="va">M</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># arange in grid</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/gridExtra/man/arrangeGrob.html" class="external-link">grid.arrange</a></span><span class="op">(</span>grobs<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">g1</span>,<span class="va">g2</span><span class="op">)</span>,nrow<span class="op">=</span><span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<p><img src="data_analysis_omics_using_the_structtoolbox_files/figure-html/unnamed-chunk-18-1.png"><!-- --></p>
</div>
<div class="section level4">
<h4 id="ontology">Ontology<a class="anchor" aria-label="anchor" href="#ontology"></a>
</h4>
<p>Within the <code>struct</code> framework (and
<code>structToolbox</code>) an <code>ontology</code> slot is provided to
allow for standardardised definitions for objects and its inputs and
outputs using the Ontology Lookup Service (OLS).</p>
<p>For example, STATO is a general purpose STATistics Ontology (<a href="http://stato-ontology.org" class="external-link uri">http://stato-ontology.org</a>). From the webpage:</p>
<blockquote>
<p>Its aim is to provide coverage for processes such as statistical
tests, their conditions of application, and information needed or
resulting from statistical methods, such as probability distributions,
variables, spread and variation metrics. STATO also covers aspects of
experimental design and description of plots and graphical
representations commonly used to provide visual cues of data
distribution or layout and to assist review of the results.</p>
</blockquote>
<p>The ontology for an object can be set by assigning the ontology term
identifier to the ontology slot of any <code>struct_class</code> object
at design time. The ids can be listed using <code>$</code> notation:</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># create an example PCA object</span></span>
<span><span class="va">P</span><span class="op">=</span><span class="fu"><a href="../reference/PCA.html">PCA</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># ontology for the PCA object</span></span>
<span><span class="va">P</span><span class="op">$</span><span class="va">ontology</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "OBI:0200051"</span></span></code></pre>
<p>The <code>ontology</code> method can be used obtain more detailed
ontology information. When <code>cache = NULL</code> the
<code>struct</code> package will automatically attempt to use the OLS
API (via the <code>rols</code> package) to obtain a name and description
for the provided identifiers. Here we used cached versions of the
ontology definitions provided in the <code>structToolbox</code> package
to prevent issues connecting to the OLS API when building the
package.</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/struct/man/ontology.html" class="external-link">ontology</a></span><span class="op">(</span><span class="va">P</span>,cache <span class="op">=</span> <span class="fu"><a href="../reference/ontology_cache.html">ontology_cache</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="co"># set cache = NULL (default) for online use</span></span></code></pre></div>
<pre><code><span><span class="co">## [[1]]</span></span>
<span><span class="co">## An object of class "ontology_list"</span></span>
<span><span class="co">## Slot "terms":</span></span>
<span><span class="co">## [[1]]</span></span>
<span><span class="co">## term id:       OBI:0200051</span></span>
<span><span class="co">## ontology:      obi</span></span>
<span><span class="co">## label:         principal components analysis dimensionality reduction</span></span>
<span><span class="co">## description:   A principal components analysis dimensionality reduction is a dimensionality reduction</span></span>
<span><span class="co">##                  achieved by applying principal components analysis and by keeping low-order principal</span></span>
<span><span class="co">##                  components and excluding higher-order ones.</span></span>
<span><span class="co">## iri:           http://purl.obolibrary.org/obo/OBI_0200051</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## [[2]]</span></span>
<span><span class="co">## An object of class "ontology_list"</span></span>
<span><span class="co">## Slot "terms":</span></span>
<span><span class="co">## [[1]]</span></span>
<span><span class="co">## term id:       STATO:0000555</span></span>
<span><span class="co">## ontology:      stato</span></span>
<span><span class="co">## label:         number of predictive components</span></span>
<span><span class="co">## description:   number of predictive components is a count used as input to the principle component analysis</span></span>
<span><span class="co">##                  (PCA)</span></span>
<span><span class="co">## iri:           http://purl.obolibrary.org/obo/STATO_0000555</span></span></code></pre>
<p>Note that the <code>ontology</code> method returns definitions for
the object (PCA) and the inputs/outputs (number_of_components).</p>
</div>
</div>
<div class="section level3">
<h3 id="validating-supervised-statistical-models">Validating supervised statistical models<a class="anchor" aria-label="anchor" href="#validating-supervised-statistical-models"></a>
</h3>
<p>Validation is an important aspect of chemometric modelling. The
<code>struct</code> framework enables this kind of iterative model
testing through <code>iterator</code> objects.</p>
<div class="section level4">
<h4 id="cross-validation">Cross-validation<a class="anchor" aria-label="anchor" href="#cross-validation"></a>
</h4>
<p>Cross validation is a common technique for assessing the performance
of classification models. For this example we will use a Partial least
squares-discriminant analysis (PLS-DA) model. Data should be mean
centred prior to PLS, so we will build a model sequence first.</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">M</span> <span class="op">=</span> <span class="fu"><a href="../reference/mean_centre.html">mean_centre</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="../reference/PLSDA.html">PLSDA</a></span><span class="op">(</span>number_components<span class="op">=</span><span class="fl">2</span>,factor_name<span class="op">=</span><span class="st">'class'</span><span class="op">)</span></span>
<span><span class="va">M</span></span></code></pre></div>
<pre><code><span><span class="co">## A model_seq object containing:</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## [1]</span></span>
<span><span class="co">## A "mean_centre" object</span></span>
<span><span class="co">## ----------------------</span></span>
<span><span class="co">## name:          Mean centre</span></span>
<span><span class="co">## description:   The mean sample is subtracted from all samples in the data matrix. The features in the centred</span></span>
<span><span class="co">##                  matrix all have zero mean.</span></span>
<span><span class="co">## input params:  mode </span></span>
<span><span class="co">## outputs:       centred, mean_data, mean_sample_meta </span></span>
<span><span class="co">## predicted:     centred</span></span>
<span><span class="co">## seq_in:        data</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## [2]</span></span>
<span><span class="co">## A "PLSDA" object</span></span>
<span><span class="co">## ----------------</span></span>
<span><span class="co">## name:          Partial least squares discriminant analysis</span></span>
<span><span class="co">## description:   PLS is a multivariate regression technique that extracts latent variables maximising</span></span>
<span><span class="co">##                  covariance between the input data and the response. The Discriminant Analysis</span></span>
<span><span class="co">##                  variant uses group labels in the response variable. For &gt;2 groups a 1-vs-all</span></span>
<span><span class="co">##                  approach is used. Group membership can be predicted for test samples based on</span></span>
<span><span class="co">##                  a probability estimate of group membership, or the estimated y-value.</span></span>
<span><span class="co">## input params:  number_components, factor_name, pred_method </span></span>
<span><span class="co">## outputs:       scores, loadings, yhat, design_matrix, y, reg_coeff, probability, vip, pls_model, pred, threshold, sr, sr_pvalue </span></span>
<span><span class="co">## predicted:     pred</span></span>
<span><span class="co">## seq_in:        data</span></span></code></pre>
<p><code>iterator</code> objects like the k-fold cross-validation object
(<code>kfold_xval</code>) can be created just like any other
<code>struct</code> object. Parameters can be set at creation using the
equals sign, and accessed or changed later using dollar notation.</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># create object</span></span>
<span><span class="va">XCV</span> <span class="op">=</span> <span class="fu"><a href="../reference/kfold_xval.html">kfold_xval</a></span><span class="op">(</span>folds<span class="op">=</span><span class="fl">5</span>,factor_name<span class="op">=</span><span class="st">'class'</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># change the number of folds</span></span>
<span><span class="va">XCV</span><span class="op">$</span><span class="va">folds</span><span class="op">=</span><span class="fl">10</span></span>
<span><span class="va">XCV</span><span class="op">$</span><span class="va">folds</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 10</span></span></code></pre>
<p>The model to be cross-validated can be set/accessed using the
<code>models</code> method.</p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/struct/man/models.html" class="external-link">models</a></span><span class="op">(</span><span class="va">XCV</span><span class="op">)</span><span class="op">=</span><span class="va">M</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/struct/man/models.html" class="external-link">models</a></span><span class="op">(</span><span class="va">XCV</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## A model_seq object containing:</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## [1]</span></span>
<span><span class="co">## A "mean_centre" object</span></span>
<span><span class="co">## ----------------------</span></span>
<span><span class="co">## name:          Mean centre</span></span>
<span><span class="co">## description:   The mean sample is subtracted from all samples in the data matrix. The features in the centred</span></span>
<span><span class="co">##                  matrix all have zero mean.</span></span>
<span><span class="co">## input params:  mode </span></span>
<span><span class="co">## outputs:       centred, mean_data, mean_sample_meta </span></span>
<span><span class="co">## predicted:     centred</span></span>
<span><span class="co">## seq_in:        data</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## [2]</span></span>
<span><span class="co">## A "PLSDA" object</span></span>
<span><span class="co">## ----------------</span></span>
<span><span class="co">## name:          Partial least squares discriminant analysis</span></span>
<span><span class="co">## description:   PLS is a multivariate regression technique that extracts latent variables maximising</span></span>
<span><span class="co">##                  covariance between the input data and the response. The Discriminant Analysis</span></span>
<span><span class="co">##                  variant uses group labels in the response variable. For &gt;2 groups a 1-vs-all</span></span>
<span><span class="co">##                  approach is used. Group membership can be predicted for test samples based on</span></span>
<span><span class="co">##                  a probability estimate of group membership, or the estimated y-value.</span></span>
<span><span class="co">## input params:  number_components, factor_name, pred_method </span></span>
<span><span class="co">## outputs:       scores, loadings, yhat, design_matrix, y, reg_coeff, probability, vip, pls_model, pred, threshold, sr, sr_pvalue </span></span>
<span><span class="co">## predicted:     pred</span></span>
<span><span class="co">## seq_in:        data</span></span></code></pre>
<p>Alternatively, iterators can be combined with models using the
multiplication symbol was shorthand for the <code>models</code>
assignement method:</p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># cross validation of a mean centred PLSDA model</span></span>
<span><span class="va">XCV</span> <span class="op">=</span> <span class="fu"><a href="../reference/kfold_xval.html">kfold_xval</a></span><span class="op">(</span></span>
<span>        folds<span class="op">=</span><span class="fl">5</span>,</span>
<span>        method<span class="op">=</span><span class="st">'venetian'</span>,</span>
<span>        factor_name<span class="op">=</span><span class="st">'class'</span><span class="op">)</span> <span class="op">*</span> </span>
<span>      <span class="op">(</span><span class="fu"><a href="../reference/mean_centre.html">mean_centre</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="../reference/PLSDA.html">PLSDA</a></span><span class="op">(</span>factor_name<span class="op">=</span><span class="st">'class'</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>The <code>run</code> method can be used with any
<code>iterator</code> object. The iterator will then run the set model
or model sequence multiple times.</p>
<p>In this case we run cross-validation 5 times, splitting the data into
different training and test sets each time.</p>
<p>The <code>run</code> method also needs a <code>metric</code> to be
specified, which is another type of <code>struct</code> object. This
metric may be calculated once after all iterations, or after each
iteration, depending on the iterator type (resampling, permutation etc).
For cross-validation we will calculate “balanced accuracy” after all
iterations.</p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">XCV</span> <span class="op">=</span> <span class="fu"><a href="../reference/run.html">run</a></span><span class="op">(</span><span class="va">XCV</span>,<span class="va">D</span>,<span class="fu"><a href="../reference/balanced_accuracy.html">balanced_accuracy</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">XCV</span><span class="op">$</span><span class="va">metric</span></span></code></pre></div>
<pre><code><span><span class="co">##              metric mean sd</span></span>
<span><span class="co">## 1 balanced_accuracy 0.11 NA</span></span></code></pre>
<p>Note The <code>balanced_accuracy</code> metric actually reports
1-accuracy, so a value of 0 indicates perfect performance. The standard
deviation “sd” is NA in this example because there is only one
permutation.</p>
<p>Like other <code>struct</code> objects, iterators can have chart
objects associated with them. The <code>chart_names</code> function will
list them for an object.</p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/struct/man/chart_names.html" class="external-link">chart_names</a></span><span class="op">(</span><span class="va">XCV</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "kfoldxcv_grid"   "kfoldxcv_metric"</span></span></code></pre>
<p>Charts for <code>iterator</code> objects can be plotted in the same
way as charts for any other object.</p>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">C</span> <span class="op">=</span> <span class="fu"><a href="../reference/kfoldxcv_grid.html">kfoldxcv_grid</a></span><span class="op">(</span></span>
<span>  factor_name<span class="op">=</span><span class="st">'class'</span>,</span>
<span>  level<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/levels.html" class="external-link">levels</a></span><span class="op">(</span><span class="va">D</span><span class="op">$</span><span class="va">sample_meta</span><span class="op">$</span><span class="va">class</span><span class="op">)</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span> <span class="co"># first level</span></span>
<span><span class="fu"><a href="../reference/chart_plot.html">chart_plot</a></span><span class="op">(</span><span class="va">C</span>,<span class="va">XCV</span><span class="op">)</span></span></code></pre></div>
<p><img src="data_analysis_omics_using_the_structtoolbox_files/figure-html/unnamed-chunk-27-1.png"><!-- --></p>
<p>It is possible to combine multiple iterators by using the
multiplication symbol. This is equivalent to nesting one iterator inside
the other. For example, we can repeat our cross-validation multiple
times by permuting the sample order.</p>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># permute sample order 10 times and run cross-validation</span></span>
<span><span class="va">P</span> <span class="op">=</span> <span class="fu"><a href="../reference/permute_sample_order.html">permute_sample_order</a></span><span class="op">(</span>number_of_permutations <span class="op">=</span> <span class="fl">10</span><span class="op">)</span> <span class="op">*</span> </span>
<span>    <span class="fu"><a href="../reference/kfold_xval.html">kfold_xval</a></span><span class="op">(</span>folds<span class="op">=</span><span class="fl">5</span>,factor_name<span class="op">=</span><span class="st">'class'</span><span class="op">)</span><span class="op">*</span></span>
<span>    <span class="op">(</span><span class="fu"><a href="../reference/mean_centre.html">mean_centre</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="../reference/PLSDA.html">PLSDA</a></span><span class="op">(</span>factor_name<span class="op">=</span><span class="st">'class'</span>,number_components<span class="op">=</span><span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">P</span> <span class="op">=</span> <span class="fu"><a href="../reference/run.html">run</a></span><span class="op">(</span><span class="va">P</span>,<span class="va">D</span>,<span class="fu"><a href="../reference/balanced_accuracy.html">balanced_accuracy</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">P</span><span class="op">$</span><span class="va">metric</span></span></code></pre></div>
<pre><code><span><span class="co">##              metric   mean          sd</span></span>
<span><span class="co">## 1 balanced_accuracy 0.1095 0.004972145</span></span></code></pre>
</div>
</div>
</div>
<div class="section level2">
<h2 id="a-typical-workflow-for-processing-and-analysing-mass-spectrometry-based-metabolomics-data-">A typical workflow for processing and analysing mass
spectrometry-based metabolomics data.<a class="anchor" aria-label="anchor" href="#a-typical-workflow-for-processing-and-analysing-mass-spectrometry-based-metabolomics-data-"></a>
</h2>
<div class="section level3">
<h3 id="introduction-2">Introduction<a class="anchor" aria-label="anchor" href="#introduction-2"></a>
</h3>
<p>This vignette provides an overview of a <code>structToolbox</code>
workflow implemented to process (e.g. filter features, signal drift and
batch correction, normalise and missing value imputation) mass
spectrometry data. The workflow exists of methods that are part of the
Peak Matrix Processing (<em><a href="https://bioconductor.org/packages/3.19/pmp" class="external-link">pmp</a></em>) package,
including a range of additional filters that are described in Kirwan et
al., <a href="https://link.springer.com/article/10.1007/s00216-013-6856-7" class="external-link">2013</a>,
<a href="https://doi.org/10.1038/sdata.2014.12" class="external-link">2014</a>.</p>
<p>Some packages are required for this vignette in addition
<code>structToolbox</code>:</p>
</div>
<div class="section level3">
<h3 id="dataset-1">Dataset<a class="anchor" aria-label="anchor" href="#dataset-1"></a>
</h3>
<p>For demonstration purposes we will process and analyse the MTBLS79
dataset (‘Dataset 7:SFPM’ Kirwan et al., <a href="https://doi.org/10.1038/sdata.2014.12" class="external-link">2014</a>. This dataset
represents a systematic evaluation of the reproducibility of a
multi-batch direct-infusion mass spectrometry (DIMS)-based metabolomics
study of cardiac tissue extracts. It comprises twenty biological samples
(cow vs. sheep) that were analysed repeatedly, in 8 batches across 7
days, together with a concurrent set of quality control (QC) samples.
Data are presented from each step of the data processing workflow and
are available through MetaboLights (<a href="https://www.ebi.ac.uk/metabolights/MTBLS79" class="external-link uri">https://www.ebi.ac.uk/metabolights/MTBLS79</a>).</p>
<p>The <code>MTBLS79_DatasetExperiment</code> object included in the
<code>structToolbox</code> package is a processed version of the MTBLS79
dataset available in peak matrix processing (<em><a href="https://bioconductor.org/packages/3.19/pmp" class="external-link">pmp</a></em>) package.
This vignette describes step by step how the <code>structToolbox</code>
version was created from the <code>pmp</code> version (i.e. ‘Dataset
7:SFPM’ from the Scientific Data publication - <a href="https://doi.org/10.1038/sdata.2014.12" class="external-link uri">https://doi.org/10.1038/sdata.2014.12</a>).</p>
<p>The <code>SummarizedExperiment</code> object from the
<code>pmp</code> package needs to be converted to a
<code>DatasetExperiment</code> object for use with
<code>structToolbox</code>.</p>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># the pmp SE object</span></span>
<span><span class="va">SE</span> <span class="op">=</span> <span class="va">MTBLS79</span></span>
<span></span>
<span><span class="co"># convert to DE</span></span>
<span><span class="va">DE</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/struct/man/as.DatasetExperiment.html" class="external-link">as.DatasetExperiment</a></span><span class="op">(</span><span class="va">SE</span><span class="op">)</span></span>
<span><span class="va">DE</span><span class="op">$</span><span class="va">name</span> <span class="op">=</span> <span class="st">'MTBLS79'</span></span>
<span><span class="va">DE</span><span class="op">$</span><span class="va">description</span> <span class="op">=</span> <span class="st">'Converted from SE provided by the pmp package'</span></span>
<span></span>
<span><span class="co"># add a column indicating the order the samples were measured in</span></span>
<span><span class="va">DE</span><span class="op">$</span><span class="va">sample_meta</span><span class="op">$</span><span class="va">run_order</span> <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">DE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># add a column indicating if the sample is biological or a QC</span></span>
<span><span class="va">Type</span><span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="va">DE</span><span class="op">$</span><span class="va">sample_meta</span><span class="op">$</span><span class="va">Class</span><span class="op">)</span></span>
<span><span class="va">Type</span><span class="op">[</span><span class="va">Type</span> <span class="op">!=</span> <span class="st">'QC'</span><span class="op">]</span> <span class="op">=</span> <span class="st">'Sample'</span></span>
<span><span class="va">DE</span><span class="op">$</span><span class="va">sample_meta</span><span class="op">$</span><span class="va">Type</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">Type</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># add a column for plotting batches</span></span>
<span><span class="va">DE</span><span class="op">$</span><span class="va">sample_meta</span><span class="op">$</span><span class="va">batch_qc</span> <span class="op">=</span> <span class="va">DE</span><span class="op">$</span><span class="va">sample_meta</span><span class="op">$</span><span class="va">Batch</span></span>
<span><span class="va">DE</span><span class="op">$</span><span class="va">sample_meta</span><span class="op">$</span><span class="va">batch_qc</span><span class="op">[</span><span class="va">DE</span><span class="op">$</span><span class="va">sample_meta</span><span class="op">$</span><span class="va">Type</span><span class="op">==</span><span class="st">'QC'</span><span class="op">]</span><span class="op">=</span><span class="st">'QC'</span></span>
<span></span>
<span><span class="co"># convert to factors</span></span>
<span><span class="va">DE</span><span class="op">$</span><span class="va">sample_meta</span><span class="op">$</span><span class="va">Batch</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">DE</span><span class="op">$</span><span class="va">sample_meta</span><span class="op">$</span><span class="va">Batch</span><span class="op">)</span></span>
<span><span class="va">DE</span><span class="op">$</span><span class="va">sample_meta</span><span class="op">$</span><span class="va">Type</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">DE</span><span class="op">$</span><span class="va">sample_meta</span><span class="op">$</span><span class="va">Type</span><span class="op">)</span></span>
<span><span class="va">DE</span><span class="op">$</span><span class="va">sample_meta</span><span class="op">$</span><span class="va">Class</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">DE</span><span class="op">$</span><span class="va">sample_meta</span><span class="op">$</span><span class="va">Class</span><span class="op">)</span></span>
<span><span class="va">DE</span><span class="op">$</span><span class="va">sample_meta</span><span class="op">$</span><span class="va">batch_qc</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">DE</span><span class="op">$</span><span class="va">sample_meta</span><span class="op">$</span><span class="va">batch_qc</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># print summary</span></span>
<span><span class="va">DE</span></span></code></pre></div>
<pre><code><span><span class="co">## A "DatasetExperiment" object</span></span>
<span><span class="co">## ----------------------------</span></span>
<span><span class="co">## name:          MTBLS79</span></span>
<span><span class="co">## description:   Converted from SE provided by the pmp package</span></span>
<span><span class="co">## data:          172 rows x 2488 columns</span></span>
<span><span class="co">## sample_meta:   172 rows x 7 columns</span></span>
<span><span class="co">## variable_meta: 2488 rows x 0 columns</span></span></code></pre>
<p>Full processing of the data set requires a number of steps. These
will be applied using a single <code>struct</code> model sequence
(<code>model_seq</code>).</p>
</div>
<div class="section level3">
<h3 id="signal-drift-and-batch-correction">Signal drift and batch correction<a class="anchor" aria-label="anchor" href="#signal-drift-and-batch-correction"></a>
</h3>
<p>A batch correction algorithm is applied to reduce intra- and inter-
batch variations in the dataset. Quality Control-Robust Spline
Correction (QC-RSC) is provided in the <code>pmp</code> package, and it
has been wrapped into a <code>structToolbox</code> object called
<code>sb_corr</code>.</p>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">M</span> <span class="op">=</span> <span class="co"># batch correction</span></span>
<span>    <span class="fu"><a href="../reference/sb_corr.html">sb_corr</a></span><span class="op">(</span></span>
<span>      order_col<span class="op">=</span><span class="st">'run_order'</span>,</span>
<span>      batch_col<span class="op">=</span><span class="st">'Batch'</span>, </span>
<span>      qc_col<span class="op">=</span><span class="st">'Type'</span>, </span>
<span>      qc_label<span class="op">=</span><span class="st">'QC'</span>,</span>
<span>      spar_lim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.6</span>,<span class="fl">0.8</span><span class="op">)</span> </span>
<span>    <span class="op">)</span></span>
<span></span>
<span><span class="va">M</span> <span class="op">=</span> <span class="fu"><a href="../reference/model_apply.html">model_apply</a></span><span class="op">(</span><span class="va">M</span>,<span class="va">DE</span><span class="op">)</span></span></code></pre></div>
<p>The figure below shows a plot of a feature vs run order, before and
after the correction. The fitted spline for each batch is shown in grey.
It can be seen that the correction has removed instrument drift within
and between batches.</p>
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">C</span> <span class="op">=</span> <span class="fu"><a href="../reference/feature_profile.html">feature_profile</a></span><span class="op">(</span></span>
<span>      run_order<span class="op">=</span><span class="st">'run_order'</span>,</span>
<span>      qc_label<span class="op">=</span><span class="st">'QC'</span>,</span>
<span>      qc_column<span class="op">=</span><span class="st">'Type'</span>,</span>
<span>      colour_by<span class="op">=</span><span class="st">'batch_qc'</span>,</span>
<span>      feature_to_plot<span class="op">=</span><span class="st">'200.03196'</span>,</span>
<span>      plot_sd<span class="op">=</span><span class="cn">FALSE</span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span><span class="co"># plot and modify using ggplot2 </span></span>
<span><span class="fu"><a href="../reference/chart_plot.html">chart_plot</a></span><span class="op">(</span><span class="va">C</span>,<span class="va">M</span>,<span class="va">DE</span><span class="op">)</span><span class="op">+</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="st">'Peak area'</span><span class="op">)</span><span class="op">+</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">'Before'</span><span class="op">)</span></span></code></pre></div>
<p><img src="data_analysis_omics_using_the_structtoolbox_files/figure-html/unnamed-chunk-31-1.png"><!-- --></p>
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/chart_plot.html">chart_plot</a></span><span class="op">(</span><span class="va">C</span>,<span class="fu"><a href="https://rdrr.io/pkg/struct/man/predicted.html" class="external-link">predicted</a></span><span class="op">(</span><span class="va">M</span><span class="op">)</span><span class="op">)</span><span class="op">+</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="st">'Peak area'</span><span class="op">)</span><span class="op">+</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">'After'</span><span class="op">)</span></span></code></pre></div>
<p><img src="data_analysis_omics_using_the_structtoolbox_files/figure-html/unnamed-chunk-31-2.png"><!-- --></p>
<p>An additional step is added to the published workflow to remove any
feature not corrected by QCRCMS. This can occur if there are not enough
measured QC values within a batch. <code>QCRMS</code> in the
<code>pmp</code> package currently returns NA for all samples in the
feature where this occurs. Features where this occurs will be
excluded.</p>
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">M2</span> <span class="op">=</span> <span class="fu"><a href="../reference/filter_na_count.html">filter_na_count</a></span><span class="op">(</span></span>
<span>      threshold<span class="op">=</span><span class="fl">3</span>,</span>
<span>      factor_name<span class="op">=</span><span class="st">'Batch'</span></span>
<span>    <span class="op">)</span></span>
<span><span class="va">M2</span> <span class="op">=</span> <span class="fu"><a href="../reference/model_apply.html">model_apply</a></span><span class="op">(</span><span class="va">M2</span>,<span class="fu"><a href="https://rdrr.io/pkg/struct/man/predicted.html" class="external-link">predicted</a></span><span class="op">(</span><span class="va">M</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># calculate number of features removed</span></span>
<span><span class="va">nc</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">DE</span><span class="op">)</span> <span class="op">-</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/struct/man/predicted.html" class="external-link">predicted</a></span><span class="op">(</span><span class="va">M2</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/cat.html" class="external-link">cat</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">'Number of features removed: '</span>, <span class="va">nc</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Number of features removed: 425</span></span></code></pre>
<p>The output of this step is the output of
<code>MTBLS79_DatasetExperiment(filtered=FALSE)</code>.</p>
</div>
<div class="section level3">
<h3 id="feature-filtering">Feature filtering<a class="anchor" aria-label="anchor" href="#feature-filtering"></a>
</h3>
<p>In the journal article three spectral cleaning steps are applied. In
the first filter a Kruskal-Wallis test is used to identify features not
reliably detected in the QC samples (p &lt; 0.0001) of all batches. We
follow the same parameters as the original article and do not use
multiple test correction (<code>mtc = 'none'</code>).</p>
<div class="sourceCode" id="cb53"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">M3</span> <span class="op">=</span> <span class="fu"><a href="../reference/kw_rank_sum.html">kw_rank_sum</a></span><span class="op">(</span></span>
<span>      alpha<span class="op">=</span><span class="fl">0.0001</span>,</span>
<span>      mtc<span class="op">=</span><span class="st">'none'</span>,</span>
<span>      factor_names<span class="op">=</span><span class="st">'Batch'</span>,</span>
<span>      predicted<span class="op">=</span><span class="st">'significant'</span></span>
<span>    <span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="../reference/filter_by_name.html">filter_by_name</a></span><span class="op">(</span></span>
<span>      mode<span class="op">=</span><span class="st">'exclude'</span>,</span>
<span>      dimension <span class="op">=</span> <span class="st">'variable'</span>,</span>
<span>      seq_in <span class="op">=</span> <span class="st">'names'</span>, </span>
<span>      names<span class="op">=</span><span class="st">'seq_fcn'</span>, <span class="co"># this is a placeholder and will be replaced by seq_fcn</span></span>
<span>      seq_fcn<span class="op">=</span><span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">{</span><span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="va">x</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span><span class="op">)</span><span class="op">}</span></span>
<span>    <span class="op">)</span></span>
<span><span class="va">M3</span> <span class="op">=</span> <span class="fu"><a href="../reference/model_apply.html">model_apply</a></span><span class="op">(</span><span class="va">M3</span>, <span class="fu"><a href="https://rdrr.io/pkg/struct/man/predicted.html" class="external-link">predicted</a></span><span class="op">(</span><span class="va">M2</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">nc</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/struct/man/predicted.html" class="external-link">predicted</a></span><span class="op">(</span><span class="va">M2</span><span class="op">)</span><span class="op">)</span> <span class="op">-</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/struct/man/predicted.html" class="external-link">predicted</a></span><span class="op">(</span><span class="va">M3</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/cat.html" class="external-link">cat</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">'Number of features removed: '</span>, <span class="va">nc</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Number of features removed: 262</span></span></code></pre>
<p>To make use of univariate tests such as <code>kw_rank_sum</code> as a
filter some advanced features of <code>struct</code> are needed. Slots
<code>predicted</code>, and <code>seq_in</code> are used to ensure the
correct output of the univariate test is connected to the correct input
of a feature filter using <code>filter_by_name</code>. Another slot
<code>seq_fcn</code> is used to extract the relevant column of the
<code>predicted</code> output so that it is compatible with the
<code>seq_in</code> input. A placeholder is used for the “names”
parameter (<code>names = 'place_holder'</code>) as this input will be
replaced by the output from <code>seq_fcn</code>.</p>
<p>The second filter is a Wilcoxon Signed-Rank test. It is used to
identify features that are not representative of the average of the
biological samples (p &lt; 1e-14). Again we make use of
<code>seq_in</code> and <code>seq_fcn</code>.</p>
<div class="sourceCode" id="cb55"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">M4</span> <span class="op">=</span> <span class="fu"><a href="../reference/wilcox_test.html">wilcox_test</a></span><span class="op">(</span></span>
<span>      alpha<span class="op">=</span><span class="fl">1e-14</span>,</span>
<span>      factor_names<span class="op">=</span><span class="st">'Type'</span>, </span>
<span>      mtc<span class="op">=</span><span class="st">'none'</span>, </span>
<span>      predicted <span class="op">=</span> <span class="st">'significant'</span></span>
<span>    <span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="../reference/filter_by_name.html">filter_by_name</a></span><span class="op">(</span></span>
<span>      mode<span class="op">=</span><span class="st">'exclude'</span>,</span>
<span>      dimension<span class="op">=</span><span class="st">'variable'</span>,</span>
<span>      seq_in<span class="op">=</span><span class="st">'names'</span>, </span>
<span>      names<span class="op">=</span><span class="st">'place_holder'</span>,</span>
<span>      seq_fcn<span class="op">=</span><span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">{</span><span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="va">x</span><span class="op">$</span><span class="va">significant</span><span class="op">)</span><span class="op">}</span></span>
<span>    <span class="op">)</span></span>
<span><span class="va">M4</span> <span class="op">=</span> <span class="fu"><a href="../reference/model_apply.html">model_apply</a></span><span class="op">(</span><span class="va">M4</span>, <span class="fu"><a href="https://rdrr.io/pkg/struct/man/predicted.html" class="external-link">predicted</a></span><span class="op">(</span><span class="va">M3</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">nc</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/struct/man/predicted.html" class="external-link">predicted</a></span><span class="op">(</span><span class="va">M3</span><span class="op">)</span><span class="op">)</span> <span class="op">-</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/struct/man/predicted.html" class="external-link">predicted</a></span><span class="op">(</span><span class="va">M4</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/cat.html" class="external-link">cat</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">'Number of features removed: '</span>, <span class="va">nc</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Number of features removed: 169</span></span></code></pre>
<p>Finally, an RSD filter is used to remove features with high
analytical variation (QC RSD &gt; 20 removed)</p>
<div class="sourceCode" id="cb57"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">M5</span> <span class="op">=</span> <span class="fu"><a href="../reference/rsd_filter.html">rsd_filter</a></span><span class="op">(</span></span>
<span>     rsd_threshold<span class="op">=</span><span class="fl">20</span>,</span>
<span>     factor_name<span class="op">=</span><span class="st">'Type'</span></span>
<span><span class="op">)</span></span>
<span><span class="va">M5</span> <span class="op">=</span> <span class="fu"><a href="../reference/model_apply.html">model_apply</a></span><span class="op">(</span><span class="va">M5</span>,<span class="fu"><a href="https://rdrr.io/pkg/struct/man/predicted.html" class="external-link">predicted</a></span><span class="op">(</span><span class="va">M4</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">nc</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/struct/man/predicted.html" class="external-link">predicted</a></span><span class="op">(</span><span class="va">M4</span><span class="op">)</span><span class="op">)</span> <span class="op">-</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/struct/man/predicted.html" class="external-link">predicted</a></span><span class="op">(</span><span class="va">M5</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/cat.html" class="external-link">cat</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">'Number of features removed: '</span>, <span class="va">nc</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Number of features removed: 53</span></span></code></pre>
<p>The output of this filter is the output of
<code>MTBLS79_DatasetExperiment(filtered=TRUE)</code>.</p>
</div>
<div class="section level3">
<h3 id="normalisation-missing-value-imputation-and-scaling">Normalisation, missing value imputation and scaling<a class="anchor" aria-label="anchor" href="#normalisation-missing-value-imputation-and-scaling"></a>
</h3>
<p>We will apply a number of common pre-processing steps to the filtered
peak matrix that are identical to steps applied in are described in
Kirwan et al. <a href="https://link.springer.com/article/10.1007/s00216-013-6856-7" class="external-link">2013</a>,
<a href="https://doi.org/10.1038/sdata.2014.12" class="external-link">2014</a>.</p>
<ul>
<li>Probabilistic Quotient Normalisation (PQN)</li>
<li>k-nearest neighbours imputation (k = 5)</li>
<li>Generalised log transform (glog)</li>
</ul>
<p>These steps prepare the data for multivariate analysis by accounting
for sample concentration differences, imputing missing values and
scaling the data.</p>
<div class="sourceCode" id="cb59"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># peak matrix processing</span></span>
<span><span class="va">M6</span> <span class="op">=</span> <span class="fu"><a href="../reference/pqn_norm.html">pqn_norm</a></span><span class="op">(</span>qc_label<span class="op">=</span><span class="st">'QC'</span>,factor_name<span class="op">=</span><span class="st">'Type'</span><span class="op">)</span> <span class="op">+</span> </span>
<span>     <span class="fu"><a href="../reference/knn_impute.html">knn_impute</a></span><span class="op">(</span>neighbours<span class="op">=</span><span class="fl">5</span><span class="op">)</span> <span class="op">+</span></span>
<span>     <span class="fu"><a href="../reference/glog_transform.html">glog_transform</a></span><span class="op">(</span>qc_label<span class="op">=</span><span class="st">'QC'</span>,factor_name<span class="op">=</span><span class="st">'Type'</span><span class="op">)</span></span>
<span><span class="va">M6</span> <span class="op">=</span> <span class="fu"><a href="../reference/model_apply.html">model_apply</a></span><span class="op">(</span><span class="va">M6</span>,<span class="fu"><a href="https://rdrr.io/pkg/struct/man/predicted.html" class="external-link">predicted</a></span><span class="op">(</span><span class="va">M5</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="exploratory-analysis">Exploratory Analysis<a class="anchor" aria-label="anchor" href="#exploratory-analysis"></a>
</h3>
<p>Principal Component Analysis (PCA) can be used to visualise
high-dimensional data. It is an unsupervised method that maximises
variance in a reduced number of latent variables, or principal
components.</p>
<div class="sourceCode" id="cb60"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># PCA</span></span>
<span><span class="va">M7</span>  <span class="op">=</span> <span class="fu"><a href="../reference/mean_centre.html">mean_centre</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="../reference/PCA.html">PCA</a></span><span class="op">(</span>number_components <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># apply model sequence to data</span></span>
<span><span class="va">M7</span> <span class="op">=</span> <span class="fu"><a href="../reference/model_apply.html">model_apply</a></span><span class="op">(</span><span class="va">M7</span>,<span class="fu"><a href="https://rdrr.io/pkg/struct/man/predicted.html" class="external-link">predicted</a></span><span class="op">(</span><span class="va">M6</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># plot pca scores</span></span>
<span><span class="va">C</span> <span class="op">=</span> <span class="fu"><a href="../reference/pca_scores_plot.html">pca_scores_plot</a></span><span class="op">(</span>factor_name<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'Sample_Rep'</span>,<span class="st">'Class'</span><span class="op">)</span>,ellipse<span class="op">=</span><span class="st">'none'</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/chart_plot.html">chart_plot</a></span><span class="op">(</span><span class="va">C</span>,<span class="va">M7</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_fixed.html" class="external-link">coord_fixed</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guides.html" class="external-link">guides</a></span><span class="op">(</span>colour<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Warning: The `&lt;scale&gt;` argument of `guides()` cannot be `FALSE`. Use "none" instead as</span></span>
<span><span class="co">## of ggplot2 3.3.4.</span></span>
<span><span class="co">## <span style="color: #555555;">This warning is displayed once every 8 hours.</span></span></span>
<span><span class="co">## <span style="color: #555555;">Call `lifecycle::last_lifecycle_warnings()` to see where this warning was</span></span></span>
<span><span class="co">## <span style="color: #555555;">generated.</span></span></span></code></pre>
<p><img src="data_analysis_omics_using_the_structtoolbox_files/figure-html/unnamed-chunk-37-1.png"><!-- --></p>
<p>This plot is very similar to Figure 3b of the original publication <a href="https://www.nature.com/articles/sdata201412/figures/3" class="external-link">link</a>.
Sample replicates are represented by colours and samples groups (C = cow
and S = Sheep) by different shapes.</p>
<p>Plotting the scores and colouring by Batch indicates that the
signal/batch correction was effective as all batches are
overlapping.</p>
<div class="sourceCode" id="cb62"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># chart object</span></span>
<span><span class="va">C</span> <span class="op">=</span> <span class="fu"><a href="../reference/pca_scores_plot.html">pca_scores_plot</a></span><span class="op">(</span>factor_name<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'Batch'</span><span class="op">)</span>,ellipse<span class="op">=</span><span class="st">'none'</span><span class="op">)</span></span>
<span><span class="co"># plot</span></span>
<span><span class="fu"><a href="../reference/chart_plot.html">chart_plot</a></span><span class="op">(</span><span class="va">C</span>,<span class="va">M7</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_fixed.html" class="external-link">coord_fixed</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="data_analysis_omics_using_the_structtoolbox_files/figure-html/unnamed-chunk-38-1.png"><!-- --></p>
</div>
</div>
<div class="section level2">
<h2 id="partial-least-squares-pls-analysis-of-a-untargeted-lc-ms-based-clinical-metabolomics-dataset-">Partial Least Squares (PLS) analysis of a untargeted LC-MS-based
clinical metabolomics dataset.<a class="anchor" aria-label="anchor" href="#partial-least-squares-pls-analysis-of-a-untargeted-lc-ms-based-clinical-metabolomics-dataset-"></a>
</h2>
<div class="section level3">
<h3 id="introduction-3">Introduction<a class="anchor" aria-label="anchor" href="#introduction-3"></a>
</h3>
<p>The aim of this vignette is to demonstrate how to 1) apply and
validate Partial Least Squares (PLS) analysis using the structToolbox,
2) reproduce statistical analysis in <a href="https://pubs.acs.org/doi/10.1021/acs.jproteome.5b00354" class="external-link">Thevenot
et al. (2015)</a> and 3. compare different implementations of PLS.</p>
</div>
<div class="section level3">
<h3 id="dataset-2">Dataset<a class="anchor" aria-label="anchor" href="#dataset-2"></a>
</h3>
<p>The objective of the original study was to: &gt; …study the influence
of age, body mass index (bmi), and gender on metabolite concentrations
in urine, by analysing 183 samples from a cohort of adults with liquid
chromatography coupled to high-resolution mass spectrometry.</p>
<p><a href="https://pubs.acs.org/doi/10.1021/acs.jproteome.5b00354" class="external-link">Thevenot
et al. (2015)</a></p>
<p>The “Sacurine” dataset needs to be converted to a
<code>DatasetExperiment</code> object. The <em><a href="https://bioconductor.org/packages/3.19/ropls" class="external-link">ropls</a></em>
package provides the data as a list containing a
<code>dataMatrix</code>, <code>sampleMetadata</code> and
<code>variableMetadata</code>.</p>
<div class="sourceCode" id="cb63"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">'sacurine'</span>,package <span class="op">=</span> <span class="st">'ropls'</span><span class="op">)</span></span>
<span><span class="co"># the 'sacurine' list should now be available</span></span>
<span></span>
<span><span class="co"># move the annotations to a new column and rename the features by index to avoid issues</span></span>
<span><span class="co"># later when data.frames get transposed and names get checked/changed</span></span>
<span><span class="va">sacurine</span><span class="op">$</span><span class="va">variableMetadata</span><span class="op">$</span><span class="va">annotation</span><span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">sacurine</span><span class="op">$</span><span class="va">variableMetadata</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">sacurine</span><span class="op">$</span><span class="va">variableMetadata</span><span class="op">)</span><span class="op">=</span><span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">sacurine</span><span class="op">$</span><span class="va">variableMetadata</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">sacurine</span><span class="op">$</span><span class="va">dataMatrix</span><span class="op">)</span><span class="op">=</span><span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">sacurine</span><span class="op">$</span><span class="va">dataMatrix</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># create DatasetExperiment</span></span>
<span><span class="va">DE</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/struct/man/struct_DatasetExperiment.html" class="external-link">DatasetExperiment</a></span><span class="op">(</span>data <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="va">sacurine</span><span class="op">$</span><span class="va">dataMatrix</span><span class="op">)</span>,</span>
<span>                       sample_meta <span class="op">=</span> <span class="va">sacurine</span><span class="op">$</span><span class="va">sampleMetadata</span>,</span>
<span>                       variable_meta <span class="op">=</span> <span class="va">sacurine</span><span class="op">$</span><span class="va">variableMetadata</span>,</span>
<span>                       name <span class="op">=</span> <span class="st">'Sacurine data'</span>,</span>
<span>                       description <span class="op">=</span> <span class="st">'See ropls package documentation for details'</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># print summary</span></span>
<span><span class="va">DE</span></span></code></pre></div>
<pre><code><span><span class="co">## A "DatasetExperiment" object</span></span>
<span><span class="co">## ----------------------------</span></span>
<span><span class="co">## name:          Sacurine data</span></span>
<span><span class="co">## description:   See ropls package documentation for details</span></span>
<span><span class="co">## data:          183 rows x 109 columns</span></span>
<span><span class="co">## sample_meta:   183 rows x 3 columns</span></span>
<span><span class="co">## variable_meta: 109 rows x 4 columns</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="data-preprocessing">Data preprocessing<a class="anchor" aria-label="anchor" href="#data-preprocessing"></a>
</h3>
<p>The Sacurine dataset used within this vignette has already been
pre-processed:</p>
<blockquote>
<p>After signal drift and batch effect correction of intensities, each
urine profile was normalized to the osmolality of the sample. Finally,
the data were log10 transformed.</p>
</blockquote>
</div>
<div class="section level3">
<h3 id="exploratory-data-analysis">Exploratory data analysis<a class="anchor" aria-label="anchor" href="#exploratory-data-analysis"></a>
</h3>
<p>Since the data has already been processed the data can be visualised
using Principal Component Analysis (PCA) without further pre-processing.
The <code>ropls</code> package automatically applies unit variance
scaling (autoscaling) by default. The same approach is applied here.</p>
<div class="sourceCode" id="cb65"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># prepare model sequence</span></span>
<span><span class="va">M</span> <span class="op">=</span> <span class="fu"><a href="../reference/autoscale.html">autoscale</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="../reference/PCA.html">PCA</a></span><span class="op">(</span>number_components <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span>
<span><span class="co"># apply model sequence to dataset</span></span>
<span><span class="va">M</span> <span class="op">=</span> <span class="fu"><a href="../reference/model_apply.html">model_apply</a></span><span class="op">(</span><span class="va">M</span>,<span class="va">DE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># pca scores plots</span></span>
<span><span class="va">g</span><span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">k</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">DE</span><span class="op">$</span><span class="va">sample_meta</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">C</span> <span class="op">=</span> <span class="fu"><a href="../reference/pca_scores_plot.html">pca_scores_plot</a></span><span class="op">(</span>factor_name <span class="op">=</span> <span class="va">k</span><span class="op">)</span></span>
<span>    <span class="va">g</span><span class="op">[[</span><span class="va">k</span><span class="op">]</span><span class="op">]</span> <span class="op">=</span> <span class="fu"><a href="../reference/chart_plot.html">chart_plot</a></span><span class="op">(</span><span class="va">C</span>,<span class="va">M</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="co"># plot using cowplot</span></span>
<span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html" class="external-link">plot_grid</a></span><span class="op">(</span>plotlist<span class="op">=</span><span class="va">g</span>, nrow<span class="op">=</span><span class="fl">1</span>, align<span class="op">=</span><span class="st">'vh'</span>, labels<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'A'</span>,<span class="st">'B'</span>,<span class="st">'C'</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="data_analysis_omics_using_the_structtoolbox_files/figure-html/unnamed-chunk-40-1.png"><!-- --></p>
<p>The third plot coloured by gender (C) is identical to Figure 2 of the
<em><a href="https://bioconductor.org/packages/3.19/ropls" class="external-link">ropls</a></em>
package vignette. The <code>structToolbox</code> package provides a
range of PCA-related diagnostic plots, including D-statistic, scree, and
loadings plots. These plots can be used to further explore the variance
of the data.</p>
<div class="sourceCode" id="cb66"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">C</span> <span class="op">=</span> <span class="fu"><a href="../reference/pca_scree_plot.html">pca_scree_plot</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">g1</span> <span class="op">=</span> <span class="fu"><a href="../reference/chart_plot.html">chart_plot</a></span><span class="op">(</span><span class="va">C</span>,<span class="va">M</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="va">C</span> <span class="op">=</span> <span class="fu"><a href="../reference/pca_loadings_plot.html">pca_loadings_plot</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">g2</span> <span class="op">=</span> <span class="fu"><a href="../reference/chart_plot.html">chart_plot</a></span><span class="op">(</span><span class="va">C</span>,<span class="va">M</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="va">C</span> <span class="op">=</span> <span class="fu"><a href="../reference/pca_dstat_plot.html">pca_dstat_plot</a></span><span class="op">(</span>alpha<span class="op">=</span><span class="fl">0.95</span><span class="op">)</span></span>
<span><span class="va">g3</span> <span class="op">=</span> <span class="fu"><a href="../reference/chart_plot.html">chart_plot</a></span><span class="op">(</span><span class="va">C</span>,<span class="va">M</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="va">p1</span><span class="op">=</span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html" class="external-link">plot_grid</a></span><span class="op">(</span>plotlist <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">g1</span>,<span class="va">g2</span><span class="op">)</span>,align<span class="op">=</span><span class="st">'h'</span>,nrow<span class="op">=</span><span class="fl">1</span>,axis<span class="op">=</span><span class="st">'b'</span><span class="op">)</span></span>
<span><span class="va">p2</span><span class="op">=</span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html" class="external-link">plot_grid</a></span><span class="op">(</span>plotlist <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">g3</span><span class="op">)</span>,nrow<span class="op">=</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html" class="external-link">plot_grid</a></span><span class="op">(</span><span class="va">p1</span>,<span class="va">p2</span>,nrow<span class="op">=</span><span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<p><img src="data_analysis_omics_using_the_structtoolbox_files/figure-html/unnamed-chunk-41-1.png"><!-- --></p>
</div>
<div class="section level3">
<h3 id="partial-least-squares-pls-analysis">Partial Least Squares (PLS) analysis<a class="anchor" aria-label="anchor" href="#partial-least-squares-pls-analysis"></a>
</h3>
<p>The <code>ropls</code> package uses its own implementation of the
(O)PLS algorithms. <code>structToolbox</code> uses the <code>pls</code>
package, so it is interesting to compare the outputs from both
approaches. For simplicity only the scores plots are compared.</p>
<div class="sourceCode" id="cb67"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># prepare model sequence</span></span>
<span><span class="va">M</span> <span class="op">=</span> <span class="fu"><a href="../reference/autoscale.html">autoscale</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="../reference/PLSDA.html">PLSDA</a></span><span class="op">(</span>factor_name<span class="op">=</span><span class="st">'gender'</span><span class="op">)</span></span>
<span><span class="va">M</span> <span class="op">=</span> <span class="fu"><a href="../reference/model_apply.html">model_apply</a></span><span class="op">(</span><span class="va">M</span>,<span class="va">DE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">C</span> <span class="op">=</span> <span class="fu"><a href="../reference/pls_scores_plot.html">pls_scores_plot</a></span><span class="op">(</span>factor_name <span class="op">=</span> <span class="st">'gender'</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/chart_plot.html">chart_plot</a></span><span class="op">(</span><span class="va">C</span>,<span class="va">M</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<p><img src="data_analysis_omics_using_the_structtoolbox_files/figure-html/unnamed-chunk-42-1.png"><!-- --></p>
<p>The plot is similar to fig.3 of the <em><a href="https://bioconductor.org/packages/3.19/ropls" class="external-link">ropls</a></em>
vignette. Differences are due to inverted LV axes, a common occurrence
with the NIPALS algorithm (used by both <code>structToolbox</code> and
<code>ropls</code>) which depends on how the algorithm is
initialised.</p>
<p>To compare the R2 values for the model in structToolbox we have to
use a regression model, instead of a discriminant model. For this we
convert the gender factor to a numeric variable before applying the
model.</p>
<div class="sourceCode" id="cb68"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># convert gender to numeric</span></span>
<span><span class="va">DE</span><span class="op">$</span><span class="va">sample_meta</span><span class="op">$</span><span class="va">gender</span><span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="va">DE</span><span class="op">$</span><span class="va">sample_meta</span><span class="op">$</span><span class="va">gender</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># models sequence</span></span>
<span><span class="va">M</span> <span class="op">=</span> <span class="fu"><a href="../reference/autoscale.html">autoscale</a></span><span class="op">(</span>mode<span class="op">=</span><span class="st">'both'</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="../reference/PLSR.html">PLSR</a></span><span class="op">(</span>factor_name<span class="op">=</span><span class="st">'gender'</span>,number_components<span class="op">=</span><span class="fl">3</span><span class="op">)</span></span>
<span><span class="va">M</span> <span class="op">=</span> <span class="fu"><a href="../reference/model_apply.html">model_apply</a></span><span class="op">(</span><span class="va">M</span>,<span class="va">DE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># some diagnostic charts</span></span>
<span><span class="va">C</span> <span class="op">=</span> <span class="fu"><a href="../reference/plsr_cook_dist.html">plsr_cook_dist</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">g1</span> <span class="op">=</span> <span class="fu"><a href="../reference/chart_plot.html">chart_plot</a></span><span class="op">(</span><span class="va">C</span>,<span class="va">M</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="va">C</span> <span class="op">=</span> <span class="fu"><a href="../reference/plsr_prediction_plot.html">plsr_prediction_plot</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">g2</span> <span class="op">=</span> <span class="fu"><a href="../reference/chart_plot.html">chart_plot</a></span><span class="op">(</span><span class="va">C</span>,<span class="va">M</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="va">C</span> <span class="op">=</span> <span class="fu"><a href="../reference/plsr_qq_plot.html">plsr_qq_plot</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">g3</span> <span class="op">=</span> <span class="fu"><a href="../reference/chart_plot.html">chart_plot</a></span><span class="op">(</span><span class="va">C</span>,<span class="va">M</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="va">C</span> <span class="op">=</span> <span class="fu"><a href="../reference/plsr_residual_hist.html">plsr_residual_hist</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">g4</span> <span class="op">=</span> <span class="fu"><a href="../reference/chart_plot.html">chart_plot</a></span><span class="op">(</span><span class="va">C</span>,<span class="va">M</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html" class="external-link">plot_grid</a></span><span class="op">(</span>plotlist <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">g1</span>,<span class="va">g2</span>,<span class="va">g3</span>,<span class="va">g4</span><span class="op">)</span>, nrow<span class="op">=</span><span class="fl">2</span>,align<span class="op">=</span><span class="st">'vh'</span><span class="op">)</span></span></code></pre></div>
<p><img src="data_analysis_omics_using_the_structtoolbox_files/figure-html/unnamed-chunk-43-1.png"><!-- --></p>
<p>The <code>ropls</code> package automatically applies cross-validation
to asses the performance of the PLSDA model. In
<code>structToolbox</code> this is applied separately to give more
control over the approach used if desired. The default cross-validation
used by the <code>ropls</code> package is 7-fold cross-validation and we
replicate that here.</p>
<div class="sourceCode" id="cb69"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># model sequence</span></span>
<span><span class="va">M</span> <span class="op">=</span> <span class="fu"><a href="../reference/kfold_xval.html">kfold_xval</a></span><span class="op">(</span>folds<span class="op">=</span><span class="fl">7</span>, factor_name<span class="op">=</span><span class="st">'gender'</span><span class="op">)</span> <span class="op">*</span> </span>
<span>    <span class="op">(</span><span class="fu"><a href="../reference/autoscale.html">autoscale</a></span><span class="op">(</span>mode<span class="op">=</span><span class="st">'both'</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="../reference/PLSR.html">PLSR</a></span><span class="op">(</span>factor_name<span class="op">=</span><span class="st">'gender'</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">M</span> <span class="op">=</span> <span class="fu"><a href="../reference/run.html">run</a></span><span class="op">(</span><span class="va">M</span>,<span class="va">DE</span>,<span class="fu"><a href="../reference/r_squared.html">r_squared</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Training set R2: 0.6975706 0.6798415 0.646671 0.6532914 0.7109769
0.670777 0.6935344</p>
<p>Test set Q2: 0.5460723</p>
<p>The validity of the model can further be assessed using permutation
testing. For this we will return to a discriminant model.</p>
<div class="sourceCode" id="cb70"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># reset gender to original factor</span></span>
<span><span class="va">DE</span><span class="op">$</span><span class="va">sample_meta</span><span class="op">$</span><span class="va">gender</span><span class="op">=</span><span class="va">sacurine</span><span class="op">$</span><span class="va">sampleMetadata</span><span class="op">$</span><span class="va">gender</span></span>
<span><span class="co"># model sequence</span></span>
<span><span class="va">M</span> <span class="op">=</span> <span class="fu"><a href="../reference/permutation_test.html">permutation_test</a></span><span class="op">(</span>number_of_permutations <span class="op">=</span> <span class="fl">10</span>, factor_name<span class="op">=</span><span class="st">'gender'</span><span class="op">)</span> <span class="op">*</span></span>
<span>    <span class="fu"><a href="../reference/kfold_xval.html">kfold_xval</a></span><span class="op">(</span>folds<span class="op">=</span><span class="fl">7</span>,factor_name<span class="op">=</span><span class="st">'gender'</span><span class="op">)</span> <span class="op">*</span></span>
<span>    <span class="op">(</span><span class="fu"><a href="../reference/autoscale.html">autoscale</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="../reference/PLSDA.html">PLSDA</a></span><span class="op">(</span>factor_name<span class="op">=</span><span class="st">'gender'</span>,number_components <span class="op">=</span> <span class="fl">3</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">M</span> <span class="op">=</span> <span class="fu"><a href="../reference/run.html">run</a></span><span class="op">(</span><span class="va">M</span>,<span class="va">DE</span>,<span class="fu"><a href="../reference/balanced_accuracy.html">balanced_accuracy</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">C</span> <span class="op">=</span> <span class="fu"><a href="../reference/permutation_test_plot.html">permutation_test_plot</a></span><span class="op">(</span>style<span class="op">=</span><span class="st">'boxplot'</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/chart_plot.html">chart_plot</a></span><span class="op">(</span><span class="va">C</span>,<span class="va">M</span><span class="op">)</span><span class="op">+</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="st">'1 - balanced accuracy'</span><span class="op">)</span></span></code></pre></div>
<p><img src="data_analysis_omics_using_the_structtoolbox_files/figure-html/unnamed-chunk-46-1.png"><!-- --></p>
<p>The permuted models have a balanced accuracy of around 50%, which is
to be expected for a dataset with two groups. The unpermuted models have
a balanced accuracy of around 90% and is therefore much better than
might be expected to occur by chance.</p>
</div>
</div>
<div class="section level2">
<h2 id="univariate-and-multivariate-statistical-analysis-of-a-nmr-based-clinical-metabolomics-dataset-">Univariate and multivariate statistical analysis of a NMR-based
clinical metabolomics dataset.<a class="anchor" aria-label="anchor" href="#univariate-and-multivariate-statistical-analysis-of-a-nmr-based-clinical-metabolomics-dataset-"></a>
</h2>
<div class="section level3">
<h3 id="introduction-4">Introduction<a class="anchor" aria-label="anchor" href="#introduction-4"></a>
</h3>
<p>The purpose of this vignette is to demonstrate the different
functionalities and methods that are available as part of the
structToolbox and reproduce the data analysis reported in <a href="https://link.springer.com/article/10.1007/s11306-019-1588-0" class="external-link">Mendez
et al., (2020)</a> and <a href="https://www.nature.com/articles/bjc2015414" class="external-link">Chan et al.,
(2016)</a>.</p>
</div>
<div class="section level3">
<h3 id="dataset-3">Dataset<a class="anchor" aria-label="anchor" href="#dataset-3"></a>
</h3>
<p>The 1H-NMR dataset used and described in <a href="https://link.springer.com/article/10.1007/s11306-019-1588-0" class="external-link">Mendez
et al., (2020)</a> and in this vignette contains processed spectra of
urine samples obtained from gastric cancer and healthy patients <a href="https://www.nature.com/articles/bjc2015414" class="external-link">Chan et al.,
(2016)</a>. The experimental raw data is available through Metabolomics
Workbench (<a href="http://dx.doi.org/10.21228/M8B10B" class="external-link">PR000699</a>) and
the processed version is available from <a href="https://github.com/CIMCB/MetabWorkflowTutorial/raw/master/GastricCancer_NMR.xlsx" class="external-link">here</a>
as an Excel data file.</p>
<p>As a first step we need to reorganise and convert the Excel data file
into a DatasetExperiment object. Using the <code>openxlsx</code> package
the file can be read directly into an R <code>data.frame</code> and then
manipulated as required.</p>
<div class="sourceCode" id="cb71"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">url</span> <span class="op">=</span> <span class="st">'https://github.com/CIMCB/MetabWorkflowTutorial/raw/master/GastricCancer_NMR.xlsx'</span></span>
<span></span>
<span><span class="co"># read in file directly from github...</span></span>
<span><span class="co"># X=read.xlsx(url)</span></span>
<span></span>
<span><span class="co"># ...or use BiocFileCache</span></span>
<span><span class="va">path</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocFileCache/man/BiocFileCache-class.html" class="external-link">bfcrpath</a></span><span class="op">(</span><span class="va">bfc</span>,<span class="va">url</span><span class="op">)</span></span>
<span><span class="va">X</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/openxlsx/man/read.xlsx.html" class="external-link">read.xlsx</a></span><span class="op">(</span><span class="va">path</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># sample meta data</span></span>
<span><span class="va">SM</span><span class="op">=</span><span class="va">X</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">SM</span><span class="op">)</span><span class="op">=</span><span class="va">SM</span><span class="op">$</span><span class="va">SampleID</span></span>
<span><span class="co"># convert to factors</span></span>
<span><span class="va">SM</span><span class="op">$</span><span class="va">SampleType</span><span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">SM</span><span class="op">$</span><span class="va">SampleType</span><span class="op">)</span></span>
<span><span class="va">SM</span><span class="op">$</span><span class="va">Class</span><span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">SM</span><span class="op">$</span><span class="va">Class</span><span class="op">)</span></span>
<span><span class="co"># keep a numeric version of class for regression</span></span>
<span><span class="va">SM</span><span class="op">$</span><span class="va">Class_num</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="va">SM</span><span class="op">$</span><span class="va">Class</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## data matrix</span></span>
<span><span class="co"># remove meta data</span></span>
<span><span class="va">X</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span><span class="op">=</span><span class="cn">NULL</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">X</span><span class="op">)</span><span class="op">=</span><span class="va">SM</span><span class="op">$</span><span class="va">SampleID</span></span>
<span></span>
<span><span class="co"># feature meta data</span></span>
<span><span class="va">VM</span><span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>idx<span class="op">=</span><span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">X</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">VM</span><span class="op">)</span><span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">X</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># prepare DatasetExperiment</span></span>
<span><span class="va">DE</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/struct/man/struct_DatasetExperiment.html" class="external-link">DatasetExperiment</a></span><span class="op">(</span></span>
<span>    data<span class="op">=</span><span class="va">X</span>,</span>
<span>    sample_meta<span class="op">=</span><span class="va">SM</span>,</span>
<span>    variable_meta<span class="op">=</span><span class="va">VM</span>,</span>
<span>    description<span class="op">=</span><span class="st">'1H-NMR urinary metabolomic profiling for diagnosis of gastric cancer'</span>,</span>
<span>    name<span class="op">=</span><span class="st">'Gastric cancer (NMR)'</span><span class="op">)</span></span>
<span></span>
<span><span class="va">DE</span></span></code></pre></div>
<pre><code><span><span class="co">## A "DatasetExperiment" object</span></span>
<span><span class="co">## ----------------------------</span></span>
<span><span class="co">## name:          Gastric cancer (NMR)</span></span>
<span><span class="co">## description:   1H-NMR urinary metabolomic profiling for diagnosis of gastric cancer</span></span>
<span><span class="co">## data:          140 rows x 149 columns</span></span>
<span><span class="co">## sample_meta:   140 rows x 5 columns</span></span>
<span><span class="co">## variable_meta: 149 rows x 1 columns</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="data-pre-processing-and-quality-assessment">Data pre-processing and quality assessment<a class="anchor" aria-label="anchor" href="#data-pre-processing-and-quality-assessment"></a>
</h3>
<p>It is good practice to remove any features that may be of low
quality, and to assess the quality of the data in general. In the
Tutorial features with QC-RSD &gt; 20% and where more than 10% of the
features are missing are retained.</p>
<div class="sourceCode" id="cb73"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># prepare model sequence</span></span>
<span><span class="va">M</span> <span class="op">=</span> <span class="fu"><a href="../reference/rsd_filter.html">rsd_filter</a></span><span class="op">(</span>rsd_threshold<span class="op">=</span><span class="fl">20</span>,qc_label<span class="op">=</span><span class="st">'QC'</span>,factor_name<span class="op">=</span><span class="st">'Class'</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="../reference/mv_feature_filter.html">mv_feature_filter</a></span><span class="op">(</span>threshold <span class="op">=</span> <span class="fl">10</span>,method<span class="op">=</span><span class="st">'across'</span>,factor_name<span class="op">=</span><span class="st">'Class'</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># apply model</span></span>
<span><span class="va">M</span> <span class="op">=</span> <span class="fu"><a href="../reference/model_apply.html">model_apply</a></span><span class="op">(</span><span class="va">M</span>,<span class="va">DE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># get the model output</span></span>
<span><span class="va">filtered</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/struct/man/predicted.html" class="external-link">predicted</a></span><span class="op">(</span><span class="va">M</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># summary of filtered data</span></span>
<span><span class="va">filtered</span></span></code></pre></div>
<pre><code><span><span class="co">## A "DatasetExperiment" object</span></span>
<span><span class="co">## ----------------------------</span></span>
<span><span class="co">## name:          Gastric cancer (NMR)</span></span>
<span><span class="co">## description:   1H-NMR urinary metabolomic profiling for diagnosis of gastric cancer</span></span>
<span><span class="co">## data:          140 rows x 53 columns</span></span>
<span><span class="co">## sample_meta:   140 rows x 5 columns</span></span>
<span><span class="co">## variable_meta: 53 rows x 1 columns</span></span></code></pre>
<p>Note there is an additional feature vs the the processing reported by
Mendez et. al. because the filters here use &gt;= or &lt;= instead of
&gt; and &lt;.</p>
<p>After suitable scaling and transformation PCA can be used to assess
data quality. It is expected that the biological variance (samples) will
be larger than the technical variance (QCs). In the workflow that we are
reproducing (<a href="https://cimcb.github.io/MetabWorkflowTutorial/Tutorial1.html" class="external-link">link</a>)
the following steps were applied:</p>
<ul>
<li>log10 transform</li>
<li>autoscaling (scaled to unit variance)</li>
<li>knn imputation (3 neighbours)</li>
</ul>
<p>The transformed and scaled matrix in then used as input to PCA. Using
<code>struct</code> we can chain all of these steps into a single model
sequence.</p>
<div class="sourceCode" id="cb75"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># prepare the model sequence</span></span>
<span><span class="va">M</span> <span class="op">=</span> <span class="fu"><a href="../reference/log_transform.html">log_transform</a></span><span class="op">(</span>base <span class="op">=</span> <span class="fl">10</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="../reference/autoscale.html">autoscale</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="../reference/knn_impute.html">knn_impute</a></span><span class="op">(</span>neighbours <span class="op">=</span> <span class="fl">3</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="../reference/PCA.html">PCA</a></span><span class="op">(</span>number_components <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># apply model sequence to data</span></span>
<span><span class="va">M</span> <span class="op">=</span> <span class="fu"><a href="../reference/model_apply.html">model_apply</a></span><span class="op">(</span><span class="va">M</span>,<span class="va">filtered</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># get the transformed, scaled and imputed matrix</span></span>
<span><span class="va">TSI</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/struct/man/predicted.html" class="external-link">predicted</a></span><span class="op">(</span><span class="va">M</span><span class="op">[</span><span class="fl">3</span><span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># scores plot</span></span>
<span><span class="va">C</span> <span class="op">=</span> <span class="fu"><a href="../reference/pca_scores_plot.html">pca_scores_plot</a></span><span class="op">(</span>factor_name <span class="op">=</span> <span class="st">'SampleType'</span><span class="op">)</span></span>
<span><span class="va">g1</span> <span class="op">=</span> <span class="fu"><a href="../reference/chart_plot.html">chart_plot</a></span><span class="op">(</span><span class="va">C</span>,<span class="va">M</span><span class="op">[</span><span class="fl">4</span><span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># loadings plot</span></span>
<span><span class="va">C</span> <span class="op">=</span> <span class="fu"><a href="../reference/pca_loadings_plot.html">pca_loadings_plot</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">g2</span> <span class="op">=</span> <span class="fu"><a href="../reference/chart_plot.html">chart_plot</a></span><span class="op">(</span><span class="va">C</span>,<span class="va">M</span><span class="op">[</span><span class="fl">4</span><span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html" class="external-link">plot_grid</a></span><span class="op">(</span><span class="va">g1</span>,<span class="va">g2</span>,align<span class="op">=</span><span class="st">'hv'</span>,nrow<span class="op">=</span><span class="fl">1</span>,axis<span class="op">=</span><span class="st">'tblr'</span><span class="op">)</span></span></code></pre></div>
<p><img src="data_analysis_omics_using_the_structtoolbox_files/figure-html/unnamed-chunk-49-1.png"><!-- --></p>
</div>
<div class="section level3">
<h3 id="univariate-statistics">Univariate statistics<a class="anchor" aria-label="anchor" href="#univariate-statistics"></a>
</h3>
<p><code>structToolbox</code> provides a number of objects for ttest,
counting numbers of features etc. For brevity only the ttest is
calculated for comparison with the workflow we are following (<a href="https://cimcb.github.io/MetabWorkflowTutorial/Tutorial1.html" class="external-link">link</a>).
The QC samples need to be excluded, and the data reduced to only the GC
and HE groups.</p>
<div class="sourceCode" id="cb76"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># prepare model</span></span>
<span><span class="va">TT</span> <span class="op">=</span> <span class="fu"><a href="../reference/filter_smeta.html">filter_smeta</a></span><span class="op">(</span>mode<span class="op">=</span><span class="st">'include'</span>,factor_name<span class="op">=</span><span class="st">'Class'</span>,levels<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'GC'</span>,<span class="st">'HE'</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>  </span>
<span>     <span class="fu"><a href="../reference/ttest.html">ttest</a></span><span class="op">(</span>alpha<span class="op">=</span><span class="fl">0.05</span>,mtc<span class="op">=</span><span class="st">'fdr'</span>,factor_names<span class="op">=</span><span class="st">'Class'</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># apply model</span></span>
<span><span class="va">TT</span> <span class="op">=</span> <span class="fu"><a href="../reference/model_apply.html">model_apply</a></span><span class="op">(</span><span class="va">TT</span>,<span class="va">filtered</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># keep the data filtered by group for later</span></span>
<span><span class="va">filtered</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/struct/man/predicted.html" class="external-link">predicted</a></span><span class="op">(</span><span class="va">TT</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># convert to data frame</span></span>
<span><span class="va">out</span><span class="op">=</span><span class="fu"><a href="../reference/as_data_frame.html">as_data_frame</a></span><span class="op">(</span><span class="va">TT</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># show first few features</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">out</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##     t_statistic   t_p_value t_significant estimate.mean.GC estimate.mean.HE</span></span>
<span><span class="co">## M4   -3.5392652 0.008421042          TRUE         26.47778         51.73947</span></span>
<span><span class="co">## M5    1.4296604 0.410396437         FALSE        265.11860        169.91500</span></span>
<span><span class="co">## M7    2.7456506 0.051494976         FALSE        118.52558         53.98718</span></span>
<span><span class="co">## M8   -2.1294198 0.178392032         FALSE         54.39535         79.26750</span></span>
<span><span class="co">## M11   0.5106536 0.776939682         FALSE        201.34390        171.27949</span></span>
<span><span class="co">## M14  -1.4786810 0.403091881         FALSE         61.53171         83.90250</span></span>
<span><span class="co">##         lower      upper</span></span>
<span><span class="co">## M4  -39.56162 -10.961769</span></span>
<span><span class="co">## M5  -38.04747 228.454679</span></span>
<span><span class="co">## M7   17.60818 111.468619</span></span>
<span><span class="co">## M8  -48.20069  -1.543611</span></span>
<span><span class="co">## M11 -87.30604 147.434869</span></span>
<span><span class="co">## M14 -52.57754   7.835950</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="multivariate-statistics-and-machine-learning">Multivariate statistics and machine learning<a class="anchor" aria-label="anchor" href="#multivariate-statistics-and-machine-learning"></a>
</h3>
<div class="section level4">
<h4 id="training-and-test-sets">Training and Test sets<a class="anchor" aria-label="anchor" href="#training-and-test-sets"></a>
</h4>
<p>Splitting data into training and test sets is an important aspect of
machine learning. In <code>structToolbox</code> this is implemented
using the <code>split_data</code> object for random subsampling across
the whole dataset, and <code>stratified_split</code> for splitting based
on group sizes, which is the approach used by Mendez et al.</p>
<div class="sourceCode" id="cb78"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># prepare model</span></span>
<span><span class="va">M</span> <span class="op">=</span> <span class="fu"><a href="../reference/stratified_split.html">stratified_split</a></span><span class="op">(</span>p_train<span class="op">=</span><span class="fl">0.75</span>,factor_name<span class="op">=</span><span class="st">'Class'</span><span class="op">)</span></span>
<span><span class="co"># apply to filtered data</span></span>
<span><span class="va">M</span> <span class="op">=</span> <span class="fu"><a href="../reference/model_apply.html">model_apply</a></span><span class="op">(</span><span class="va">M</span>,<span class="va">filtered</span><span class="op">)</span></span>
<span><span class="co"># get data from object</span></span>
<span><span class="va">train</span> <span class="op">=</span> <span class="va">M</span><span class="op">$</span><span class="va">training</span></span>
<span><span class="va">train</span></span></code></pre></div>
<pre><code><span><span class="co">## A "DatasetExperiment" object</span></span>
<span><span class="co">## ----------------------------</span></span>
<span><span class="co">## name:          Gastric cancer (NMR)</span></span>
<span><span class="co">##                (Training set)</span></span>
<span><span class="co">## description:   • 1H-NMR urinary metabolomic profiling for diagnosis of gastric cancer</span></span>
<span><span class="co">##                • A subset of the data has been selected as a training set</span></span>
<span><span class="co">## data:          62 rows x 53 columns</span></span>
<span><span class="co">## sample_meta:   62 rows x 5 columns</span></span>
<span><span class="co">## variable_meta: 53 rows x 1 columns</span></span></code></pre>
<div class="sourceCode" id="cb80"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/cat.html" class="external-link">cat</a></span><span class="op">(</span><span class="st">'\n'</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb81"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">test</span> <span class="op">=</span> <span class="va">M</span><span class="op">$</span><span class="va">testing</span></span>
<span><span class="va">test</span></span></code></pre></div>
<pre><code><span><span class="co">## A "DatasetExperiment" object</span></span>
<span><span class="co">## ----------------------------</span></span>
<span><span class="co">## name:          Gastric cancer (NMR)</span></span>
<span><span class="co">##                (Testing set)</span></span>
<span><span class="co">## description:   • 1H-NMR urinary metabolomic profiling for diagnosis of gastric cancer</span></span>
<span><span class="co">##                • A subset of the data has been selected as a test set</span></span>
<span><span class="co">## data:          21 rows x 53 columns</span></span>
<span><span class="co">## sample_meta:   21 rows x 5 columns</span></span>
<span><span class="co">## variable_meta: 53 rows x 1 columns</span></span></code></pre>
</div>
<div class="section level4">
<h4 id="optimal-number-of-pls-components">Optimal number of PLS components<a class="anchor" aria-label="anchor" href="#optimal-number-of-pls-components"></a>
</h4>
<p>In Mendez et al a k-fold cross-validation is used to determine the
optimal number of PLS components. 100 bootstrap iterations are used to
generate confidence intervals. In <code>strucToolbox</code> these are
implemented using “iterator” objects, that can be combined with model
objects. R2 is used as the metric for optimisation, so the
<code>PLSR</code> model in structToolbox will be used. For speed only 10
bootstrap iterations are used here.</p>
<div class="sourceCode" id="cb83"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># scale/transform training data</span></span>
<span><span class="va">M</span> <span class="op">=</span> <span class="fu"><a href="../reference/log_transform.html">log_transform</a></span><span class="op">(</span>base <span class="op">=</span> <span class="fl">10</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="../reference/autoscale.html">autoscale</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="../reference/knn_impute.html">knn_impute</a></span><span class="op">(</span>neighbours <span class="op">=</span> <span class="fl">3</span>,by<span class="op">=</span><span class="st">'samples'</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># apply model</span></span>
<span><span class="va">M</span> <span class="op">=</span> <span class="fu"><a href="../reference/model_apply.html">model_apply</a></span><span class="op">(</span><span class="va">M</span>,<span class="va">train</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># get scaled/transformed training data</span></span>
<span><span class="va">train_st</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/struct/man/predicted.html" class="external-link">predicted</a></span><span class="op">(</span><span class="va">M</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># prepare model sequence</span></span>
<span><span class="va">MS</span> <span class="op">=</span> <span class="fu"><a href="../reference/grid_search_1d.html">grid_search_1d</a></span><span class="op">(</span></span>
<span>        param_to_optimise <span class="op">=</span> <span class="st">'number_components'</span>,</span>
<span>        search_values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">)</span><span class="op">)</span>,</span>
<span>        model_index <span class="op">=</span> <span class="fl">2</span>,</span>
<span>        factor_name <span class="op">=</span> <span class="st">'Class_num'</span>,</span>
<span>        max_min <span class="op">=</span> <span class="st">'max'</span><span class="op">)</span> <span class="op">*</span></span>
<span>     <span class="fu"><a href="../reference/permute_sample_order.html">permute_sample_order</a></span><span class="op">(</span></span>
<span>        number_of_permutations <span class="op">=</span> <span class="fl">10</span><span class="op">)</span> <span class="op">*</span></span>
<span>     <span class="fu"><a href="../reference/kfold_xval.html">kfold_xval</a></span><span class="op">(</span></span>
<span>        folds <span class="op">=</span> <span class="fl">5</span>,</span>
<span>        factor_name <span class="op">=</span> <span class="st">'Class_num'</span><span class="op">)</span> <span class="op">*</span> </span>
<span>     <span class="op">(</span><span class="fu"><a href="../reference/mean_centre.html">mean_centre</a></span><span class="op">(</span>mode<span class="op">=</span><span class="st">'sample_meta'</span><span class="op">)</span><span class="op">+</span></span>
<span>      <span class="fu"><a href="../reference/PLSR.html">PLSR</a></span><span class="op">(</span>factor_name<span class="op">=</span><span class="st">'Class_num'</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># run the validation</span></span>
<span><span class="va">MS</span> <span class="op">=</span> <span class="fu">struct</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/struct/man/iterator.html" class="external-link">run</a></span><span class="op">(</span><span class="va">MS</span>,<span class="va">train_st</span>,<span class="fu"><a href="../reference/r_squared.html">r_squared</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#</span></span>
<span><span class="va">C</span> <span class="op">=</span> <span class="fu"><a href="../reference/gs_line.html">gs_line</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/chart_plot.html">chart_plot</a></span><span class="op">(</span><span class="va">C</span>,<span class="va">MS</span><span class="op">)</span></span></code></pre></div>
<p><img src="data_analysis_omics_using_the_structtoolbox_files/figure-html/unnamed-chunk-52-1.png"><!-- --></p>
<p>The chart plotted shows Q2, which is comparable with Figure 13 of <a href="(https://cimcb.github.io/MetabWorkflowTutorial/Tutorial1.html)" class="external-link">Mendez
et al</a> . Two components were selected by Mendez et al, so we will use
that here.</p>
</div>
<div class="section level4">
<h4 id="pls-model-evalutation">PLS model evalutation<a class="anchor" aria-label="anchor" href="#pls-model-evalutation"></a>
</h4>
<p>To evaluate the model for discriminant analysis in structToolbox the
<code>PLSDA</code> model is appropriate.</p>
<div class="sourceCode" id="cb84"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># prepare the discriminant model</span></span>
<span><span class="va">P</span> <span class="op">=</span> <span class="fu"><a href="../reference/PLSDA.html">PLSDA</a></span><span class="op">(</span>number_components <span class="op">=</span> <span class="fl">2</span>, factor_name<span class="op">=</span><span class="st">'Class'</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># apply the model</span></span>
<span><span class="va">P</span> <span class="op">=</span> <span class="fu"><a href="../reference/model_apply.html">model_apply</a></span><span class="op">(</span><span class="va">P</span>,<span class="va">train_st</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># charts</span></span>
<span><span class="va">C</span> <span class="op">=</span> <span class="fu"><a href="../reference/plsda_predicted_plot.html">plsda_predicted_plot</a></span><span class="op">(</span>factor_name<span class="op">=</span><span class="st">'Class'</span>,style<span class="op">=</span><span class="st">'boxplot'</span><span class="op">)</span></span>
<span><span class="va">g1</span> <span class="op">=</span> <span class="fu"><a href="../reference/chart_plot.html">chart_plot</a></span><span class="op">(</span><span class="va">C</span>,<span class="va">P</span><span class="op">)</span></span>
<span></span>
<span><span class="va">C</span> <span class="op">=</span> <span class="fu"><a href="../reference/plsda_predicted_plot.html">plsda_predicted_plot</a></span><span class="op">(</span>factor_name<span class="op">=</span><span class="st">'Class'</span>,style<span class="op">=</span><span class="st">'density'</span><span class="op">)</span></span>
<span><span class="va">g2</span> <span class="op">=</span> <span class="fu"><a href="../reference/chart_plot.html">chart_plot</a></span><span class="op">(</span><span class="va">C</span>,<span class="va">P</span><span class="op">)</span><span class="op">+</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/lims.html" class="external-link">xlim</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">2</span>,<span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">C</span> <span class="op">=</span> <span class="fu"><a href="../reference/plsda_roc_plot.html">plsda_roc_plot</a></span><span class="op">(</span>factor_name<span class="op">=</span><span class="st">'Class'</span><span class="op">)</span></span>
<span><span class="va">g3</span> <span class="op">=</span> <span class="fu"><a href="../reference/chart_plot.html">chart_plot</a></span><span class="op">(</span><span class="va">C</span>,<span class="va">P</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html" class="external-link">plot_grid</a></span><span class="op">(</span><span class="va">g1</span>,<span class="va">g2</span>,<span class="va">g3</span>,align<span class="op">=</span><span class="st">'vh'</span>,axis<span class="op">=</span><span class="st">'tblr'</span>,nrow<span class="op">=</span><span class="fl">1</span>, labels<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'A'</span>,<span class="st">'B'</span>,<span class="st">'C'</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="data_analysis_omics_using_the_structtoolbox_files/figure-html/unnamed-chunk-53-1.png"><!-- --></p>
<div class="sourceCode" id="cb85"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># AUC for comparison with Mendez et al</span></span>
<span><span class="va">MET</span> <span class="op">=</span> <span class="fu"><a href="../reference/calculate.html">calculate</a></span><span class="op">(</span><span class="fu"><a href="../reference/AUC.html">AUC</a></span><span class="op">(</span><span class="op">)</span>,<span class="va">P</span><span class="op">$</span><span class="va">y</span><span class="op">$</span><span class="va">Class</span>,<span class="va">P</span><span class="op">$</span><span class="va">yhat</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="va">MET</span></span></code></pre></div>
<pre><code><span><span class="co">## A "AUC" object</span></span>
<span><span class="co">## --------------</span></span>
<span><span class="co">## name:          Area under ROC curve</span></span>
<span><span class="co">## description:   The area under the ROC curve of a classifier is estimated using the trapezoid method.</span></span>
<span><span class="co">## value:         0.9739583</span></span></code></pre>
<p>Note that the default cutoff in A and B of the figure above for the
PLS models in <code>structToolbox</code> is 0, because groups are
encoded as +/-1. This has no impact on the overall performance of the
model.</p>
</div>
<div class="section level4">
<h4 id="permutation-test">Permutation test<a class="anchor" aria-label="anchor" href="#permutation-test"></a>
</h4>
<p>A permutation test can be used to assess how likely the observed
result is to have occurred by chance. In structToolbox
<code>permutation_test</code> is an iterator object that can be combined
with other iterators and models.</p>
<div class="sourceCode" id="cb87"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># model sequence</span></span>
<span><span class="va">MS</span> <span class="op">=</span> <span class="fu"><a href="../reference/permutation_test.html">permutation_test</a></span><span class="op">(</span>number_of_permutations <span class="op">=</span> <span class="fl">20</span>,factor_name <span class="op">=</span> <span class="st">'Class_num'</span><span class="op">)</span> <span class="op">*</span> </span>
<span>     <span class="fu"><a href="../reference/kfold_xval.html">kfold_xval</a></span><span class="op">(</span>folds <span class="op">=</span> <span class="fl">5</span>,factor_name <span class="op">=</span> <span class="st">'Class_num'</span><span class="op">)</span> <span class="op">*</span></span>
<span>     <span class="op">(</span><span class="fu"><a href="../reference/mean_centre.html">mean_centre</a></span><span class="op">(</span>mode<span class="op">=</span><span class="st">'sample_meta'</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="../reference/PLSR.html">PLSR</a></span><span class="op">(</span>factor_name<span class="op">=</span><span class="st">'Class_num'</span>, number_components <span class="op">=</span> <span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># run iterator</span></span>
<span><span class="va">MS</span> <span class="op">=</span> <span class="fu">struct</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/struct/man/iterator.html" class="external-link">run</a></span><span class="op">(</span><span class="va">MS</span>,<span class="va">train_st</span>,<span class="fu"><a href="../reference/r_squared.html">r_squared</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># chart</span></span>
<span><span class="va">C</span> <span class="op">=</span> <span class="fu"><a href="../reference/permutation_test_plot.html">permutation_test_plot</a></span><span class="op">(</span>style <span class="op">=</span> <span class="st">'density'</span><span class="op">)</span> </span>
<span><span class="fu"><a href="../reference/chart_plot.html">chart_plot</a></span><span class="op">(</span><span class="va">C</span>,<span class="va">MS</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/lims.html" class="external-link">xlim</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">1</span>,<span class="fl">1</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">xlab</a></span><span class="op">(</span><span class="st">'R Squared'</span><span class="op">)</span></span></code></pre></div>
<p><img src="data_analysis_omics_using_the_structtoolbox_files/figure-html/unnamed-chunk-55-1.png"><!-- --></p>
<p>This plot is comparable to the bottom half of Figure 17 in <a href="https://cimcb.github.io/MetabWorkflowTutorial/Tutorial1.html" class="external-link">Mendez
et. al.</a>. The unpermuted (true) Q2 values are consistently better
than the permuted (null) models. i.e. the model is reliable.</p>
</div>
<div class="section level4">
<h4 id="pls-projection-plots">PLS projection plots<a class="anchor" aria-label="anchor" href="#pls-projection-plots"></a>
</h4>
<p>PLS can also be used to visualise the model and interpret the latent
variables.</p>
<div class="sourceCode" id="cb88"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># prepare the discriminant model</span></span>
<span><span class="va">P</span> <span class="op">=</span> <span class="fu"><a href="../reference/PLSDA.html">PLSDA</a></span><span class="op">(</span>number_components <span class="op">=</span> <span class="fl">2</span>, factor_name<span class="op">=</span><span class="st">'Class'</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># apply the model</span></span>
<span><span class="va">P</span> <span class="op">=</span> <span class="fu"><a href="../reference/model_apply.html">model_apply</a></span><span class="op">(</span><span class="va">P</span>,<span class="va">train_st</span><span class="op">)</span></span>
<span></span>
<span><span class="va">C</span> <span class="op">=</span> <span class="fu"><a href="../reference/pls_scores_plot.html">pls_scores_plot</a></span><span class="op">(</span>components<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span><span class="op">)</span>,factor_name <span class="op">=</span> <span class="st">'Class'</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/chart_plot.html">chart_plot</a></span><span class="op">(</span><span class="va">C</span>,<span class="va">P</span><span class="op">)</span></span></code></pre></div>
<p><img src="data_analysis_omics_using_the_structtoolbox_files/figure-html/unnamed-chunk-56-1.png"><!-- --></p>
</div>
<div class="section level4">
<h4 id="pls-feature-importance">PLS feature importance<a class="anchor" aria-label="anchor" href="#pls-feature-importance"></a>
</h4>
<p>Regression coefficients and VIP scores can be used to estimate the
importance of individual features to the PLS model. In Mendez et al
bootstrapping is used to estimate the confidence intervals, but for
brevity here we will skip this.</p>
<div class="sourceCode" id="cb89"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># prepare chart</span></span>
<span><span class="va">C</span> <span class="op">=</span> <span class="fu"><a href="../reference/pls_vip_plot.html">pls_vip_plot</a></span><span class="op">(</span>ycol <span class="op">=</span> <span class="st">'HE'</span><span class="op">)</span></span>
<span><span class="va">g1</span> <span class="op">=</span> <span class="fu"><a href="../reference/chart_plot.html">chart_plot</a></span><span class="op">(</span><span class="va">C</span>,<span class="va">P</span><span class="op">)</span></span>
<span></span>
<span><span class="va">C</span> <span class="op">=</span> <span class="fu"><a href="../reference/pls_regcoeff_plot.html">pls_regcoeff_plot</a></span><span class="op">(</span>ycol<span class="op">=</span><span class="st">'HE'</span><span class="op">)</span></span>
<span><span class="va">g2</span> <span class="op">=</span> <span class="fu"><a href="../reference/chart_plot.html">chart_plot</a></span><span class="op">(</span><span class="va">C</span>,<span class="va">P</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html" class="external-link">plot_grid</a></span><span class="op">(</span><span class="va">g1</span>,<span class="va">g2</span>,align<span class="op">=</span><span class="st">'hv'</span>,axis<span class="op">=</span><span class="st">'tblr'</span>,nrow<span class="op">=</span><span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<p><img src="data_analysis_omics_using_the_structtoolbox_files/figure-html/unnamed-chunk-57-1.png"><!-- --></p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="classification-of-metabolomics-data-using-support-vector-machines-">Classification of Metabolomics Data using Support Vector
Machines.<a class="anchor" aria-label="anchor" href="#classification-of-metabolomics-data-using-support-vector-machines-"></a>
</h2>
<div class="section level3">
<h3 id="introduction-5">Introduction<a class="anchor" aria-label="anchor" href="#introduction-5"></a>
</h3>
<p>The aim of this vignette is to illustrate how to apply SVM analysis
for Classifying Metabolomics data.</p>
<p>Support vector Machines (SVM) are a commonly used method in Machine
Learning. For classification tasks they are used to generate a boundary
between groups of samples in the training set. As well as generating
linear boundaries, SVM can be extended to exploit the use of kernels and
generate complex non-linear boundaries between groups if required.</p>
<p>For the <code>structToolbox</code> package, SVM functionality
provided by the <em><a href="https://CRAN.R-project.org/package=e1071" class="external-link">e1071</a></em> package
has been incorporated into a <code>model</code> object. A chart object
(<code>svm_plot_2d</code>) is also available to plot SVM boundaries for
data with two variables.</p>
</div>
<div class="section level3">
<h3 id="dataset-4">Dataset<a class="anchor" aria-label="anchor" href="#dataset-4"></a>
</h3>
<p>The 1H-NMR dataset used and described in <a href="https://link.springer.com/article/10.1007/s11306-019-1588-0,%20https://github.com/CIMCB/MetabWorkflowTutorial" class="external-link">Mendez
et al., (2020)</a> and in this vignette contains processed spectra of
urine samples obtained from gastric cancer and healthy patients <a href="https://www.nature.com/articles/bjc2015414" class="external-link">Chan et al.,
(2016)</a>. The raw experimental data is available through Metabolomics
Workbench (<a href="http://dx.doi.org/10.21228/M8B10B" class="external-link">PR000699</a>) and
the processed version is available from <a href="https://github.com/CIMCB/MetabWorkflowTutorial/raw/master/GastricCancer_NMR.xlsx" class="external-link">here</a>
as an Excel data file.</p>
<p>For simplicity we will use a pre-processed version of the 1H-NMR
“Gastric cancer” dataset using the <code>structToolbox</code> package.
Details in regards to pre-processing are reported in the
“NMR_clinical_metabolomics” vignette of the `r Biocpkg(“structToolbox”)
package.</p>
<div class="sourceCode" id="cb90"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># summary of DatasetExperiment object</span></span>
<span><span class="va">DE</span></span></code></pre></div>
<pre><code><span><span class="co">## A "DatasetExperiment" object</span></span>
<span><span class="co">## ----------------------------</span></span>
<span><span class="co">## name:          </span></span>
<span><span class="co">## description:   </span></span>
<span><span class="co">## data:          140 rows x 53 columns</span></span>
<span><span class="co">## sample_meta:   140 rows x 5 columns</span></span>
<span><span class="co">## variable_meta: 53 rows x 1 columns</span></span></code></pre>
<p>For the purposes of illustrating the effect of SVM parameters on the
boundary between groups, we reduce the data to include only the GC and
HE groups and apply PLS to reduce the data to two components. We then
treat the PLS scores as as a two group dataset with only two
features.</p>
<div class="sourceCode" id="cb92"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># model sequence and pls model (NB data already centred)</span></span>
<span><span class="va">MS</span> <span class="op">=</span> <span class="fu"><a href="../reference/filter_smeta.html">filter_smeta</a></span><span class="op">(</span>mode <span class="op">=</span> <span class="st">'include'</span>, levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'GC'</span>,<span class="st">'HE'</span><span class="op">)</span>, factor_name <span class="op">=</span> <span class="st">'Class'</span><span class="op">)</span> <span class="op">+</span></span>
<span>     <span class="fu"><a href="../reference/PLSDA.html">PLSDA</a></span><span class="op">(</span>factor_name <span class="op">=</span> <span class="st">'Class'</span>,number_components <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># apply PLS model</span></span>
<span><span class="va">MS</span> <span class="op">=</span> <span class="fu"><a href="../reference/model_apply.html">model_apply</a></span><span class="op">(</span><span class="va">MS</span>,<span class="va">DE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># plot the data</span></span>
<span><span class="va">C</span> <span class="op">=</span> <span class="fu"><a href="../reference/pls_scores_plot.html">pls_scores_plot</a></span><span class="op">(</span>factor_name <span class="op">=</span> <span class="st">'Class'</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/chart_plot.html">chart_plot</a></span><span class="op">(</span><span class="va">C</span>,<span class="va">MS</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<p><img src="data_analysis_omics_using_the_structtoolbox_files/figure-html/unnamed-chunk-60-1.png"><!-- --></p>
<div class="sourceCode" id="cb93"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># new DatasetExperiment object from the PLS scores</span></span>
<span><span class="va">DE2</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/struct/man/struct_DatasetExperiment.html" class="external-link">DatasetExperiment</a></span><span class="op">(</span></span>
<span>  data <span class="op">=</span> <span class="va">MS</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">$</span><span class="va">scores</span><span class="op">$</span><span class="va">data</span>, </span>
<span>  sample_meta <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/struct/man/predicted.html" class="external-link">predicted</a></span><span class="op">(</span><span class="va">MS</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span><span class="op">$</span><span class="va">sample_meta</span>,</span>
<span>  variable_meta <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="st">'LV'</span><span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span><span class="op">)</span>,row.names <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">MS</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">$</span><span class="va">scores</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  name <span class="op">=</span> <span class="st">'Illustrativate SVM dataset'</span>,</span>
<span>  description <span class="op">=</span> <span class="st">'Generated by applying PLS to the processed Gastric cancer (NMR) dataset'</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">DE2</span></span></code></pre></div>
<pre><code><span><span class="co">## A "DatasetExperiment" object</span></span>
<span><span class="co">## ----------------------------</span></span>
<span><span class="co">## name:          Illustrativate SVM dataset</span></span>
<span><span class="co">## description:   Generated by applying PLS to the processed Gastric cancer (NMR) dataset</span></span>
<span><span class="co">## data:          83 rows x 2 columns</span></span>
<span><span class="co">## sample_meta:   83 rows x 5 columns</span></span>
<span><span class="co">## variable_meta: 2 rows x 1 columns</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="basic-svm-model">Basic SVM model<a class="anchor" aria-label="anchor" href="#basic-svm-model"></a>
</h3>
<p>The simplest SVM model uses a linear kernel. In
<code>structToolbox</code> the <code>SVM</code> model can be used to
train and apply SVM models. A <code>svm_plot_2d</code> chart object is
provided for visualisation of boundaries in two dimensions.</p>
<div class="sourceCode" id="cb95"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># SVM model</span></span>
<span><span class="va">M</span> <span class="op">=</span> <span class="fu"><a href="../reference/SVM.html">SVM</a></span><span class="op">(</span></span>
<span>  factor_name <span class="op">=</span> <span class="st">'Class'</span>,</span>
<span>  kernel <span class="op">=</span> <span class="st">'linear'</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># apply model</span></span>
<span><span class="va">M</span> <span class="op">=</span> <span class="fu"><a href="../reference/model_apply.html">model_apply</a></span><span class="op">(</span><span class="va">M</span>,<span class="va">DE2</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># plot boundary</span></span>
<span><span class="va">C</span> <span class="op">=</span> <span class="fu"><a href="../reference/svm_plot_2d.html">svm_plot_2d</a></span><span class="op">(</span>factor_name <span class="op">=</span> <span class="st">'Class'</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/chart_plot.html">chart_plot</a></span><span class="op">(</span><span class="va">C</span>,<span class="va">M</span>, <span class="va">DE2</span><span class="op">)</span></span></code></pre></div>
<p><img src="data_analysis_omics_using_the_structtoolbox_files/figure-html/unnamed-chunk-61-1.png"><!-- --></p>
<p>The SVM boundary is plotted in black, the margins in grey and support
vectors are indicated by grey circles.</p>
</div>
<div class="section level3">
<h3 id="svm-cost-function">SVM cost function<a class="anchor" aria-label="anchor" href="#svm-cost-function"></a>
</h3>
<p>The SVM cost function applies a penalty to samples on the wrong side
of the margins. A high penalty results in a narrow margin and tries to
force more samples to be on the correct side of the boundary. A low
penalty makes for a wider margin and is less strict about samples being
misclassified. The optimal cost to use is data dependent.</p>
<div class="sourceCode" id="cb96"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># low cost</span></span>
<span><span class="va">M</span><span class="op">$</span><span class="va">cost</span><span class="op">=</span><span class="fl">0.01</span></span>
<span><span class="va">M</span><span class="op">=</span><span class="fu"><a href="../reference/model_apply.html">model_apply</a></span><span class="op">(</span><span class="va">M</span>,<span class="va">DE2</span><span class="op">)</span></span>
<span><span class="va">C</span><span class="op">=</span><span class="fu"><a href="../reference/svm_plot_2d.html">svm_plot_2d</a></span><span class="op">(</span>factor_name<span class="op">=</span><span class="st">'Species'</span><span class="op">)</span></span>
<span><span class="va">g1</span><span class="op">=</span><span class="fu"><a href="../reference/chart_plot.html">chart_plot</a></span><span class="op">(</span><span class="va">C</span>,<span class="va">M</span>,<span class="va">DE2</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># medium cost</span></span>
<span><span class="va">M</span><span class="op">$</span><span class="va">cost</span><span class="op">=</span><span class="fl">0.05</span></span>
<span><span class="va">M</span><span class="op">=</span><span class="fu"><a href="../reference/model_apply.html">model_apply</a></span><span class="op">(</span><span class="va">M</span>,<span class="va">DE2</span><span class="op">)</span></span>
<span><span class="va">C</span><span class="op">=</span><span class="fu"><a href="../reference/svm_plot_2d.html">svm_plot_2d</a></span><span class="op">(</span>factor_name<span class="op">=</span><span class="st">'Species'</span><span class="op">)</span></span>
<span><span class="va">g2</span><span class="op">=</span><span class="fu"><a href="../reference/chart_plot.html">chart_plot</a></span><span class="op">(</span><span class="va">C</span>,<span class="va">M</span>,<span class="va">DE2</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># high cost</span></span>
<span><span class="va">M</span><span class="op">$</span><span class="va">cost</span><span class="op">=</span><span class="fl">100</span></span>
<span><span class="va">M</span><span class="op">=</span><span class="fu"><a href="../reference/model_apply.html">model_apply</a></span><span class="op">(</span><span class="va">M</span>,<span class="va">DE2</span><span class="op">)</span></span>
<span></span>
<span><span class="va">C</span><span class="op">=</span><span class="fu"><a href="../reference/svm_plot_2d.html">svm_plot_2d</a></span><span class="op">(</span>factor_name<span class="op">=</span><span class="st">'Species'</span><span class="op">)</span></span>
<span><span class="va">g3</span><span class="op">=</span><span class="fu"><a href="../reference/chart_plot.html">chart_plot</a></span><span class="op">(</span><span class="va">C</span>,<span class="va">M</span>,<span class="va">DE2</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># plot</span></span>
<span><span class="va">prow</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html" class="external-link">plot_grid</a></span><span class="op">(</span></span>
<span>  <span class="va">g1</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position<span class="op">=</span><span class="st">"none"</span><span class="op">)</span>,</span>
<span>  <span class="va">g2</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position<span class="op">=</span><span class="st">"none"</span><span class="op">)</span>,</span>
<span>  <span class="va">g3</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position<span class="op">=</span><span class="st">"none"</span><span class="op">)</span>,</span>
<span>  align <span class="op">=</span> <span class="st">'vh'</span>,</span>
<span>  labels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Low cost"</span>, <span class="st">"Medium cost"</span>, <span class="st">"High cost"</span><span class="op">)</span>,</span>
<span>  hjust <span class="op">=</span> <span class="op">-</span><span class="fl">1</span>,</span>
<span>  nrow <span class="op">=</span> <span class="fl">2</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">legend</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://wilkelab.org/cowplot/reference/get_legend.html" class="external-link">get_legend</a></span><span class="op">(</span></span>
<span>  <span class="co"># create some space to the left of the legend</span></span>
<span>  <span class="va">g1</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guides.html" class="external-link">guides</a></span><span class="op">(</span>color <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guide_legend.html" class="external-link">guide_legend</a></span><span class="op">(</span>nrow <span class="op">=</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"bottom"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html" class="external-link">plot_grid</a></span><span class="op">(</span><span class="va">prow</span>, <span class="va">legend</span>, ncol<span class="op">=</span><span class="fl">1</span>, rel_heights <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">.1</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="data_analysis_omics_using_the_structtoolbox_files/figure-html/unnamed-chunk-62-1.png"><!-- --></p>
</div>
<div class="section level3">
<h3 id="kernel-functions">Kernel functions<a class="anchor" aria-label="anchor" href="#kernel-functions"></a>
</h3>
<p>A number of different kernels can be used with support vector
machines. For the <code>structToolbox</code> wrapper ‘linear’,
‘polynomial’,‘radial’ and ‘sigmoid’ kernels can be specified. Using
kernels allows the boundary to be more flexible, but often require
additional parameters to be specified. The best kernel to use will vary
depending on the dataset, but a common choice is the radial kernel as it
allows high flexibility with a single parameter.</p>
<div class="sourceCode" id="cb97"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># set a fixed cost for this comparison</span></span>
<span><span class="va">M</span><span class="op">$</span><span class="va">cost</span><span class="op">=</span><span class="fl">1</span></span>
<span></span>
<span><span class="co"># linear kernel</span></span>
<span><span class="va">M</span><span class="op">$</span><span class="va">kernel</span><span class="op">=</span><span class="st">'linear'</span></span>
<span><span class="va">M</span><span class="op">=</span><span class="fu"><a href="../reference/model_apply.html">model_apply</a></span><span class="op">(</span><span class="va">M</span>,<span class="va">DE2</span><span class="op">)</span></span>
<span><span class="va">C</span><span class="op">=</span><span class="fu"><a href="../reference/svm_plot_2d.html">svm_plot_2d</a></span><span class="op">(</span>factor_name<span class="op">=</span><span class="st">'Species'</span><span class="op">)</span></span>
<span><span class="va">g1</span><span class="op">=</span><span class="fu"><a href="../reference/chart_plot.html">chart_plot</a></span><span class="op">(</span><span class="va">C</span>,<span class="va">M</span>,<span class="va">DE2</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># polynomial kernel</span></span>
<span><span class="va">M</span><span class="op">$</span><span class="va">kernel</span><span class="op">=</span><span class="st">'polynomial'</span></span>
<span><span class="va">M</span><span class="op">$</span><span class="va">gamma</span><span class="op">=</span><span class="fl">1</span></span>
<span><span class="va">M</span><span class="op">$</span><span class="va">coef0</span><span class="op">=</span><span class="fl">0</span></span>
<span><span class="va">M</span><span class="op">=</span><span class="fu"><a href="../reference/model_apply.html">model_apply</a></span><span class="op">(</span><span class="va">M</span>,<span class="va">DE2</span><span class="op">)</span></span>
<span><span class="va">C</span><span class="op">=</span><span class="fu"><a href="../reference/svm_plot_2d.html">svm_plot_2d</a></span><span class="op">(</span>factor_name<span class="op">=</span><span class="st">'Species'</span><span class="op">)</span></span>
<span><span class="va">g2</span><span class="op">=</span><span class="fu"><a href="../reference/chart_plot.html">chart_plot</a></span><span class="op">(</span><span class="va">C</span>,<span class="va">M</span>,<span class="va">DE2</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># rbf kernel</span></span>
<span><span class="va">M</span><span class="op">$</span><span class="va">kernel</span><span class="op">=</span><span class="st">'radial'</span></span>
<span><span class="va">M</span><span class="op">$</span><span class="va">gamma</span><span class="op">=</span><span class="fl">1</span></span>
<span><span class="va">M</span><span class="op">=</span><span class="fu"><a href="../reference/model_apply.html">model_apply</a></span><span class="op">(</span><span class="va">M</span>,<span class="va">DE2</span><span class="op">)</span></span>
<span><span class="va">C</span><span class="op">=</span><span class="fu"><a href="../reference/svm_plot_2d.html">svm_plot_2d</a></span><span class="op">(</span>factor_name<span class="op">=</span><span class="st">'Species'</span><span class="op">)</span></span>
<span><span class="va">g3</span><span class="op">=</span><span class="fu"><a href="../reference/chart_plot.html">chart_plot</a></span><span class="op">(</span><span class="va">C</span>,<span class="va">M</span>,<span class="va">DE2</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># sigmoid kernel</span></span>
<span><span class="va">M</span><span class="op">$</span><span class="va">kernel</span><span class="op">=</span><span class="st">'sigmoid'</span></span>
<span><span class="va">M</span><span class="op">$</span><span class="va">gamma</span><span class="op">=</span><span class="fl">1</span></span>
<span><span class="va">M</span><span class="op">$</span><span class="va">coef0</span><span class="op">=</span><span class="fl">0</span></span>
<span><span class="va">M</span><span class="op">=</span><span class="fu"><a href="../reference/model_apply.html">model_apply</a></span><span class="op">(</span><span class="va">M</span>,<span class="va">DE2</span><span class="op">)</span></span>
<span><span class="va">C</span><span class="op">=</span><span class="fu"><a href="../reference/svm_plot_2d.html">svm_plot_2d</a></span><span class="op">(</span>factor_name<span class="op">=</span><span class="st">'Species'</span><span class="op">)</span></span>
<span><span class="va">g4</span><span class="op">=</span><span class="fu"><a href="../reference/chart_plot.html">chart_plot</a></span><span class="op">(</span><span class="va">C</span>,<span class="va">M</span>,<span class="va">DE2</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># plot</span></span>
<span><span class="va">prow</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html" class="external-link">plot_grid</a></span><span class="op">(</span></span>
<span>  <span class="va">g1</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position<span class="op">=</span><span class="st">"none"</span><span class="op">)</span>,</span>
<span>  <span class="va">g2</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position<span class="op">=</span><span class="st">"none"</span><span class="op">)</span>,</span>
<span>  <span class="va">g3</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position<span class="op">=</span><span class="st">"none"</span><span class="op">)</span>,</span>
<span>  <span class="va">g4</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position<span class="op">=</span><span class="st">"none"</span><span class="op">)</span>,</span>
<span>  align <span class="op">=</span> <span class="st">'vh'</span>,</span>
<span>  labels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Linear"</span>, <span class="st">"Polynomial"</span>, <span class="st">"Radial"</span>,<span class="st">"Sigmoid"</span><span class="op">)</span>,</span>
<span>  hjust <span class="op">=</span> <span class="op">-</span><span class="fl">1</span>,</span>
<span>  nrow <span class="op">=</span> <span class="fl">2</span></span>
<span><span class="op">)</span></span>
<span><span class="va">legend</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://wilkelab.org/cowplot/reference/get_legend.html" class="external-link">get_legend</a></span><span class="op">(</span></span>
<span>  <span class="co"># create some space to the left of the legend</span></span>
<span>  <span class="va">g1</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guides.html" class="external-link">guides</a></span><span class="op">(</span>color <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guide_legend.html" class="external-link">guide_legend</a></span><span class="op">(</span>nrow <span class="op">=</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"bottom"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html" class="external-link">plot_grid</a></span><span class="op">(</span><span class="va">prow</span>, <span class="va">legend</span>, ncol <span class="op">=</span> <span class="fl">1</span>, rel_heights <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">.1</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="data_analysis_omics_using_the_structtoolbox_files/figure-html/unnamed-chunk-63-1.png"><!-- --></p>
<p>The parameters of a kernel can be used to control the complexity of
the boundary. Here I show how the radial kernel parameter “gamma” can be
used to change the complexity of the boundary. In combination with the
cost parameter (which I keep constant here) this allows for highly
flexible boundary models.</p>
<div class="sourceCode" id="cb98"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># rbf kernel and cost</span></span>
<span><span class="va">M</span><span class="op">$</span><span class="va">kernel</span> <span class="op">=</span> <span class="st">'radial'</span></span>
<span><span class="va">M</span><span class="op">$</span><span class="va">cost</span> <span class="op">=</span> <span class="fl">1</span></span>
<span></span>
<span><span class="co"># low gamma</span></span>
<span><span class="va">M</span><span class="op">$</span><span class="va">gamma</span><span class="op">=</span><span class="fl">0.01</span></span>
<span><span class="va">M</span><span class="op">=</span><span class="fu"><a href="../reference/model_apply.html">model_apply</a></span><span class="op">(</span><span class="va">M</span>,<span class="va">DE2</span><span class="op">)</span></span>
<span><span class="va">C</span><span class="op">=</span><span class="fu"><a href="../reference/svm_plot_2d.html">svm_plot_2d</a></span><span class="op">(</span>factor_name<span class="op">=</span><span class="st">'Species'</span><span class="op">)</span></span>
<span><span class="va">g1</span><span class="op">=</span><span class="fu"><a href="../reference/chart_plot.html">chart_plot</a></span><span class="op">(</span><span class="va">C</span>,<span class="va">M</span>,<span class="va">DE2</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># medium gamma</span></span>
<span><span class="va">M</span><span class="op">$</span><span class="va">gamma</span><span class="op">=</span><span class="fl">0.1</span></span>
<span><span class="va">M</span><span class="op">=</span><span class="fu"><a href="../reference/model_apply.html">model_apply</a></span><span class="op">(</span><span class="va">M</span>,<span class="va">DE2</span><span class="op">)</span></span>
<span><span class="va">C</span><span class="op">=</span><span class="fu"><a href="../reference/svm_plot_2d.html">svm_plot_2d</a></span><span class="op">(</span>factor_name<span class="op">=</span><span class="st">'Species'</span><span class="op">)</span></span>
<span><span class="va">g2</span><span class="op">=</span><span class="fu"><a href="../reference/chart_plot.html">chart_plot</a></span><span class="op">(</span><span class="va">C</span>,<span class="va">M</span>,<span class="va">DE2</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># high gamma</span></span>
<span><span class="va">M</span><span class="op">$</span><span class="va">gamma</span><span class="op">=</span><span class="fl">1</span></span>
<span><span class="va">M</span><span class="op">=</span><span class="fu"><a href="../reference/model_apply.html">model_apply</a></span><span class="op">(</span><span class="va">M</span>,<span class="va">DE2</span><span class="op">)</span></span>
<span><span class="va">C</span><span class="op">=</span><span class="fu"><a href="../reference/svm_plot_2d.html">svm_plot_2d</a></span><span class="op">(</span>factor_name<span class="op">=</span><span class="st">'Species'</span><span class="op">)</span></span>
<span><span class="va">g3</span><span class="op">=</span><span class="fu"><a href="../reference/chart_plot.html">chart_plot</a></span><span class="op">(</span><span class="va">C</span>,<span class="va">M</span>,<span class="va">DE2</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># plot</span></span>
<span><span class="va">prow</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html" class="external-link">plot_grid</a></span><span class="op">(</span></span>
<span> <span class="va">g1</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position<span class="op">=</span><span class="st">"none"</span><span class="op">)</span>,</span>
<span> <span class="va">g2</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position<span class="op">=</span><span class="st">"none"</span><span class="op">)</span>,</span>
<span> <span class="va">g3</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position<span class="op">=</span><span class="st">"none"</span><span class="op">)</span>,</span>
<span> align <span class="op">=</span> <span class="st">'vh'</span>,</span>
<span> labels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Low gamma"</span>, <span class="st">"Medium gamma"</span>, <span class="st">"High gamma"</span><span class="op">)</span>,</span>
<span> hjust <span class="op">=</span> <span class="op">-</span><span class="fl">1</span>,</span>
<span> nrow <span class="op">=</span> <span class="fl">2</span></span>
<span><span class="op">)</span></span>
<span><span class="va">legend</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://wilkelab.org/cowplot/reference/get_legend.html" class="external-link">get_legend</a></span><span class="op">(</span></span>
<span>  <span class="co"># create some space to the left of the legend</span></span>
<span>  <span class="va">g1</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guides.html" class="external-link">guides</a></span><span class="op">(</span>color <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guide_legend.html" class="external-link">guide_legend</a></span><span class="op">(</span>nrow <span class="op">=</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"bottom"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html" class="external-link">plot_grid</a></span><span class="op">(</span><span class="va">prow</span>, <span class="va">legend</span>, ncol <span class="op">=</span> <span class="fl">1</span>, rel_heights <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">.1</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="data_analysis_omics_using_the_structtoolbox_files/figure-html/unnamed-chunk-64-1.png"><!-- --></p>
<p>Note that best practice would be to select the optimal kernel
parameter(s) in combination with the cost parameter (e.g. by 2d grid
search) so that the best combination of both is identified.</p>
</div>
</div>
<div class="section level2">
<h2 id="exploratory-data-analysis-of-lc-ms-based-proteomics-and-metabolomics-datasets-stategra-project">Exploratory data analysis of LC-MS-based proteomics and metabolomics
datasets (STATegra project)<a class="anchor" aria-label="anchor" href="#exploratory-data-analysis-of-lc-ms-based-proteomics-and-metabolomics-datasets-stategra-project"></a>
</h2>
<div class="section level3">
<h3 id="introduction-6">Introduction<a class="anchor" aria-label="anchor" href="#introduction-6"></a>
</h3>
<p>The aim of this vignette is to conduct data preprocessing and
exploratory analysis of data from the STATegra project (<a href="https://www.nature.com/articles/s41597-019-0202-7" class="external-link uri">https://www.nature.com/articles/s41597-019-0202-7</a>). For
demonstration purposes we will focus on the Proteomics and Metabolomics
datasets that are publicly available as part of the STATegra multi-omics
dataset.</p>
<blockquote>
<p>…the STATegra multi-omics dataset combines measurements from up to 10
different omics technologies applied to the same biological system,
namely the well-studied mouse pre-B-cell differentiation. STATegra
includes high-throughput measurements of chromatin structure, gene
expression, proteomics and metabolomics, and it is complemented with
single-cell data. <a href="https://www.nature.com/articles/s41597-019-0202-7" class="external-link">Gomez-Cabrero
et al</a></p>
</blockquote>
<p>STATegra includes high-throughput measurements of chromatin
structure, gene expression, proteomics and metabolomics, and it is
complemented with single-cell data.</p>
</div>
<div class="section level3">
<h3 id="lc-ms-based-proteomics-dataset">LC-MS-based proteomics dataset<a class="anchor" aria-label="anchor" href="#lc-ms-based-proteomics-dataset"></a>
</h3>
<p>The LC-MS-based proteomics dataset from the STATegra multi-omics
dataset (see Introduction) can be found on <a href="https://github.com/STATegraData/STATegraData" class="external-link">github</a> and must
be extracted from the zip file prior to data analysis.</p>
<div class="sourceCode" id="cb99"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># path to zip</span></span>
<span><span class="va">zipfile</span> <span class="op">=</span> <span class="st">"https://raw.github.com/STATegraData/STATegraData/master/Script_STATegra_Proteomics.zip"</span></span>
<span></span>
<span><span class="co">## retrieve from BiocFileCache</span></span>
<span><span class="va">path</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocFileCache/man/BiocFileCache-class.html" class="external-link">bfcrpath</a></span><span class="op">(</span><span class="va">bfc</span>,<span class="va">zipfile</span><span class="op">)</span></span>
<span><span class="va">temp</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocFileCache/man/BiocFileCache-class.html" class="external-link">bfccache</a></span><span class="op">(</span><span class="va">bfc</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## ... or download to temp location</span></span>
<span><span class="co"># path = tempfile()</span></span>
<span><span class="co"># temp = tempdir()</span></span>
<span><span class="co"># download.file(zipfile,path)</span></span>
<span></span>
<span><span class="co"># unzip</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/unzip.html" class="external-link">unzip</a></span><span class="op">(</span><span class="va">path</span>, files <span class="op">=</span> <span class="st">"Proteomics_01_uniprot_canonical_normalized.txt"</span>, exdir<span class="op">=</span><span class="va">temp</span><span class="op">)</span></span>
<span><span class="co"># read samples</span></span>
<span><span class="va">all_data</span> <span class="op">&lt;-</span>  <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.delim</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">temp</span>,<span class="st">"Proteomics_01_uniprot_canonical_normalized.txt"</span><span class="op">)</span>, as.is <span class="op">=</span> <span class="cn">TRUE</span>, header <span class="op">=</span> <span class="cn">TRUE</span>, sep <span class="op">=</span> <span class="st">"\t"</span><span class="op">)</span></span></code></pre></div>
<p>The imported data needs to be converted to
<code>DatasetExperiment</code> format for use with
<code>structToolbox</code>.</p>
<div class="sourceCode" id="cb100"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># extract data matrix</span></span>
<span><span class="va">data</span> <span class="op">=</span> <span class="va">all_data</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">2527</span>,<span class="fl">51</span><span class="op">:</span><span class="fl">86</span><span class="op">]</span></span>
<span><span class="co"># shorten sample names</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span>, <span class="kw">function</span> <span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/substr.html" class="external-link">substr</a></span><span class="op">(</span><span class="va">x</span>, <span class="fl">27</span>, <span class="fu"><a href="https://rdrr.io/r/base/nchar.html" class="external-link">nchar</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># replace 0 with NA</span></span>
<span><span class="va">data</span><span class="op">[</span><span class="va">data</span> <span class="op">==</span> <span class="fl">0</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="cn">NA</span></span>
<span><span class="co"># transpose</span></span>
<span><span class="va">data</span><span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># prepare sample meta</span></span>
<span><span class="va">SM</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span>,<span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">s</span><span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/strsplit.html" class="external-link">strsplit</a></span><span class="op">(</span><span class="va">x</span>,<span class="st">'_'</span><span class="op">)</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span> <span class="co"># split at underscore</span></span>
<span>  <span class="va">out</span><span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>    <span class="st">'treatment'</span> <span class="op">=</span> <span class="va">s</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>,</span>
<span>    <span class="st">'time'</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/substr.html" class="external-link">substr</a></span><span class="op">(</span><span class="va">s</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span>,<span class="fl">1</span>,<span class="fu"><a href="https://rdrr.io/r/base/nchar.html" class="external-link">nchar</a></span><span class="op">(</span><span class="va">s</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">-</span><span class="fl">1</span><span class="op">)</span> ,    </span>
<span>    <span class="st">'batch'</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/substr.html" class="external-link">substr</a></span><span class="op">(</span><span class="va">s</span><span class="op">[[</span><span class="fl">3</span><span class="op">]</span><span class="op">]</span>,<span class="fl">6</span>,<span class="fu"><a href="https://rdrr.io/r/base/nchar.html" class="external-link">nchar</a></span><span class="op">(</span><span class="va">s</span><span class="op">[[</span><span class="fl">3</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    <span class="st">'condition'</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/substr.html" class="external-link">substr</a></span><span class="op">(</span><span class="va">x</span>,<span class="fl">1</span>,<span class="fl">6</span><span class="op">)</span> <span class="co"># interaction between treatment and time</span></span>
<span>  <span class="op">)</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="va">out</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span><span class="va">SM</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/do.call.html" class="external-link">do.call</a></span><span class="op">(</span><span class="va">rbind</span>,<span class="va">SM</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">SM</span><span class="op">)</span><span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span></span>
<span><span class="co"># convert to factors</span></span>
<span><span class="va">SM</span><span class="op">$</span><span class="va">treatment</span><span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">SM</span><span class="op">$</span><span class="va">treatment</span><span class="op">)</span></span>
<span><span class="va">SM</span><span class="op">$</span><span class="va">time</span><span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">ordered</a></span><span class="op">(</span><span class="va">SM</span><span class="op">$</span><span class="va">time</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"0"</span>,<span class="st">"2"</span>,<span class="st">"6"</span>,<span class="st">"12"</span>,<span class="st">"18"</span>,<span class="st">"24"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">SM</span><span class="op">$</span><span class="va">batch</span><span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">ordered</a></span><span class="op">(</span><span class="va">SM</span><span class="op">$</span><span class="va">batch</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">3</span>,<span class="fl">4</span>,<span class="fl">5</span>,<span class="fl">6</span>,<span class="fl">7</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">SM</span><span class="op">$</span><span class="va">condition</span><span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">SM</span><span class="op">$</span><span class="va">condition</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># variable meta data</span></span>
<span><span class="va">VM</span> <span class="op">=</span> <span class="va">all_data</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">2527</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">6</span>,<span class="fl">7</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">VM</span><span class="op">)</span><span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># prepare DatasetExperiment</span></span>
<span><span class="va">DS</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/struct/man/struct_DatasetExperiment.html" class="external-link">DatasetExperiment</a></span><span class="op">(</span></span>
<span>  data <span class="op">=</span> <span class="va">data</span>, </span>
<span>  sample_meta <span class="op">=</span> <span class="va">SM</span>, </span>
<span>  variable_meta <span class="op">=</span> <span class="va">VM</span>, </span>
<span>  name <span class="op">=</span> <span class="st">'STATegra Proteomics'</span>,</span>
<span>  description <span class="op">=</span> <span class="st">'downloaded from: https://github.com/STATegraData/STATegraData/'</span></span>
<span><span class="op">)</span></span>
<span><span class="va">DS</span></span></code></pre></div>
<pre><code><span><span class="co">## A "DatasetExperiment" object</span></span>
<span><span class="co">## ----------------------------</span></span>
<span><span class="co">## name:          STATegra Proteomics</span></span>
<span><span class="co">## description:   downloaded from: https://github.com/STATegraData/STATegraData/</span></span>
<span><span class="co">## data:          36 rows x 2527 columns</span></span>
<span><span class="co">## sample_meta:   36 rows x 4 columns</span></span>
<span><span class="co">## variable_meta: 2527 rows x 3 columns</span></span></code></pre>
<p>A number of Reporter genes were included in the study. We plot two of
them here to illustrate some trends in the data.</p>
<div class="sourceCode" id="cb102"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># find id of reporters</span></span>
<span><span class="va">Ldha</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">DS</span><span class="op">$</span><span class="va">variable_meta</span><span class="op">$</span><span class="va">Gene.names</span><span class="op">==</span><span class="st">'Ldha'</span><span class="op">)</span></span>
<span><span class="va">Hk2</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">DS</span><span class="op">$</span><span class="va">variable_meta</span><span class="op">$</span><span class="va">Gene.names</span><span class="op">==</span><span class="st">'Hk2'</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># chart object</span></span>
<span><span class="va">C</span> <span class="op">=</span> <span class="fu"><a href="../reference/feature_boxplot.html">feature_boxplot</a></span><span class="op">(</span>feature_to_plot<span class="op">=</span><span class="va">Ldha</span>,factor_name<span class="op">=</span><span class="st">'time'</span>,label_outliers<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">g1</span><span class="op">=</span><span class="fu"><a href="../reference/chart_plot.html">chart_plot</a></span><span class="op">(</span><span class="va">C</span>,<span class="va">DS</span><span class="op">)</span><span class="op">+</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">'Ldha'</span><span class="op">)</span><span class="op">+</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="st">'expression'</span><span class="op">)</span></span>
<span><span class="va">C</span> <span class="op">=</span> <span class="fu"><a href="../reference/feature_boxplot.html">feature_boxplot</a></span><span class="op">(</span>feature_to_plot<span class="op">=</span><span class="va">Hk2</span>,factor_name<span class="op">=</span><span class="st">'time'</span>,label_outliers<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">g2</span><span class="op">=</span><span class="fu"><a href="../reference/chart_plot.html">chart_plot</a></span><span class="op">(</span><span class="va">C</span>,<span class="va">DS</span><span class="op">)</span><span class="op">+</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">'Hk2'</span><span class="op">)</span><span class="op">+</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="st">'expression'</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html" class="external-link">plot_grid</a></span><span class="op">(</span><span class="va">g1</span>,<span class="va">g2</span>,nrow<span class="op">=</span><span class="fl">1</span>,align<span class="op">=</span><span class="st">'vh'</span>,axis<span class="op">=</span><span class="st">'tblr'</span><span class="op">)</span></span></code></pre></div>
<p><img src="data_analysis_omics_using_the_structtoolbox_files/figure-html/unnamed-chunk-67-1.png"><!-- --></p>
<div class="section level4">
<h4 id="data-transformation">Data transformation<a class="anchor" aria-label="anchor" href="#data-transformation"></a>
</h4>
<p>The data is log2 transformed, then scaled such that the mean of the
medians is equal for all conditions. These steps are available in
<code>structToolbox</code> using <code>log_transform</code> and
<code>mean_of_medians</code> objects.</p>
<div class="sourceCode" id="cb103"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># prepare model sequence</span></span>
<span><span class="va">M</span> <span class="op">=</span> <span class="fu"><a href="../reference/log_transform.html">log_transform</a></span><span class="op">(</span></span>
<span>  base<span class="op">=</span><span class="fl">2</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="../reference/mean_of_medians.html">mean_of_medians</a></span><span class="op">(</span></span>
<span>    factor_name <span class="op">=</span> <span class="st">'condition'</span><span class="op">)</span></span>
<span><span class="co"># apply model sequence</span></span>
<span><span class="va">M</span> <span class="op">=</span> <span class="fu"><a href="../reference/model_apply.html">model_apply</a></span><span class="op">(</span><span class="va">M</span>,<span class="va">DS</span><span class="op">)</span> </span>
<span></span>
<span><span class="co"># get transformed data</span></span>
<span><span class="va">DST</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/struct/man/predicted.html" class="external-link">predicted</a></span><span class="op">(</span><span class="va">M</span><span class="op">)</span></span></code></pre></div>
<p>The Reporter genes are plotted again for comparison.</p>
<div class="sourceCode" id="cb104"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># chart object</span></span>
<span><span class="va">C</span> <span class="op">=</span> <span class="fu"><a href="../reference/feature_boxplot.html">feature_boxplot</a></span><span class="op">(</span>feature_to_plot<span class="op">=</span><span class="va">Ldha</span>,factor_name<span class="op">=</span><span class="st">'time'</span>,label_outliers<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">g1</span><span class="op">=</span><span class="fu"><a href="../reference/chart_plot.html">chart_plot</a></span><span class="op">(</span><span class="va">C</span>,<span class="va">DST</span><span class="op">)</span><span class="op">+</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">'Ldha'</span><span class="op">)</span><span class="op">+</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="st">'log2(expression)'</span><span class="op">)</span></span>
<span><span class="va">C</span> <span class="op">=</span> <span class="fu"><a href="../reference/feature_boxplot.html">feature_boxplot</a></span><span class="op">(</span>feature_to_plot<span class="op">=</span><span class="va">Hk2</span>,factor_name<span class="op">=</span><span class="st">'time'</span>,label_outliers<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">g2</span><span class="op">=</span><span class="fu"><a href="../reference/chart_plot.html">chart_plot</a></span><span class="op">(</span><span class="va">C</span>,<span class="va">DST</span><span class="op">)</span><span class="op">+</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">'Hk2'</span><span class="op">)</span><span class="op">+</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="st">'log2(expression)'</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html" class="external-link">plot_grid</a></span><span class="op">(</span><span class="va">g1</span>,<span class="va">g2</span>,nrow<span class="op">=</span><span class="fl">1</span>,align<span class="op">=</span><span class="st">'vh'</span>,axis<span class="op">=</span><span class="st">'tblr'</span><span class="op">)</span></span></code></pre></div>
<p><img src="data_analysis_omics_using_the_structtoolbox_files/figure-html/unnamed-chunk-69-1.png"><!-- --></p>
</div>
<div class="section level4">
<h4 id="missing-value-filtering">Missing value filtering<a class="anchor" aria-label="anchor" href="#missing-value-filtering"></a>
</h4>
<p>Missing value filtering involves removing any feature (gene) where
there are at least 3 missing values per group in at least 11
samples.</p>
<p>This specific filter is not in <code>structToolbox</code> at this
time, but can be achieved by combining <code>filter_na_count</code> and
<code>filter_by_name</code> objects.</p>
<p>Specifically, the default output of <code>filter_na_count</code> is
changed to return a matrix of NA counts per class. This output is then
connected to the ‘names’ input of <code>filter_by_names</code> and
converted to TRUE/FALSE using the ‘seq_fcn’ input.</p>
<p>The ‘seq_fcn’ function processes the NA counts <em>before</em> they
are used as inputs for <code>filter_by_names</code>. When data is passed
along the model sequence it passes unchanged through the
<code>filter_na_count</code> object becuase the default output has been
changed, so the <code>filter_na_count</code> and
<code>filter_by_name</code> objects are working together as a single
filter.</p>
<div class="sourceCode" id="cb105"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># build model sequence</span></span>
<span><span class="va">M2</span> <span class="op">=</span> <span class="fu"><a href="../reference/filter_na_count.html">filter_na_count</a></span><span class="op">(</span></span>
<span>  threshold<span class="op">=</span><span class="fl">2</span>,</span>
<span>  factor_name<span class="op">=</span><span class="st">'condition'</span>, </span>
<span>  predicted<span class="op">=</span><span class="st">'na_count'</span><span class="op">)</span> <span class="op">+</span>    <span class="co"># override the default output</span></span>
<span>  <span class="fu"><a href="../reference/filter_by_name.html">filter_by_name</a></span><span class="op">(</span></span>
<span>    mode<span class="op">=</span><span class="st">'exclude'</span>,</span>
<span>    dimension<span class="op">=</span><span class="st">'variable'</span>,</span>
<span>    names<span class="op">=</span><span class="st">'place_holder'</span>,</span>
<span>    seq_in<span class="op">=</span><span class="st">'names'</span>,</span>
<span>    seq_fcn<span class="op">=</span><span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span>      <span class="co"># convert NA count pre group to true/false</span></span>
<span>      <span class="va">x</span><span class="op">=</span><span class="va">x</span><span class="op">&gt;</span><span class="fl">2</span>                  <span class="co"># more the two missing per group</span></span>
<span>      <span class="va">x</span><span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">&gt;</span><span class="fl">10</span>        <span class="co"># in more than 10 groups</span></span>
<span>      <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span></span>
<span>    <span class="op">}</span></span>
<span>  <span class="op">)</span></span>
<span><span class="co"># apply to transformed data</span></span>
<span><span class="va">M2</span> <span class="op">=</span> <span class="fu"><a href="../reference/model_apply.html">model_apply</a></span><span class="op">(</span><span class="va">M2</span>,<span class="va">DST</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># get the filtered data</span></span>
<span><span class="va">DSTF</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/struct/man/predicted.html" class="external-link">predicted</a></span><span class="op">(</span><span class="va">M2</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="missing-value-imputation">Missing value imputation<a class="anchor" aria-label="anchor" href="#missing-value-imputation"></a>
</h4>
<p>STATegra uses two imputation methods that are not available as
<code>struct</code> objects, so we create temporary
<code>STATegra_impute</code> objects to do this using some functions
from the <code>struct</code> package.</p>
<p>The first imputation method imputes missing values for any treatment
where values are missing for all samples using a “random value below
discovery”. We create a new struct object using
<code>set_struct_obj</code> in the global environment, and a
“method_apply” method that implements the imputation.</p>
<div class="sourceCode" id="cb106"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># create new imputation object</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/struct/man/set_struct_obj.html" class="external-link">set_struct_obj</a></span><span class="op">(</span></span>
<span>  class_name <span class="op">=</span> <span class="st">'STATegra_impute1'</span>,</span>
<span>  struct_obj <span class="op">=</span> <span class="st">'model'</span>,</span>
<span>  params<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>factor_sd<span class="op">=</span><span class="st">'character'</span>,factor_name<span class="op">=</span><span class="st">'character'</span><span class="op">)</span>,</span>
<span>  outputs<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>imputed<span class="op">=</span><span class="st">'DatasetExperiment'</span><span class="op">)</span>,</span>
<span>  prototype <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>    name <span class="op">=</span> <span class="st">'STATegra imputation 1'</span>,</span>
<span>    description <span class="op">=</span> <span class="st">'If missing values are present for all one group then they are replaced with min/2 + "random value below discovery".'</span>,</span>
<span>    predicted <span class="op">=</span> <span class="st">'imputed'</span></span>
<span>  <span class="op">)</span> </span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># create method_apply for imputation method 1</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/struct/man/set_obj_method.html" class="external-link">set_obj_method</a></span><span class="op">(</span></span>
<span>  class_name<span class="op">=</span><span class="st">'STATegra_impute1'</span>,</span>
<span>  method_name<span class="op">=</span><span class="st">'model_apply'</span>,</span>
<span>  definition<span class="op">=</span><span class="kw">function</span><span class="op">(</span><span class="va">M</span>,<span class="va">D</span><span class="op">)</span> <span class="op">{</span></span>
<span>    </span>
<span>    <span class="co"># for each feature count NA within each level</span></span>
<span>    <span class="va">na</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="va">D</span><span class="op">$</span><span class="va">data</span>,<span class="fl">2</span>,<span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">{</span></span>
<span>      <span class="fu"><a href="https://rdrr.io/r/base/tapply.html" class="external-link">tapply</a></span><span class="op">(</span><span class="va">x</span>,<span class="va">D</span><span class="op">$</span><span class="va">sample_meta</span><span class="op">[[</span><span class="va">M</span><span class="op">$</span><span class="va">factor_name</span><span class="op">]</span><span class="op">]</span>,<span class="kw">function</span><span class="op">(</span><span class="va">y</span><span class="op">)</span><span class="op">{</span></span>
<span>        <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">y</span><span class="op">)</span><span class="op">)</span></span>
<span>      <span class="op">}</span><span class="op">)</span></span>
<span>    <span class="op">}</span><span class="op">)</span></span>
<span>    <span class="co"># count number of samples in each group</span></span>
<span>    <span class="va">count</span><span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">D</span><span class="op">$</span><span class="va">sample_meta</span><span class="op">[[</span><span class="va">M</span><span class="op">$</span><span class="va">factor_name</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span>    <span class="co"># standard deviation of features within levels of factor_sd</span></span>
<span>    <span class="va">sd</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="va">D</span><span class="op">$</span><span class="va">data</span>,<span class="fl">2</span>,<span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span><span class="fu"><a href="https://rdrr.io/r/base/tapply.html" class="external-link">tapply</a></span><span class="op">(</span><span class="va">x</span>,<span class="va">D</span><span class="op">$</span><span class="va">sample_meta</span><span class="op">[[</span><span class="va">M</span><span class="op">$</span><span class="va">factor_sd</span><span class="op">]</span><span class="op">]</span>,<span class="va">sd</span>,na.rm<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span><span class="op">}</span><span class="op">)</span></span>
<span>    <span class="va">sd</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/median.html" class="external-link">median</a></span><span class="op">(</span><span class="va">sd</span>,na.rm<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span>    </span>
<span>    <span class="co"># impute or not</span></span>
<span>    <span class="va">check</span><span class="op">=</span><span class="va">na</span> <span class="op">==</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="va">count</span>,nrow<span class="op">=</span><span class="fl">2</span>,ncol<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">D</span><span class="op">)</span><span class="op">)</span> <span class="co"># all missing in one class</span></span>
<span>    </span>
<span>    <span class="co"># impute matrix</span></span>
<span>    <span class="va">mi</span> <span class="op">=</span> <span class="va">D</span><span class="op">$</span><span class="va">data</span></span>
<span>    <span class="kw">for</span> <span class="op">(</span><span class="va">j</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">mi</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>      <span class="co"># index of group for this sample</span></span>
<span>      <span class="va">g</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/levels.html" class="external-link">levels</a></span><span class="op">(</span><span class="va">D</span><span class="op">$</span><span class="va">sample_meta</span><span class="op">[[</span><span class="va">M</span><span class="op">$</span><span class="va">factor_name</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">==</span><span class="va">D</span><span class="op">$</span><span class="va">sample_meta</span><span class="op">[[</span><span class="va">M</span><span class="op">$</span><span class="va">factor_name</span><span class="op">]</span><span class="op">]</span><span class="op">[</span><span class="va">j</span><span class="op">]</span><span class="op">)</span></span>
<span>      <span class="va">iv</span><span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/stats/Normal.html" class="external-link">rnorm</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">D</span><span class="op">)</span>,<span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">min</a></span><span class="op">(</span><span class="va">D</span><span class="op">$</span><span class="va">data</span><span class="op">[</span><span class="va">j</span>,<span class="op">]</span>,na.rm<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span><span class="op">/</span><span class="fl">2</span>,<span class="va">sd</span><span class="op">)</span></span>
<span>      <span class="va">mi</span><span class="op">[</span><span class="va">j</span>,<span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">mi</span><span class="op">[</span><span class="va">j</span>,<span class="op">]</span><span class="op">)</span> <span class="op">&amp;</span> <span class="va">check</span><span class="op">[</span><span class="va">g</span>,<span class="op">]</span><span class="op">]</span> <span class="op">=</span> <span class="va">iv</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">mi</span><span class="op">[</span><span class="va">j</span>,<span class="op">]</span><span class="op">)</span> <span class="op">&amp;</span> <span class="va">check</span><span class="op">[</span><span class="va">g</span>,<span class="op">]</span><span class="op">]</span></span>
<span>    <span class="op">}</span></span>
<span>    <span class="va">D</span><span class="op">$</span><span class="va">data</span> <span class="op">=</span> <span class="va">mi</span></span>
<span>    <span class="va">M</span><span class="op">$</span><span class="va">imputed</span><span class="op">=</span><span class="va">D</span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="va">M</span><span class="op">)</span></span>
<span>  <span class="op">}</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>The second imputation method replacing missing values in any
condition with exactly 1 missing value with the mean of the values for
that condition. Again we create a new struct object and corresponding
method for the the new object to implement the filter.</p>
<div class="sourceCode" id="cb107"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># create new imputation object</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/struct/man/set_struct_obj.html" class="external-link">set_struct_obj</a></span><span class="op">(</span></span>
<span>  class_name <span class="op">=</span> <span class="st">'STATegra_impute2'</span>,</span>
<span>  struct_obj <span class="op">=</span> <span class="st">'model'</span>,</span>
<span>  params<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>factor_name<span class="op">=</span><span class="st">'character'</span><span class="op">)</span>,</span>
<span>  outputs<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>imputed<span class="op">=</span><span class="st">'DatasetExperiment'</span><span class="op">)</span>,</span>
<span>  prototype <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>    name <span class="op">=</span> <span class="st">'STATegra imputation 2'</span>,</span>
<span>    description <span class="op">=</span> <span class="st">'For those conditions with only 1 NA impute with the mean of the condition.'</span>,</span>
<span>    predicted <span class="op">=</span> <span class="st">'imputed'</span></span>
<span>  <span class="op">)</span> </span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># create method_apply for imputation method 2</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/struct/man/set_obj_method.html" class="external-link">set_obj_method</a></span><span class="op">(</span></span>
<span>  class_name<span class="op">=</span><span class="st">'STATegra_impute2'</span>,</span>
<span>  method_name<span class="op">=</span><span class="st">'model_apply'</span>,</span>
<span>  definition<span class="op">=</span><span class="kw">function</span><span class="op">(</span><span class="va">M</span>,<span class="va">D</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="co"># levels in condition</span></span>
<span>    <span class="va">L</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/levels.html" class="external-link">levels</a></span><span class="op">(</span><span class="va">D</span><span class="op">$</span><span class="va">sample_meta</span><span class="op">[[</span><span class="va">M</span><span class="op">$</span><span class="va">factor_name</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span>    </span>
<span>    <span class="co"># for each feature count NA within each level</span></span>
<span>    <span class="va">na</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="va">D</span><span class="op">$</span><span class="va">data</span>,<span class="fl">2</span>,<span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">{</span></span>
<span>      <span class="fu"><a href="https://rdrr.io/r/base/tapply.html" class="external-link">tapply</a></span><span class="op">(</span><span class="va">x</span>,<span class="va">D</span><span class="op">$</span><span class="va">sample_meta</span><span class="op">[[</span><span class="va">M</span><span class="op">$</span><span class="va">factor_name</span><span class="op">]</span><span class="op">]</span>,<span class="kw">function</span><span class="op">(</span><span class="va">y</span><span class="op">)</span><span class="op">{</span></span>
<span>        <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">y</span><span class="op">)</span><span class="op">)</span></span>
<span>      <span class="op">}</span><span class="op">)</span></span>
<span>    <span class="op">}</span><span class="op">)</span></span>
<span>    </span>
<span>    <span class="co"># standard deviation of features within levels of factor_sd</span></span>
<span>    <span class="va">sd</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="va">D</span><span class="op">$</span><span class="va">data</span>,<span class="fl">2</span>,<span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span><span class="fu"><a href="https://rdrr.io/r/base/tapply.html" class="external-link">tapply</a></span><span class="op">(</span><span class="va">x</span>,<span class="va">D</span><span class="op">$</span><span class="va">sample_meta</span><span class="op">[[</span><span class="va">M</span><span class="op">$</span><span class="va">factor_name</span><span class="op">]</span><span class="op">]</span>,<span class="va">sd</span>,na.rm<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span><span class="op">}</span><span class="op">)</span></span>
<span>    <span class="va">sd</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/median.html" class="external-link">median</a></span><span class="op">(</span><span class="va">sd</span>,na.rm<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span>    </span>
<span>    <span class="co"># impute or not</span></span>
<span>    <span class="va">check</span><span class="op">=</span><span class="va">na</span> <span class="op">==</span> <span class="fl">1</span> <span class="co"># only one missing for a condition</span></span>
<span>    </span>
<span>    <span class="co"># index of samples for each condition</span></span>
<span>    <span class="va">IDX</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span></span>
<span>    <span class="kw">for</span> <span class="op">(</span><span class="va">k</span> <span class="kw">in</span> <span class="va">L</span><span class="op">)</span> <span class="op">{</span></span>
<span>      <span class="va">IDX</span><span class="op">[[</span><span class="va">k</span><span class="op">]</span><span class="op">]</span><span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">D</span><span class="op">$</span><span class="va">sample_meta</span><span class="op">[[</span><span class="va">M</span><span class="op">$</span><span class="va">factor_name</span><span class="op">]</span><span class="op">]</span><span class="op">==</span><span class="va">k</span><span class="op">)</span></span>
<span>    <span class="op">}</span></span>
<span>    </span>
<span>    <span class="co">## impute</span></span>
<span>    <span class="co"># for each feature</span></span>
<span>    <span class="kw">for</span> <span class="op">(</span><span class="va">k</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">D</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>      <span class="co"># for each condition</span></span>
<span>      <span class="kw">for</span> <span class="op">(</span><span class="va">j</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">L</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>        <span class="co"># if passes test</span></span>
<span>        <span class="kw">if</span> <span class="op">(</span><span class="va">check</span><span class="op">[</span><span class="va">j</span>,<span class="va">k</span><span class="op">]</span><span class="op">)</span> <span class="op">{</span></span>
<span>          <span class="co"># mean of samples in group</span></span>
<span>          <span class="va">m</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">D</span><span class="op">$</span><span class="va">data</span><span class="op">[</span><span class="va">IDX</span><span class="op">[[</span><span class="va">j</span><span class="op">]</span><span class="op">]</span>,<span class="va">k</span><span class="op">]</span>,na.rm<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span>          <span class="co"># imputed value</span></span>
<span>          <span class="va">im</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/Normal.html" class="external-link">rnorm</a></span><span class="op">(</span><span class="fl">1</span>,<span class="va">m</span>,<span class="va">sd</span><span class="op">)</span></span>
<span>          <span class="co"># replace NA with imputed</span></span>
<span>          <span class="va">D</span><span class="op">$</span><span class="va">data</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">D</span><span class="op">$</span><span class="va">data</span><span class="op">[</span>,<span class="va">k</span><span class="op">]</span><span class="op">)</span> <span class="op">&amp;</span> <span class="va">D</span><span class="op">$</span><span class="va">sample_meta</span><span class="op">[[</span><span class="va">M</span><span class="op">$</span><span class="va">factor_name</span><span class="op">]</span><span class="op">]</span><span class="op">==</span><span class="va">L</span><span class="op">[</span><span class="va">j</span><span class="op">]</span>,<span class="va">k</span><span class="op">]</span><span class="op">=</span><span class="va">im</span></span>
<span>        <span class="op">}</span></span>
<span>      <span class="op">}</span></span>
<span>    <span class="op">}</span></span>
<span>    <span class="va">M</span><span class="op">$</span><span class="va">imputed</span><span class="op">=</span><span class="va">D</span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="va">M</span><span class="op">)</span></span>
<span>  <span class="op">}</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>The new STATegra imputation objects can now be used in model
sequences like any other <code>struct</code> object. A final filter is
added to remove any feature that has missing values after
imputation.</p>
<div class="sourceCode" id="cb108"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># model sequence</span></span>
<span><span class="va">M3</span> <span class="op">=</span> <span class="fu">STATegra_impute1</span><span class="op">(</span>factor_name<span class="op">=</span><span class="st">'treatment'</span>,factor_sd<span class="op">=</span><span class="st">'condition'</span><span class="op">)</span> <span class="op">+</span></span>
<span>     <span class="fu">STATegra_impute2</span><span class="op">(</span>factor_name <span class="op">=</span> <span class="st">'condition'</span><span class="op">)</span> <span class="op">+</span> </span>
<span>     <span class="fu"><a href="../reference/filter_na_count.html">filter_na_count</a></span><span class="op">(</span>threshold <span class="op">=</span> <span class="fl">3</span>, factor_name<span class="op">=</span><span class="st">'condition'</span><span class="op">)</span></span>
<span><span class="co"># apply model</span></span>
<span><span class="va">M3</span> <span class="op">=</span> <span class="fu"><a href="../reference/model_apply.html">model_apply</a></span><span class="op">(</span><span class="va">M3</span>,<span class="va">DSTF</span><span class="op">)</span></span>
<span><span class="co"># get imputed data</span></span>
<span><span class="va">DSTFI</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/struct/man/predicted.html" class="external-link">predicted</a></span><span class="op">(</span><span class="va">M3</span><span class="op">)</span></span>
<span><span class="va">DSTFI</span></span></code></pre></div>
<pre><code><span><span class="co">## A "DatasetExperiment" object</span></span>
<span><span class="co">## ----------------------------</span></span>
<span><span class="co">## name:          STATegra Proteomics</span></span>
<span><span class="co">## description:   downloaded from: https://github.com/STATegraData/STATegraData/</span></span>
<span><span class="co">## data:          36 rows x 864 columns</span></span>
<span><span class="co">## sample_meta:   36 rows x 4 columns</span></span>
<span><span class="co">## variable_meta: 864 rows x 3 columns</span></span></code></pre>
</div>
<div class="section level4">
<h4 id="exploratory-analysis-1">Exploratory analysis<a class="anchor" aria-label="anchor" href="#exploratory-analysis-1"></a>
</h4>
<p>It is often useful to visualise the distribution of values across
samples to verify that the transformations/normalisation/filtering etc
have been effective.</p>
<p><img src="data_analysis_omics_using_the_structtoolbox_files/figure-html/unnamed-chunk-74-1.png"><!-- --></p>
<p>The values are no longer skewed and show an approximately normal
distribution. The boxplots are comparable in width with very few
outliers indicated, so the transformations etc have had an overall
positive effect.</p>
<p>PCA is used to provide a graphical representation of the data. For
comparison with the outputs from STATegra a filter is included to reduce
the data to include only the treated samples (IKA)</p>
<div class="sourceCode" id="cb110"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># model sequence</span></span>
<span><span class="va">P</span> <span class="op">=</span> <span class="fu"><a href="../reference/filter_smeta.html">filter_smeta</a></span><span class="op">(</span>mode<span class="op">=</span><span class="st">'include'</span>,factor_name<span class="op">=</span><span class="st">'treatment'</span>,levels<span class="op">=</span><span class="st">'IKA'</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="../reference/mean_centre.html">mean_centre</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="../reference/PCA.html">PCA</a></span><span class="op">(</span>number_components <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span><span class="co"># apply model</span></span>
<span><span class="va">P</span> <span class="op">=</span> <span class="fu"><a href="../reference/model_apply.html">model_apply</a></span><span class="op">(</span><span class="va">P</span>,<span class="va">DSTFI</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># scores plots coloured by factors</span></span>
<span><span class="va">g</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">k</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'batch'</span>,<span class="st">'time'</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">C</span> <span class="op">=</span> <span class="fu"><a href="../reference/pca_scores_plot.html">pca_scores_plot</a></span><span class="op">(</span>factor_name<span class="op">=</span><span class="va">k</span>,ellipse<span class="op">=</span><span class="st">'none'</span><span class="op">)</span></span>
<span>  <span class="va">g</span><span class="op">[[</span><span class="va">k</span><span class="op">]</span><span class="op">]</span><span class="op">=</span><span class="fu"><a href="../reference/chart_plot.html">chart_plot</a></span><span class="op">(</span><span class="va">C</span>,<span class="va">P</span><span class="op">[</span><span class="fl">3</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html" class="external-link">plot_grid</a></span><span class="op">(</span>plotlist <span class="op">=</span> <span class="va">g</span>,nrow<span class="op">=</span><span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<p><img src="data_analysis_omics_using_the_structtoolbox_files/figure-html/unnamed-chunk-75-1.png"><!-- --></p>
<p>There does not appear to be a strong batch effect. PC1 is dominated
by time point “24” and some potentially outlying points from time points
“2” and “0”.</p>
</div>
</div>
<div class="section level3">
<h3 id="lc-ms-based-metabolomics-dataset">LC-MS-based metabolomics dataset<a class="anchor" aria-label="anchor" href="#lc-ms-based-metabolomics-dataset"></a>
</h3>
<p>The LC-MS-based metabolomics dataset from the STATegra multi-omics
dataset (see Introduction) can be found on <a href="https://github.com/STATegraData/STATegraData" class="external-link">github</a> and must
be extracted from zip file prior to data analysis.</p>
<div class="sourceCode" id="cb111"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># path to zip</span></span>
<span><span class="va">zipfile</span> <span class="op">=</span> <span class="st">"https://raw.github.com/STATegraData/STATegraData/master/Script_STATegra_Metabolomics.zip"</span></span>
<span></span>
<span><span class="co">## retrieve from BiocFileCache</span></span>
<span><span class="va">path</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocFileCache/man/BiocFileCache-class.html" class="external-link">bfcrpath</a></span><span class="op">(</span><span class="va">bfc</span>,<span class="va">zipfile</span><span class="op">)</span></span>
<span><span class="va">temp</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocFileCache/man/BiocFileCache-class.html" class="external-link">bfccache</a></span><span class="op">(</span><span class="va">bfc</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## ... or download to temp location</span></span>
<span><span class="co"># path = tempfile()</span></span>
<span><span class="co"># temp = tempdir()</span></span>
<span><span class="co"># download.file(zipfile,path)</span></span>
<span></span>
<span><span class="co"># unzip</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/unzip.html" class="external-link">unzip</a></span><span class="op">(</span>zipfile<span class="op">=</span><span class="va">path</span>, files <span class="op">=</span> <span class="st">"LC_MS_raw_data.xlsx"</span>, exdir<span class="op">=</span><span class="va">temp</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># read samples</span></span>
<span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/openxlsx/man/read.xlsx.html" class="external-link">read.xlsx</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">temp</span>,<span class="st">"LC_MS_raw_data.xlsx"</span><span class="op">)</span>,sheet <span class="op">=</span> <span class="st">'Data'</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>The imported data needs to be converted to
<code>DatasetExperiment</code> format for use with
<code>structToolbox</code>.</p>
<div class="sourceCode" id="cb112"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># extract sample meta data</span></span>
<span><span class="va">SM</span> <span class="op">=</span> <span class="va">data</span><span class="op">[</span> ,<span class="fl">1</span><span class="op">:</span><span class="fl">8</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># add coloumn for sample type (QC, blank etc)</span></span>
<span><span class="va">blanks</span><span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span>,<span class="fl">33</span>,<span class="fl">34</span>,<span class="fl">65</span>,<span class="fl">66</span><span class="op">)</span></span>
<span><span class="va">QCs</span><span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">3</span>,<span class="fl">4</span>,<span class="fl">11</span>,<span class="fl">18</span>,<span class="fl">25</span>,<span class="fl">32</span>,<span class="fl">35</span>,<span class="fl">36</span>,<span class="fl">43</span>,<span class="fl">50</span>,<span class="fl">57</span>,<span class="fl">64</span><span class="op">)</span></span>
<span><span class="va">SM</span><span class="op">$</span><span class="va">sample_type</span><span class="op">=</span><span class="st">'Sample'</span></span>
<span><span class="va">SM</span><span class="op">$</span><span class="va">sample_type</span><span class="op">[</span><span class="va">blanks</span><span class="op">]</span><span class="op">=</span><span class="st">'Blank'</span></span>
<span><span class="va">SM</span><span class="op">$</span><span class="va">sample_type</span><span class="op">[</span><span class="va">QCs</span><span class="op">]</span><span class="op">=</span><span class="st">'QC'</span></span>
<span></span>
<span><span class="co"># put qc/blank labels in all factors for plotting later</span></span>
<span><span class="va">SM</span><span class="op">$</span><span class="va">biol.batch</span><span class="op">[</span><span class="va">SM</span><span class="op">$</span><span class="va">sample_type</span><span class="op">!=</span><span class="st">'Sample'</span><span class="op">]</span><span class="op">=</span><span class="va">SM</span><span class="op">$</span><span class="va">sample_type</span><span class="op">[</span><span class="va">SM</span><span class="op">$</span><span class="va">sample_type</span><span class="op">!=</span><span class="st">'Sample'</span><span class="op">]</span></span>
<span><span class="va">SM</span><span class="op">$</span><span class="va">time.point</span><span class="op">[</span><span class="va">SM</span><span class="op">$</span><span class="va">sample_type</span><span class="op">!=</span><span class="st">'Sample'</span><span class="op">]</span><span class="op">=</span><span class="va">SM</span><span class="op">$</span><span class="va">sample_type</span><span class="op">[</span><span class="va">SM</span><span class="op">$</span><span class="va">sample_type</span><span class="op">!=</span><span class="st">'Sample'</span><span class="op">]</span></span>
<span><span class="va">SM</span><span class="op">$</span><span class="va">condition</span><span class="op">[</span><span class="va">SM</span><span class="op">$</span><span class="va">sample_type</span><span class="op">!=</span><span class="st">'Sample'</span><span class="op">]</span><span class="op">=</span><span class="va">SM</span><span class="op">$</span><span class="va">sample_type</span><span class="op">[</span><span class="va">SM</span><span class="op">$</span><span class="va">sample_type</span><span class="op">!=</span><span class="st">'Sample'</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># convert to factors</span></span>
<span><span class="va">SM</span><span class="op">$</span><span class="va">biol.batch</span><span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">ordered</a></span><span class="op">(</span><span class="va">SM</span><span class="op">$</span><span class="va">biol.batch</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'9'</span>,<span class="st">'10'</span>,<span class="st">'11'</span>,<span class="st">'12'</span>,<span class="st">'QC'</span>,<span class="st">'Blank'</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">SM</span><span class="op">$</span><span class="va">time.point</span><span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">ordered</a></span><span class="op">(</span><span class="va">SM</span><span class="op">$</span><span class="va">time.point</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'0h'</span>,<span class="st">'2h'</span>,<span class="st">'6h'</span>,<span class="st">'12h'</span>,<span class="st">'18h'</span>,<span class="st">'24h'</span>,<span class="st">'QC'</span>,<span class="st">'Blank'</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">SM</span><span class="op">$</span><span class="va">condition</span><span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">SM</span><span class="op">$</span><span class="va">condition</span><span class="op">)</span></span>
<span><span class="va">SM</span><span class="op">$</span><span class="va">sample_type</span><span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">SM</span><span class="op">$</span><span class="va">sample_type</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># variable meta data</span></span>
<span><span class="va">VM</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="st">'annotation'</span><span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span><span class="op">[</span><span class="fl">9</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># raw data</span></span>
<span><span class="va">X</span> <span class="op">=</span> <span class="va">data</span><span class="op">[</span>,<span class="fl">9</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="co"># convert 0 to NA</span></span>
<span><span class="va">X</span><span class="op">[</span><span class="va">X</span><span class="op">==</span><span class="fl">0</span><span class="op">]</span><span class="op">=</span><span class="cn">NA</span></span>
<span><span class="co"># force to numeric; any non-numerics will become NA</span></span>
<span><span class="va">X</span><span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">X</span>,<span class="va">as.numeric</span><span class="op">)</span>,check.names <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># make sure row/col names match</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">X</span><span class="op">)</span><span class="op">=</span><span class="va">data</span><span class="op">$</span><span class="va">label</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">SM</span><span class="op">)</span><span class="op">=</span><span class="va">data</span><span class="op">$</span><span class="va">label</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">VM</span><span class="op">)</span><span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">X</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co"># create DatasetExperiment object</span></span>
<span><span class="va">DE</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/struct/man/struct_DatasetExperiment.html" class="external-link">DatasetExperiment</a></span><span class="op">(</span></span>
<span>  data <span class="op">=</span> <span class="va">X</span>, </span>
<span>  sample_meta <span class="op">=</span> <span class="va">SM</span>, </span>
<span>  variable_meta <span class="op">=</span> <span class="va">VM</span>, </span>
<span>  name <span class="op">=</span> <span class="st">'STATegra Metabolomics LCMS'</span>,</span>
<span>  description <span class="op">=</span> <span class="st">'https://www.nature.com/articles/s41597-019-0202-7'</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">DE</span></span></code></pre></div>
<pre><code><span><span class="co">## A "DatasetExperiment" object</span></span>
<span><span class="co">## ----------------------------</span></span>
<span><span class="co">## name:          STATegra Metabolomics LCMS</span></span>
<span><span class="co">## description:   https://www.nature.com/articles/s41597-019-0202-7</span></span>
<span><span class="co">## data:          66 rows x 152 columns</span></span>
<span><span class="co">## sample_meta:   66 rows x 9 columns</span></span>
<span><span class="co">## variable_meta: 152 rows x 1 columns</span></span></code></pre>
<div class="section level4">
<h4 id="data-preprocessing-1">Data preprocessing<a class="anchor" aria-label="anchor" href="#data-preprocessing-1"></a>
</h4>
<p>In the STATegra project the LCMS data was combined with GCMS data and
multiblock analysis was conducted. Here only the LCMS will be explored,
so the data will be processed differently in comparison to <a href="https://www.nature.com/articles/s41597-019-0202-7" class="external-link">Gomez-Cabrero
et al</a>. Some basic processing steps will be applied in order to
generate a valid PCA plot from the biological and QC samples.</p>
<div class="sourceCode" id="cb114"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># prepare model sequence</span></span>
<span><span class="va">MS</span> <span class="op">=</span> <span class="fu"><a href="../reference/filter_smeta.html">filter_smeta</a></span><span class="op">(</span>mode <span class="op">=</span> <span class="st">'include'</span>, levels<span class="op">=</span><span class="st">'QC'</span>, factor_name <span class="op">=</span> <span class="st">'sample_type'</span><span class="op">)</span> <span class="op">+</span></span>
<span>     <span class="fu"><a href="../reference/knn_impute.html">knn_impute</a></span><span class="op">(</span>neighbours<span class="op">=</span><span class="fl">5</span><span class="op">)</span> <span class="op">+</span></span>
<span>     <span class="fu"><a href="../reference/vec_norm.html">vec_norm</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>     <span class="fu"><a href="../reference/log_transform.html">log_transform</a></span><span class="op">(</span>base <span class="op">=</span> <span class="fl">10</span><span class="op">)</span> </span>
<span></span>
<span><span class="co"># apply model sequence</span></span>
<span><span class="va">MS</span> <span class="op">=</span> <span class="fu"><a href="../reference/model_apply.html">model_apply</a></span><span class="op">(</span><span class="va">MS</span>, <span class="va">DE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Warning in knnimp(x, k, maxmiss = rowmax, maxp = maxp): 3 rows with more than 50 % entries missing;</span></span>
<span><span class="co">##  mean imputation used for these rows</span></span></code></pre>
</div>
<div class="section level4">
<h4 id="exploratory-analysis-2">Exploratory analysis<a class="anchor" aria-label="anchor" href="#exploratory-analysis-2"></a>
</h4>
<p>First we will use PCA to look at the QC samples in order to make an
assessment of the data quality.</p>
<div class="sourceCode" id="cb116"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># pca model sequence</span></span>
<span><span class="va">M</span> <span class="op">=</span>  <span class="fu"><a href="../reference/mean_centre.html">mean_centre</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>     <span class="fu"><a href="../reference/PCA.html">PCA</a></span><span class="op">(</span>number_components <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># apply model</span></span>
<span><span class="va">M</span> <span class="op">=</span> <span class="fu"><a href="../reference/model_apply.html">model_apply</a></span><span class="op">(</span><span class="va">M</span>,<span class="fu"><a href="https://rdrr.io/pkg/struct/man/predicted.html" class="external-link">predicted</a></span><span class="op">(</span><span class="va">MS</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co"># PCA scores plot</span></span>
<span><span class="va">C</span> <span class="op">=</span> <span class="fu"><a href="../reference/pca_scores_plot.html">pca_scores_plot</a></span><span class="op">(</span>factor_name <span class="op">=</span> <span class="st">'sample_type'</span>,label_factor <span class="op">=</span> <span class="st">'order'</span>,points_to_label <span class="op">=</span> <span class="st">'all'</span><span class="op">)</span></span>
<span><span class="co"># plot</span></span>
<span><span class="fu"><a href="../reference/chart_plot.html">chart_plot</a></span><span class="op">(</span><span class="va">C</span>,<span class="va">M</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<p><img src="data_analysis_omics_using_the_structtoolbox_files/figure-html/unnamed-chunk-79-1.png"><!-- --></p>
<p>The QC labelled “36” is clearly very different to the other QCs. In
STATegra this QC was removed, so we will exclude it here as well. This
corresponds to QC H1. STATegra also excluded QC samples measured
immediately after a blank, which we will also do here.</p>
<div class="sourceCode" id="cb117"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># prepare model sequence</span></span>
<span><span class="va">MS</span> <span class="op">=</span> <span class="fu"><a href="../reference/filter_smeta.html">filter_smeta</a></span><span class="op">(</span></span>
<span>      mode <span class="op">=</span> <span class="st">'include'</span>, </span>
<span>      levels<span class="op">=</span><span class="st">'QC'</span>, </span>
<span>      factor_name <span class="op">=</span> <span class="st">'sample_type'</span><span class="op">)</span> <span class="op">+</span></span>
<span>  </span>
<span>     <span class="fu"><a href="../reference/filter_by_name.html">filter_by_name</a></span><span class="op">(</span></span>
<span>      mode <span class="op">=</span> <span class="st">'exclude'</span>, </span>
<span>      dimension<span class="op">=</span><span class="st">'sample'</span>,</span>
<span>      names <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'1358BZU_0001QC_H1'</span>,<span class="st">'1358BZU_0001QC_A1'</span>,<span class="st">'1358BZU_0001QC_G1'</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  </span>
<span>     <span class="fu"><a href="../reference/knn_impute.html">knn_impute</a></span><span class="op">(</span></span>
<span>      neighbours<span class="op">=</span><span class="fl">5</span><span class="op">)</span> <span class="op">+</span></span>
<span>  </span>
<span>     <span class="fu"><a href="../reference/vec_norm.html">vec_norm</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  </span>
<span>     <span class="fu"><a href="../reference/log_transform.html">log_transform</a></span><span class="op">(</span></span>
<span>       base <span class="op">=</span> <span class="fl">10</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  </span>
<span>     <span class="fu"><a href="../reference/mean_centre.html">mean_centre</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  </span>
<span>     <span class="fu"><a href="../reference/PCA.html">PCA</a></span><span class="op">(</span></span>
<span>       number_components <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># apply model sequence</span></span>
<span><span class="va">MS</span> <span class="op">=</span> <span class="fu"><a href="../reference/model_apply.html">model_apply</a></span><span class="op">(</span><span class="va">MS</span>, <span class="va">DE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Warning in knnimp(x, k, maxmiss = rowmax, maxp = maxp): 4 rows with more than 50 % entries missing;</span></span>
<span><span class="co">##  mean imputation used for these rows</span></span></code></pre>
<div class="sourceCode" id="cb119"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># PCA scores plot</span></span>
<span><span class="va">C</span> <span class="op">=</span> <span class="fu"><a href="../reference/pca_scores_plot.html">pca_scores_plot</a></span><span class="op">(</span>factor_name <span class="op">=</span> <span class="st">'sample_type'</span>,label_factor <span class="op">=</span> <span class="st">'order'</span>,points_to_label <span class="op">=</span> <span class="st">'all'</span><span class="op">)</span></span>
<span><span class="co"># plot</span></span>
<span><span class="fu"><a href="../reference/chart_plot.html">chart_plot</a></span><span class="op">(</span><span class="va">C</span>,<span class="va">MS</span><span class="op">[</span><span class="fl">7</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<p><img src="data_analysis_omics_using_the_structtoolbox_files/figure-html/unnamed-chunk-80-1.png"><!-- --></p>
<p>Now we will plot the QC samples in context with the samples. There
are several possible approaches, and we will apply the approach of
applying PCA to the full dataset including the QCs. We will exclude the
blanks as it is likely that they will dominate the plot if not removed.
All samples from batch 12 were excluded from STATegra and we will
replicate that here.</p>
<div class="sourceCode" id="cb120"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># prepare model sequence</span></span>
<span><span class="va">MS</span> <span class="op">=</span> <span class="fu"><a href="../reference/filter_smeta.html">filter_smeta</a></span><span class="op">(</span></span>
<span>      mode <span class="op">=</span> <span class="st">'exclude'</span>, </span>
<span>      levels<span class="op">=</span><span class="st">'Blank'</span>, </span>
<span>      factor_name <span class="op">=</span> <span class="st">'sample_type'</span><span class="op">)</span> <span class="op">+</span></span>
<span>  </span>
<span>     <span class="fu"><a href="../reference/filter_smeta.html">filter_smeta</a></span><span class="op">(</span></span>
<span>      mode <span class="op">=</span> <span class="st">'exclude'</span>, </span>
<span>      levels<span class="op">=</span><span class="st">'12'</span>, </span>
<span>      factor_name <span class="op">=</span> <span class="st">'biol.batch'</span><span class="op">)</span> <span class="op">+</span></span>
<span>  </span>
<span>     <span class="fu"><a href="../reference/filter_by_name.html">filter_by_name</a></span><span class="op">(</span></span>
<span>      mode <span class="op">=</span> <span class="st">'exclude'</span>, </span>
<span>      dimension<span class="op">=</span><span class="st">'sample'</span>,</span>
<span>      names <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'1358BZU_0001QC_H1'</span>,</span>
<span>                <span class="st">'1358BZU_0001QC_A1'</span>,</span>
<span>                <span class="st">'1358BZU_0001QC_G1'</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  </span>
<span>     <span class="fu"><a href="../reference/knn_impute.html">knn_impute</a></span><span class="op">(</span></span>
<span>      neighbours<span class="op">=</span><span class="fl">5</span><span class="op">)</span> <span class="op">+</span></span>
<span>  </span>
<span>     <span class="fu"><a href="../reference/vec_norm.html">vec_norm</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  </span>
<span>     <span class="fu"><a href="../reference/log_transform.html">log_transform</a></span><span class="op">(</span></span>
<span>       base <span class="op">=</span> <span class="fl">10</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  </span>
<span>     <span class="fu"><a href="../reference/mean_centre.html">mean_centre</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  </span>
<span>     <span class="fu"><a href="../reference/PCA.html">PCA</a></span><span class="op">(</span></span>
<span>       number_components <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># apply model sequence</span></span>
<span><span class="va">MS</span> <span class="op">=</span> <span class="fu"><a href="../reference/model_apply.html">model_apply</a></span><span class="op">(</span><span class="va">MS</span>, <span class="va">DE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Warning in knnimp(x, k, maxmiss = rowmax, maxp = maxp): 2 rows with more than 50 % entries missing;</span></span>
<span><span class="co">##  mean imputation used for these rows</span></span></code></pre>
<div class="sourceCode" id="cb122"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># PCA scores plots</span></span>
<span><span class="va">C</span> <span class="op">=</span> <span class="fu"><a href="../reference/pca_scores_plot.html">pca_scores_plot</a></span><span class="op">(</span>factor_name <span class="op">=</span> <span class="st">'sample_type'</span><span class="op">)</span></span>
<span><span class="co"># plot</span></span>
<span><span class="fu"><a href="../reference/chart_plot.html">chart_plot</a></span><span class="op">(</span><span class="va">C</span>,<span class="va">MS</span><span class="op">[</span><span class="fl">8</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<p><img src="data_analysis_omics_using_the_structtoolbox_files/figure-html/unnamed-chunk-81-1.png"><!-- -->
The QCs appear to representative of the samples, but there are strong
clusters in the data, including the QC samples which have no biological
variation. There is likely to be a number of ‘low quality’ features that
should be excluded, so we will do that now, and use more sophisticated
normalisation (PQN) and scaling methods (glog).</p>
<div class="sourceCode" id="cb123"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">MS</span> <span class="op">=</span>  <span class="fu"><a href="../reference/filter_smeta.html">filter_smeta</a></span><span class="op">(</span></span>
<span>       mode <span class="op">=</span> <span class="st">'exclude'</span>, </span>
<span>       levels <span class="op">=</span> <span class="st">'12'</span>, </span>
<span>       factor_name <span class="op">=</span> <span class="st">'biol.batch'</span><span class="op">)</span> <span class="op">+</span></span>
<span>  </span>
<span>      <span class="fu"><a href="../reference/filter_by_name.html">filter_by_name</a></span><span class="op">(</span></span>
<span>       mode <span class="op">=</span> <span class="st">'exclude'</span>, </span>
<span>       dimension<span class="op">=</span><span class="st">'sample'</span>,</span>
<span>       names <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'1358BZU_0001QC_H1'</span>,</span>
<span>                 <span class="st">'1358BZU_0001QC_A1'</span>,</span>
<span>                 <span class="st">'1358BZU_0001QC_G1'</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span></span>
<span>      <span class="fu"><a href="../reference/blank_filter.html">blank_filter</a></span><span class="op">(</span></span>
<span>       fold_change <span class="op">=</span> <span class="fl">20</span>,</span>
<span>       qc_label <span class="op">=</span> <span class="st">'QC'</span>,</span>
<span>       factor_name <span class="op">=</span> <span class="st">'sample_type'</span><span class="op">)</span> <span class="op">+</span></span>
<span></span>
<span>      <span class="fu"><a href="../reference/filter_smeta.html">filter_smeta</a></span><span class="op">(</span></span>
<span>       mode<span class="op">=</span><span class="st">'exclude'</span>,</span>
<span>       levels<span class="op">=</span><span class="st">'Blank'</span>,</span>
<span>       factor_name<span class="op">=</span><span class="st">'sample_type'</span><span class="op">)</span> <span class="op">+</span></span>
<span>  </span>
<span>      <span class="fu"><a href="../reference/mv_feature_filter.html">mv_feature_filter</a></span><span class="op">(</span></span>
<span>       threshold <span class="op">=</span> <span class="fl">80</span>, </span>
<span>       qc_label <span class="op">=</span> <span class="st">'QC'</span>, </span>
<span>       factor_name <span class="op">=</span> <span class="st">'sample_type'</span>, </span>
<span>       method <span class="op">=</span> <span class="st">'QC'</span><span class="op">)</span> <span class="op">+</span></span>
<span>     </span>
<span>      <span class="fu"><a href="../reference/mv_feature_filter.html">mv_feature_filter</a></span><span class="op">(</span></span>
<span>        threshold <span class="op">=</span> <span class="fl">50</span>, </span>
<span>        factor_name <span class="op">=</span> <span class="st">'sample_type'</span>, </span>
<span>        method<span class="op">=</span><span class="st">'across'</span><span class="op">)</span> <span class="op">+</span></span>
<span>  </span>
<span>     <span class="fu"><a href="../reference/rsd_filter.html">rsd_filter</a></span><span class="op">(</span></span>
<span>       rsd_threshold<span class="op">=</span><span class="fl">20</span>, </span>
<span>       qc_label<span class="op">=</span><span class="st">'QC'</span>,</span>
<span>       factor_name<span class="op">=</span><span class="st">'sample_type'</span><span class="op">)</span> <span class="op">+</span></span>
<span>  </span>
<span>     <span class="fu"><a href="../reference/mv_sample_filter.html">mv_sample_filter</a></span><span class="op">(</span></span>
<span>       mv_threshold <span class="op">=</span> <span class="fl">50</span><span class="op">)</span> <span class="op">+</span></span>
<span>     </span>
<span>     <span class="fu"><a href="../reference/pqn_norm.html">pqn_norm</a></span><span class="op">(</span></span>
<span>       qc_label<span class="op">=</span><span class="st">'QC'</span>,</span>
<span>       factor_name<span class="op">=</span><span class="st">'sample_type'</span><span class="op">)</span> <span class="op">+</span></span>
<span>     </span>
<span>     <span class="fu"><a href="../reference/knn_impute.html">knn_impute</a></span><span class="op">(</span></span>
<span>       neighbours<span class="op">=</span><span class="fl">5</span>, </span>
<span>       by<span class="op">=</span><span class="st">'samples'</span><span class="op">)</span> <span class="op">+</span></span>
<span>     </span>
<span>     <span class="fu"><a href="../reference/glog_transform.html">glog_transform</a></span><span class="op">(</span></span>
<span>       qc_label <span class="op">=</span> <span class="st">'QC'</span>,</span>
<span>       factor_name <span class="op">=</span> <span class="st">'sample_type'</span><span class="op">)</span> <span class="op">+</span></span>
<span>     </span>
<span>     <span class="fu"><a href="../reference/mean_centre.html">mean_centre</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>     </span>
<span>     <span class="fu"><a href="../reference/PCA.html">PCA</a></span><span class="op">(</span></span>
<span>       number_components <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># apply model sequence</span></span>
<span><span class="va">MS</span> <span class="op">=</span> <span class="fu"><a href="../reference/model_apply.html">model_apply</a></span><span class="op">(</span><span class="va">MS</span>, <span class="va">DE</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co"># PCA plots using different factors</span></span>
<span><span class="va">g</span><span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">k</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'order'</span>,<span class="st">'biol.batch'</span>,<span class="st">'time.point'</span>,<span class="st">'condition'</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">C</span> <span class="op">=</span> <span class="fu"><a href="../reference/pca_scores_plot.html">pca_scores_plot</a></span><span class="op">(</span>factor_name <span class="op">=</span> <span class="va">k</span>,ellipse<span class="op">=</span><span class="st">'none'</span><span class="op">)</span></span>
<span>  <span class="co"># plot</span></span>
<span>  <span class="va">g</span><span class="op">[[</span><span class="va">k</span><span class="op">]</span><span class="op">]</span><span class="op">=</span><span class="fu"><a href="../reference/chart_plot.html">chart_plot</a></span><span class="op">(</span><span class="va">C</span>,<span class="va">MS</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">MS</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html" class="external-link">plot_grid</a></span><span class="op">(</span>plotlist <span class="op">=</span> <span class="va">g</span>,align<span class="op">=</span><span class="st">'vh'</span>,axis<span class="op">=</span><span class="st">'tblr'</span>,nrow<span class="op">=</span><span class="fl">2</span>,labels<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'A'</span>,<span class="st">'B'</span>,<span class="st">'C'</span>,<span class="st">'D'</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="data_analysis_omics_using_the_structtoolbox_files/figure-html/unnamed-chunk-82-1.png"><!-- --></p>
<p>We can see now that the QCs are tightly clustered. This indicates
that the biological variance of the remaining high quality features is
much greater than the technical variance represented by the QCs.</p>
<p>There does not appear to be a trend by measurement order (A), which
is an important indicator that instrument drift throughout the run is
not a large source of variation in this dataset.</p>
<p>There does not appear to be strong clustering related to biological
batch (B).</p>
<p>There does not appear to be a strong trend with time (C) but this is
likely to be a more subtle variation and might be masked by other
sources of variance at this stage.</p>
<p>There is some clustering related to condition (D) but with
overlap.</p>
<p>To further explore any trends with time, we will split the data by
the condition factor and only explore the Ikaros group. Removing the
condition factor variation will potentially make it easier to spot any
more subtle trends. We will extract the glog transformed matrix from the
previous model sequence and continue from there.</p>
<div class="sourceCode" id="cb124"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># get the glog scaled data</span></span>
<span><span class="va">GL</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/struct/man/predicted.html" class="external-link">predicted</a></span><span class="op">(</span><span class="va">MS</span><span class="op">[</span><span class="fl">11</span><span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># extract the Ikaros group and apply PCA</span></span>
<span><span class="va">IK</span> <span class="op">=</span> <span class="fu"><a href="../reference/filter_smeta.html">filter_smeta</a></span><span class="op">(</span></span>
<span>      mode<span class="op">=</span><span class="st">'include'</span>,</span>
<span>      factor_name<span class="op">=</span><span class="st">'condition'</span>,</span>
<span>      levels<span class="op">=</span><span class="st">'Ikaros'</span><span class="op">)</span> <span class="op">+</span></span>
<span>     <span class="fu"><a href="../reference/mean_centre.html">mean_centre</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>     <span class="fu"><a href="../reference/PCA.html">PCA</a></span><span class="op">(</span>number_components <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># apply the model sequence to glog transformed data</span></span>
<span><span class="va">IK</span> <span class="op">=</span> <span class="fu"><a href="../reference/model_apply.html">model_apply</a></span><span class="op">(</span><span class="va">IK</span>,<span class="va">GL</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># plot the PCA scores</span></span>
<span><span class="va">C</span> <span class="op">=</span> <span class="fu"><a href="../reference/pca_scores_plot.html">pca_scores_plot</a></span><span class="op">(</span>factor_name<span class="op">=</span><span class="st">'time.point'</span>,ellipse <span class="op">=</span> <span class="st">'sample'</span><span class="op">)</span></span>
<span><span class="va">g1</span><span class="op">=</span><span class="fu"><a href="../reference/chart_plot.html">chart_plot</a></span><span class="op">(</span><span class="va">C</span>,<span class="va">IK</span><span class="op">[</span><span class="fl">3</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="va">g2</span><span class="op">=</span><span class="va">g1</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_viridis.html" class="external-link">scale_color_viridis_d</a></span><span class="op">(</span><span class="op">)</span> <span class="co"># add continuous scale colouring</span></span>
<span></span>
<span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html" class="external-link">plot_grid</a></span><span class="op">(</span><span class="va">g1</span>,<span class="va">g2</span>,nrow<span class="op">=</span><span class="fl">2</span>,align<span class="op">=</span><span class="st">'vh'</span>,axis <span class="op">=</span> <span class="st">'tblr'</span>,labels<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'A'</span>,<span class="st">'B'</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="data_analysis_omics_using_the_structtoolbox_files/figure-html/unnamed-chunk-83-1.png"><!-- --></p>
<p>Colouring by groups (A) makes the time point trend difficult to see,
but by adding a <code>ggplot</code> continuous colour scale “viridis”
(B) the trend with time along PC1 becomes much clearer.</p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="session-info">Session Info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<div class="sourceCode" id="cb125"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R Under development (unstable) (2024-01-23 r85822)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">## Running under: Ubuntu 22.04.3 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span></span>
<span><span class="co">##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: UTC</span></span>
<span><span class="co">## tzcode source: system (glibc)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">##  [1] openxlsx_4.2.5.2     cowplot_1.1.3        gridExtra_2.3       </span></span>
<span><span class="co">##  [4] ggplot2_3.4.4        BiocFileCache_2.11.1 dbplyr_2.4.0        </span></span>
<span><span class="co">##  [7] ropls_1.35.4         pmp_1.15.0           structToolbox_1.15.1</span></span>
<span><span class="co">## [10] struct_1.15.3        BiocStyle_2.31.0    </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] DBI_1.2.1                   bitops_1.0-7               </span></span>
<span><span class="co">##   [3] MultiDataSet_1.31.0         rlang_1.1.3                </span></span>
<span><span class="co">##   [5] magrittr_2.0.3              e1071_1.7-14               </span></span>
<span><span class="co">##   [7] matrixStats_1.2.0           compiler_4.4.0             </span></span>
<span><span class="co">##   [9] RSQLite_2.3.5               systemfonts_1.0.5          </span></span>
<span><span class="co">##  [11] vctrs_0.6.5                 reshape2_1.4.4             </span></span>
<span><span class="co">##  [13] stringr_1.5.1               pkgconfig_2.0.3            </span></span>
<span><span class="co">##  [15] crayon_1.5.2                fastmap_1.1.1              </span></span>
<span><span class="co">##  [17] XVector_0.43.1              labeling_0.4.3             </span></span>
<span><span class="co">##  [19] utf8_1.2.4                  rmarkdown_2.25             </span></span>
<span><span class="co">##  [21] itertools_0.1-3             ragg_1.2.7                 </span></span>
<span><span class="co">##  [23] bit_4.0.5                   purrr_1.0.2                </span></span>
<span><span class="co">##  [25] xfun_0.41                   MultiAssayExperiment_1.29.0</span></span>
<span><span class="co">##  [27] randomForest_4.7-1.1        zlibbioc_1.49.0            </span></span>
<span><span class="co">##  [29] cachem_1.0.8                GenomeInfoDb_1.39.5        </span></span>
<span><span class="co">##  [31] jsonlite_1.8.8              progress_1.2.3             </span></span>
<span><span class="co">##  [33] blob_1.2.4                  highr_0.10                 </span></span>
<span><span class="co">##  [35] DelayedArray_0.29.1         prettyunits_1.2.0          </span></span>
<span><span class="co">##  [37] parallel_4.4.0              rols_2.31.0                </span></span>
<span><span class="co">##  [39] R6_2.5.1                    bslib_0.6.1                </span></span>
<span><span class="co">##  [41] stringi_1.8.3               limma_3.59.1               </span></span>
<span><span class="co">##  [43] GenomicRanges_1.55.2        jquerylib_0.1.4            </span></span>
<span><span class="co">##  [45] Rcpp_1.0.12                 bookdown_0.37              </span></span>
<span><span class="co">##  [47] SummarizedExperiment_1.33.3 iterators_1.0.14           </span></span>
<span><span class="co">##  [49] knitr_1.45                  IRanges_2.37.1             </span></span>
<span><span class="co">##  [51] Matrix_1.6-5                tidyselect_1.2.0           </span></span>
<span><span class="co">##  [53] abind_1.4-5                 yaml_2.3.8                 </span></span>
<span><span class="co">##  [55] codetools_0.2-19            curl_5.2.0                 </span></span>
<span><span class="co">##  [57] doRNG_1.8.6                 lattice_0.22-5             </span></span>
<span><span class="co">##  [59] tibble_3.2.1                plyr_1.8.9                 </span></span>
<span><span class="co">##  [61] withr_3.0.0                 Biobase_2.63.0             </span></span>
<span><span class="co">##  [63] evaluate_0.23               ontologyIndex_2.11         </span></span>
<span><span class="co">##  [65] desc_1.4.3                  isoband_0.2.7              </span></span>
<span><span class="co">##  [67] proxy_0.4-27                zip_2.3.1                  </span></span>
<span><span class="co">##  [69] filelock_1.0.3              pillar_1.9.0               </span></span>
<span><span class="co">##  [71] BiocManager_1.30.22         MatrixGenerics_1.15.0      </span></span>
<span><span class="co">##  [73] rngtools_1.5.2              foreach_1.5.2              </span></span>
<span><span class="co">##  [75] stats4_4.4.0                generics_0.1.3             </span></span>
<span><span class="co">##  [77] sp_2.1-3                    RCurl_1.98-1.14            </span></span>
<span><span class="co">##  [79] hms_1.1.3                   S4Vectors_0.41.3           </span></span>
<span><span class="co">##  [81] munsell_0.5.0               scales_1.3.0               </span></span>
<span><span class="co">##  [83] calibrate_1.7.7             class_7.3-22               </span></span>
<span><span class="co">##  [85] glue_1.7.0                  tools_4.4.0                </span></span>
<span><span class="co">##  [87] fs_1.6.3                    grid_4.4.0                 </span></span>
<span><span class="co">##  [89] impute_1.77.0               missForest_1.5             </span></span>
<span><span class="co">##  [91] colorspace_2.1-0            GenomeInfoDbData_1.2.11    </span></span>
<span><span class="co">##  [93] cli_3.6.2                   textshaping_0.3.7          </span></span>
<span><span class="co">##  [95] fansi_1.0.6                 viridisLite_0.4.2          </span></span>
<span><span class="co">##  [97] ggthemes_5.0.0              S4Arrays_1.3.3             </span></span>
<span><span class="co">##  [99] dplyr_1.1.4                 pls_2.8-3                  </span></span>
<span><span class="co">## [101] pcaMethods_1.95.0           gtable_0.3.4               </span></span>
<span><span class="co">## [103] sass_0.4.8                  digest_0.6.34              </span></span>
<span><span class="co">## [105] BiocGenerics_0.49.1         SparseArray_1.3.3          </span></span>
<span><span class="co">## [107] farver_2.1.1                memoise_2.0.1              </span></span>
<span><span class="co">## [109] htmltools_0.5.7             pkgdown_2.0.7.9000         </span></span>
<span><span class="co">## [111] lifecycle_1.0.4             httr_1.4.7                 </span></span>
<span><span class="co">## [113] statmod_1.5.0               bit64_4.0.5                </span></span>
<span><span class="co">## [115] qqman_0.1.9                 MASS_7.3-60.2</span></span></code></pre>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Gavin Rhys Lloyd, Ralf Johannes Maria Weber.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.9000.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
